{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a39f3c65",
   "metadata": {},
   "source": [
    "<h1>WFS1: Strom et al. (1998)</h1>\n",
    "<p>We will process <a href=\"https://pubmed.ncbi.nlm.nih.gov/9817917/\" target=\"__blank\">Strom, et al. (1998) Diabetes insipidus, diabetes mellitus, optic atrophy and deafness (DIDMOAD) caused by mutations in a novel gene (wolframin) coding for a predicted transmembrane protein</a></p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f0602c0e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pyphetools version 0.6.4\n"
     ]
    }
   ],
   "source": [
    "import phenopackets as php\n",
    "from google.protobuf.json_format import MessageToDict, MessageToJson\n",
    "from google.protobuf.json_format import Parse, ParseDict\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_colwidth', None) # show entire column contents, important!\n",
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "from pyphetools.creation import *\n",
    "from pyphetools.visualization import *\n",
    "import pyphetools\n",
    "print(f\"pyphetools version {pyphetools.__version__}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8606e7eb",
   "metadata": {},
   "source": [
    "<h2>Importing HPO data</h2>\n",
    "<p>pyphetools uses the Human Phenotype Ontology (HPO) to encode phenotypic features. The recommended way of doing this is to ingest the hp.json file using HpoParser, which in turn creates an HpoConceptRecognizer object. </p>\n",
    "<p>The HpoParser can accept a hpo_json_file argument if you want to use a specific file. If the argument is not passed, it will download the latext hp.json file from the HPO GitHub site and store it in a new subdirectory called hpo_data. It will not download the file if the file is already downloaded.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5a7789fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = HpoParser()\n",
    "hpo_cr = parser.get_hpo_concept_recognizer()\n",
    "hpo_version = parser.get_version()\n",
    "PMID = \"PMID:9817917\"\n",
    "title = \"Diabetes insipidus, diabetes mellitus, optic atrophy and deafness (DIDMOAD) caused by mutations in a novel gene (wolframin) coding for a predicted transmembrane protein\"\n",
    "metadata = MetaData(created_by=\"ORCID:0000-0002-5648-2155\")\n",
    "metadata.default_versions_with_hpo(version=hpo_version)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5102722a",
   "metadata": {},
   "source": [
    "<h2>Importing the supplemental table</h2>\n",
    "<p>Here, we use the pandas library to import this file (note that the Python package called openpyxl must be installed to read Excel files with pandas, although the library does not need to be imported in this notebook). pyphetools expects a pandas DataFrame as input, and users can choose any input format available for pandas include CSV, TSV, and Excel, or can use any other method to transform their input data into a Pandas DataFrame before using pyphetools.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f17c3924",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel('../../data/WFS1/PMID_9817917.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "06ec7dbd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Family</th>\n",
       "      <th>Patient</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Diabetes mellitus</th>\n",
       "      <th>Progressive optic atrophy</th>\n",
       "      <th>Hearing impairment</th>\n",
       "      <th>Diabetes insipidus</th>\n",
       "      <th>Abnormality of the kidney</th>\n",
       "      <th>Neurological abnormalities</th>\n",
       "      <th>Other complications</th>\n",
       "      <th>Consangui</th>\n",
       "      <th>Variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5519</td>\n",
       "      <td>f</td>\n",
       "      <td>22.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>Ataxia, nystagmus</td>\n",
       "      <td>Retarded sexual maturation, depression</td>\n",
       "      <td>-</td>\n",
       "      <td>1380del9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>13883</td>\n",
       "      <td>f</td>\n",
       "      <td>11.0</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>1380del9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.0</td>\n",
       "      <td>13775</td>\n",
       "      <td>f</td>\n",
       "      <td>20.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>460+1G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.0</td>\n",
       "      <td>13776</td>\n",
       "      <td>m</td>\n",
       "      <td>17.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>Retarded sexual maturation</td>\n",
       "      <td>-</td>\n",
       "      <td>460+1G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4.0</td>\n",
       "      <td>13070</td>\n",
       "      <td>f</td>\n",
       "      <td>22.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>Abnormal EEG</td>\n",
       "      <td>Psychiatric illness</td>\n",
       "      <td>-</td>\n",
       "      <td>599delT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5.0</td>\n",
       "      <td>13885</td>\n",
       "      <td>f</td>\n",
       "      <td>35.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>Cataract</td>\n",
       "      <td>-</td>\n",
       "      <td>1096C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6.0</td>\n",
       "      <td>13062</td>\n",
       "      <td>f</td>\n",
       "      <td>25.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>Ataxia, nystagmus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>676C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7.0</td>\n",
       "      <td>13076</td>\n",
       "      <td>m</td>\n",
       "      <td>26.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Retarded sexual maturation, mental retardation</td>\n",
       "      <td>-</td>\n",
       "      <td>599delT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8.0</td>\n",
       "      <td>13073</td>\n",
       "      <td>f</td>\n",
       "      <td>35.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>Ataxia</td>\n",
       "      <td>Cataract, psychiatric illness, ragged red fibers</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1096C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9.0</td>\n",
       "      <td>13781</td>\n",
       "      <td>m</td>\n",
       "      <td>19.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>Abnormal EEG</td>\n",
       "      <td>Retarded sexual maturation</td>\n",
       "      <td>+</td>\n",
       "      <td>1558C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10.0</td>\n",
       "      <td>13782</td>\n",
       "      <td>m</td>\n",
       "      <td>16.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11.0</td>\n",
       "      <td>13783</td>\n",
       "      <td>f</td>\n",
       "      <td>12.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Ataxia</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12.0</td>\n",
       "      <td>12131</td>\n",
       "      <td>m</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2164ins24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NaN</td>\n",
       "      <td>12132</td>\n",
       "      <td>m</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2164ins24</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Family  Patient  ... Consangui    Variant\n",
       "0      1.0     5519  ...         -   1380del9\n",
       "1      1.0    13883  ...         -   1380del9\n",
       "2      2.0    13775  ...         -   460+1G>A\n",
       "3      2.0    13776  ...         -   460+1G>A\n",
       "4      4.0    13070  ...         -    599delT\n",
       "5      5.0    13885  ...         -    1096C>T\n",
       "6      6.0    13062  ...       NaN     676C>T\n",
       "7      7.0    13076  ...         -    599delT\n",
       "8      8.0    13073  ...       NaN    1096C>T\n",
       "9      9.0    13781  ...         +    1558C>T\n",
       "10    10.0    13782  ...         -        NaN\n",
       "11    11.0    13783  ...       NaN        NaN\n",
       "12    12.0    12131  ...       NaN  2164ins24\n",
       "13     NaN    12132  ...       NaN  2164ins24\n",
       "\n",
       "[14 rows x 13 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e013666d",
   "metadata": {},
   "source": [
    "Some column names might include spaces in front or after, and a couple of columns are subheadings and only contain NaNs, so lets correct that. Furthermore, remove individuals without an age specified or a variant in this gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5d604600",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Family</th>\n",
       "      <th>Patient</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Diabetes mellitus</th>\n",
       "      <th>Progressive optic atrophy</th>\n",
       "      <th>Hearing impairment</th>\n",
       "      <th>Diabetes insipidus</th>\n",
       "      <th>Abnormality of the kidney</th>\n",
       "      <th>Neurological abnormalities</th>\n",
       "      <th>Other complications</th>\n",
       "      <th>Consangui</th>\n",
       "      <th>Variant</th>\n",
       "      <th>patient_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5519</td>\n",
       "      <td>f</td>\n",
       "      <td>22.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>Ataxia, nystagmus</td>\n",
       "      <td>Retarded sexual maturation, depression</td>\n",
       "      <td>-</td>\n",
       "      <td>1380del9</td>\n",
       "      <td>5519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>13883</td>\n",
       "      <td>f</td>\n",
       "      <td>11.0</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>1380del9</td>\n",
       "      <td>13883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.0</td>\n",
       "      <td>13775</td>\n",
       "      <td>f</td>\n",
       "      <td>20.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>460+1G&gt;A</td>\n",
       "      <td>13775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.0</td>\n",
       "      <td>13776</td>\n",
       "      <td>m</td>\n",
       "      <td>17.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>Retarded sexual maturation</td>\n",
       "      <td>-</td>\n",
       "      <td>460+1G&gt;A</td>\n",
       "      <td>13776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4.0</td>\n",
       "      <td>13070</td>\n",
       "      <td>f</td>\n",
       "      <td>22.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>Abnormal EEG</td>\n",
       "      <td>Psychiatric illness</td>\n",
       "      <td>-</td>\n",
       "      <td>599delT</td>\n",
       "      <td>13070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5.0</td>\n",
       "      <td>13885</td>\n",
       "      <td>f</td>\n",
       "      <td>35.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>Cataract</td>\n",
       "      <td>-</td>\n",
       "      <td>1096C&gt;T</td>\n",
       "      <td>13885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6.0</td>\n",
       "      <td>13062</td>\n",
       "      <td>f</td>\n",
       "      <td>25.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>Ataxia, nystagmus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>676C&gt;T</td>\n",
       "      <td>13062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7.0</td>\n",
       "      <td>13076</td>\n",
       "      <td>m</td>\n",
       "      <td>26.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Retarded sexual maturation, mental retardation</td>\n",
       "      <td>-</td>\n",
       "      <td>599delT</td>\n",
       "      <td>13076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8.0</td>\n",
       "      <td>13073</td>\n",
       "      <td>f</td>\n",
       "      <td>35.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>Ataxia</td>\n",
       "      <td>Cataract, psychiatric illness, ragged red fibers</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1096C&gt;T</td>\n",
       "      <td>13073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9.0</td>\n",
       "      <td>13781</td>\n",
       "      <td>m</td>\n",
       "      <td>19.0</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>Abnormal EEG</td>\n",
       "      <td>Retarded sexual maturation</td>\n",
       "      <td>+</td>\n",
       "      <td>1558C&gt;T</td>\n",
       "      <td>13781</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Family  Patient Sex  ...  Consangui    Variant patient_id\n",
       "0     1.0     5519   f  ...          -   1380del9       5519\n",
       "1     1.0    13883   f  ...          -   1380del9      13883\n",
       "2     2.0    13775   f  ...          -   460+1G>A      13775\n",
       "3     2.0    13776   m  ...          -   460+1G>A      13776\n",
       "4     4.0    13070   f  ...          -    599delT      13070\n",
       "5     5.0    13885   f  ...          -    1096C>T      13885\n",
       "6     6.0    13062   f  ...        NaN     676C>T      13062\n",
       "7     7.0    13076   m  ...          -    599delT      13076\n",
       "8     8.0    13073   f  ...        NaN    1096C>T      13073\n",
       "9     9.0    13781   m  ...          +    1558C>T      13781\n",
       "\n",
       "[10 rows x 14 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns = df.columns.str.strip()\n",
    "df = df.dropna(axis=1, how='all')\n",
    "df['patient_id'] = df['Patient']\n",
    "df = df[~df['Age'].isna()]\n",
    "df = df[~df['Variant'].isna()]\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8314d6bd",
   "metadata": {},
   "source": [
    "<h2>Column mappers</h2>\n",
    "<p>Please see the notebook \"Create phenopackets from tabular data with individuals in rows\" for explanations. In the following cell we create a dictionary for the ColumnMappers. Note that the code is identical except that we use the df.loc function to get the corresponding row data</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "59d9e9e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "hpo_cr = parser.get_hpo_concept_recognizer()\n",
    "generator = SimpleColumnMapperGenerator(df=df, observed='+', excluded='-', hpo_cr=hpo_cr)\n",
    "column_mapper_d = generator.try_mapping_columns()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "122946b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table style=\"border: 2px solid black;\">\n",
       "\n",
       "<tr>\n",
       "            <th>Result</th>\n",
       "            <th>Columns</th>\n",
       "        </tr>\n",
       "      \n",
       "<tr><td>Mapped</td><td>Diabetes mellitus; Hearing impairment; Diabetes insipidus; Abnormality of the kidney</td></tr>\n",
       "<tr><td>Unmapped</td><td><q>Family</q>; <q>Patient</q>; <q>Sex</q>; <q>Age</q>; <q>Progressive optic atrophy</q>; <q>Neurological abnormalities</q>; <q>Other complications</q>; <q>Consangui</q>; <q>Variant</q>; <q>patient_id</q></td></tr>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import HTML, display\n",
    "display(HTML(generator.to_html()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6daa46a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "neurological = {'Ataxia': 'Ataxia',\n",
    "                 'nystagmus': 'Nystagmus',\n",
    "               'Abnormal EEG': 'EEG abnormality'}\n",
    "neurologicalMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=neurological)\n",
    "neurologicalMapper.preview_column(df['Neurological abnormalities'])\n",
    "column_mapper_d['Neurological abnormalities'] = neurologicalMapper\n",
    "\n",
    "other = {'Retarded sexual maturation': 'Puberty and gonadal disorders',\n",
    "                 'depression': 'Depression',\n",
    "               'psychiatric illness': 'Behavioral abnormality',\n",
    "                'cataract' : 'Cataract', \n",
    "         'mental retardation': 'Intellectual disability',\n",
    "         'ragged red fibers': 'Ragged-red muscle fibers'\n",
    "        }\n",
    "otherMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=other)\n",
    "otherMapper.preview_column(df['Other complications'])\n",
    "column_mapper_d['Other complications'] = otherMapper"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e53617d",
   "metadata": {},
   "source": [
    "<h2>Variant Data</h2>\n",
    "<p>The variant data (HGVS< transcript) is listed in the Variant (hg19, NM_015133.4) column.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "da4d5706",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['1380del9' ' 460+1G>A' ' 599delT' ' 1096C>T' '676C>T' '599delT' '1096C>T'\n",
      " '1558C>T']\n",
      "1380del9 - c.1385_1393del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_006005.3%3Ac.1385_1393del/NM_006005.3?content-type=application%2Fjson\n",
      "1380del9: chr4:6301174CCACCGAGGT>C\n",
      " 460+1G>A - c. 460+1G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_006005.3%3Ac. 460+1G>A/NM_006005.3?content-type=application%2Fjson\n",
      " 460+1G>A: chr4:6289132G>A\n",
      " 599delT - c. 599delT\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_006005.3%3Ac. 599delT/NM_006005.3?content-type=application%2Fjson\n",
      " 599delT: chr4:6291334CT>C\n",
      " 1096C>T - c. 1096C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_006005.3%3Ac. 1096C>T/NM_006005.3?content-type=application%2Fjson\n",
      " 1096C>T: chr4:6300891C>T\n",
      "676C>T - c.676C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_006005.3%3Ac.676C>T/NM_006005.3?content-type=application%2Fjson\n",
      "676C>T: chr4:6291961C>T\n",
      "599delT - c.599delT\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_006005.3%3Ac.599delT/NM_006005.3?content-type=application%2Fjson\n",
      "599delT: chr4:6291334CT>C\n",
      "1096C>T - c.1096C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_006005.3%3Ac.1096C>T/NM_006005.3?content-type=application%2Fjson\n",
      "1096C>T: chr4:6300891C>T\n",
      "1558C>T - c.1558C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_006005.3%3Ac.1558C>T/NM_006005.3?content-type=application%2Fjson\n",
      "1558C>T: chr4:6301353C>T\n",
      "Extracted 8 unique variants\n"
     ]
    }
   ],
   "source": [
    "hg38 = 'hg38'\n",
    "default_genotype = 'heterozygous'\n",
    "WFS1_transcript='NM_006005.3'\n",
    "vvalidator = VariantValidator(genome_build=hg38, transcript=WFS1_transcript)\n",
    "variant_list = df['Variant'].unique()\n",
    "print(variant_list)\n",
    "variant_d = {}\n",
    "for v in variant_list:\n",
    "    if v == \"1380del9\":\n",
    "        hgvs = \"c.1385_1393del\"\n",
    "    else:\n",
    "        hgvs = f\"c.{v}\"\n",
    "    print(f\"{v} - {hgvs}\")\n",
    "    var = vvalidator.encode_hgvs(hgvs)\n",
    "    print(f\"{v}: {var}\")\n",
    "    variant_d[v] = var\n",
    "print(f\"Extracted {len(variant_d)} unique variants\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "928eeb81-5116-4cf9-972c-820347802a26",
   "metadata": {},
   "outputs": [],
   "source": [
    "varMapper = VariantColumnMapper(variant_d=variant_d,variant_column_name='Variant', default_genotype=default_genotype)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1af2052",
   "metadata": {},
   "source": [
    "<h1>Demographic data</h1>\n",
    "<p>pyphetools can be used to capture information about age, sex, and individual identifiers. This information is stored in a map of \"IndividualMapper\" objects. Special treatment may be required for the indifiers, which may be used as the column names or row index.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3e64dc08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original column contents</th>\n",
       "      <th>age</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22.0</td>\n",
       "      <td>P22Y0M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11.0</td>\n",
       "      <td>P11Y0M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20.0</td>\n",
       "      <td>P20Y0M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>17.0</td>\n",
       "      <td>P17Y0M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>35.0</td>\n",
       "      <td>P35Y0M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>25.0</td>\n",
       "      <td>P25Y0M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>26.0</td>\n",
       "      <td>P26Y0M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>19.0</td>\n",
       "      <td>P19Y0M</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   original column contents     age\n",
       "0                      22.0  P22Y0M\n",
       "1                      11.0  P11Y0M\n",
       "2                      20.0  P20Y0M\n",
       "3                      17.0  P17Y0M\n",
       "4                      35.0  P35Y0M\n",
       "5                      25.0  P25Y0M\n",
       "6                      26.0  P26Y0M\n",
       "7                      19.0  P19Y0M"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ageMapper = AgeColumnMapper.by_year('Age')\n",
    "ageMapper.preview_column(df['Age'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "71f664cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original column contents</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>f</td>\n",
       "      <td>FEMALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>f</td>\n",
       "      <td>FEMALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f</td>\n",
       "      <td>FEMALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>m</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>f</td>\n",
       "      <td>FEMALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>f</td>\n",
       "      <td>FEMALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>f</td>\n",
       "      <td>FEMALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>m</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>f</td>\n",
       "      <td>FEMALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>m</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  original column contents     sex\n",
       "0                        f  FEMALE\n",
       "1                        f  FEMALE\n",
       "2                        f  FEMALE\n",
       "3                        m    MALE\n",
       "4                        f  FEMALE\n",
       "5                        f  FEMALE\n",
       "6                        f  FEMALE\n",
       "7                        m    MALE\n",
       "8                        f  FEMALE\n",
       "9                        m    MALE"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sexMapper = SexColumnMapper(male_symbol='m', female_symbol='f', column_name='Sex')\n",
    "sexMapper.preview_column(df['Sex'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f6581a8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "encoder = CohortEncoder(df=df, hpo_cr=hpo_cr, column_mapper_d=column_mapper_d, \n",
    "                        individual_column_name=\"patient_id\", agemapper=ageMapper, sexmapper=sexMapper,\n",
    "                       variant_mapper=varMapper, metadata=metadata,\n",
    "                       pmid=PMID)\n",
    "encoder.set_disease(disease_id='OMIM:222300', label='Wolfram syndrome 1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "fd367ed6",
   "metadata": {},
   "outputs": [],
   "source": [
    "individuals = encoder.get_individuals()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5d044b78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table style=\"border: 2px solid black;\">\n",
       "\n",
       "<caption>10 phenopackets - no PMIDs (consider adding this information to the MetaData)</caption>\n",
       "\n",
       "<tr>\n",
       "            <th>Individual</th>\n",
       "            <th>Disease</th>\n",
       "            <th>Genotype</th>\n",
       "            <th>Phenotypic features</th>\n",
       "        </tr>\n",
       "        \n",
       "<tr>\n",
       "<td>5519 (FEMALE; P22Y0M)</ts>\n",
       "<td>Wolfram syndrome 1 (OMIM:222300)</ts>\n",
       "<td>NM_006005.3:c.1385_1393del (heterozygous)</td>\n",
       "<td class=\"table-data\">Diabetes mellitus (HP:0000819); Hearing impairment (HP:0000365); Diabetes insipidus (HP:0000873); Abnormality of the kidney (HP:0000077); Ataxia (HP:0001251); Nystagmus (HP:0000639); Depression (HP:0000716); Puberty and gonadal disorders (HP:0008373)</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>13883 (FEMALE; P11Y0M)</ts>\n",
       "<td>Wolfram syndrome 1 (OMIM:222300)</ts>\n",
       "<td>NM_006005.3:c.1385_1393del (heterozygous)</td>\n",
       "<td class=\"table-data\">Diabetes mellitus (HP:0000819); Abnormality of the kidney (HP:0000077)</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>13775 (FEMALE; P20Y0M)</ts>\n",
       "<td>Wolfram syndrome 1 (OMIM:222300)</ts>\n",
       "<td>NM_006005.3:c.460+1G>A (heterozygous)</td>\n",
       "<td class=\"table-data\">Diabetes mellitus (HP:0000819); Hearing impairment (HP:0000365); Diabetes insipidus (HP:0000873)</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>13776 (MALE; P17Y0M)</ts>\n",
       "<td>Wolfram syndrome 1 (OMIM:222300)</ts>\n",
       "<td>NM_006005.3:c.460+1G>A (heterozygous)</td>\n",
       "<td class=\"table-data\">Diabetes mellitus (HP:0000819); Hearing impairment (HP:0000365); Diabetes insipidus (HP:0000873); Abnormality of the kidney (HP:0000077); Puberty and gonadal disorders (HP:0008373)</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>13070 (FEMALE; P22Y0M)</ts>\n",
       "<td>Wolfram syndrome 1 (OMIM:222300)</ts>\n",
       "<td>NM_006005.3:c.599del (heterozygous)</td>\n",
       "<td class=\"table-data\">Diabetes mellitus (HP:0000819); Hearing impairment (HP:0000365); Abnormality of the kidney (HP:0000077); EEG abnormality (HP:0002353)</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>13885 (FEMALE; P35Y0M)</ts>\n",
       "<td>Wolfram syndrome 1 (OMIM:222300)</ts>\n",
       "<td>NM_006005.3:c.1096C>T (heterozygous)</td>\n",
       "<td class=\"table-data\">Diabetes mellitus (HP:0000819); Hearing impairment (HP:0000365); Abnormality of the kidney (HP:0000077)</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>13062 (FEMALE; P25Y0M)</ts>\n",
       "<td>Wolfram syndrome 1 (OMIM:222300)</ts>\n",
       "<td>NM_006005.3:c.676C>T (heterozygous)</td>\n",
       "<td class=\"table-data\">Diabetes mellitus (HP:0000819); Hearing impairment (HP:0000365); Diabetes insipidus (HP:0000873); Abnormality of the kidney (HP:0000077); Ataxia (HP:0001251); Nystagmus (HP:0000639)</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>13076 (MALE; P26Y0M)</ts>\n",
       "<td>Wolfram syndrome 1 (OMIM:222300)</ts>\n",
       "<td>NM_006005.3:c.599del (heterozygous)</td>\n",
       "<td class=\"table-data\">Diabetes mellitus (HP:0000819); Hearing impairment (HP:0000365); Diabetes insipidus (HP:0000873); Intellectual disability (HP:0001249); Puberty and gonadal disorders (HP:0008373)</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>13073 (FEMALE; P35Y0M)</ts>\n",
       "<td>Wolfram syndrome 1 (OMIM:222300)</ts>\n",
       "<td>NM_006005.3:c.1096C>T (heterozygous)</td>\n",
       "<td class=\"table-data\">Diabetes mellitus (HP:0000819); Hearing impairment (HP:0000365); Abnormality of the kidney (HP:0000077); Ataxia (HP:0001251); Behavioral abnormality (HP:0000708); Ragged-red muscle fibers (HP:0003200)</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>13781 (MALE; P19Y0M)</ts>\n",
       "<td>Wolfram syndrome 1 (OMIM:222300)</ts>\n",
       "<td>NM_006005.3:c.1558C>T (heterozygous)</td>\n",
       "<td class=\"table-data\">Diabetes mellitus (HP:0000819); Hearing impairment (HP:0000365); Diabetes insipidus (HP:0000873); Abnormality of the kidney (HP:0000077); EEG abnormality (HP:0002353); Puberty and gonadal disorders (HP:0008373)</td>\n",
       "</tr>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "phenopackets = [i.to_ga4gh_phenopacket(metadata=metadata.to_ga4gh()) for i in individuals]\n",
    "table = PhenopacketTable(phenopacket_list=phenopackets)\n",
    "display(HTML(table.to_html()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "23f1094f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We output 10 GA4GH phenopackets to the directory phenopackets\n"
     ]
    }
   ],
   "source": [
    "output_directory = \"phenopackets\"\n",
    "Individual.output_individuals_as_phenopackets(individual_list=individuals,\n",
    "                                             pmid=PMID,\n",
    "                                             metadata=metadata.to_ga4gh(),\n",
    "                                             outdir=output_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20cce252-6f66-4f14-9d9c-b3f96046f969",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
