{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "14ee3127",
   "metadata": {},
   "source": [
    "<h1>ERI1: Guo et al 2013 </h1>\n",
    "<p>Extract the clinical data from <a href=\"https://pubmed.ncbi.nlm.nih.gov/37352860/\"target=\"__blank\">Guo L, et al. (2023) Null and missense mutations of ERI1 cause a recessive phenotypic dichotomy in humans. Am J Hum Genet.  PMID:37352860</a>.<p>\n",
    "<p>The authors report a phenotypic dichotomy associated with bi-allelic ERI1 variants by reporting eight affected individuals from seven unrelated families. A severe spondyloepimetaphyseal dysplasia (SEMD) was identified in five affected individuals with missense variants but not in those with bi-allelic null variants, who showed mild intellectual disability and digital anomalies.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3d9c8a7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import phenopackets as php\n",
    "from google.protobuf.json_format import MessageToDict, MessageToJson\n",
    "from google.protobuf.json_format import Parse, ParseDict\n",
    "import pandas as pd\n",
    "import math\n",
    "from csv import DictReader\n",
    "pd.set_option('display.max_colwidth', None) # show entire column contents, important!\n",
    "from collections import defaultdict\n",
    "import re\n",
    "from pyphetools.creation import *\n",
    "from pyphetools.output import PhenopacketTable\n",
    "# last tested with pyphetools version 0.4.6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "3d06141b",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = HpoParser()\n",
    "hpo_cr = parser.get_hpo_concept_recognizer()\n",
    "hpo_version = parser.get_version()\n",
    "metadata = MetaData(created_by=\"ORCID:0000-0002-0736-9199\")\n",
    "metadata.default_versions_with_hpo(version=hpo_version)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "1f2ca0a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel(\"input/Guo_2023.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "5c8b8006",
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_id = df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "771713c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.set_index('Individual').T.reset_index()\n",
    "df[\"patient_id\"] = df[\"index\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "9eefb904",
   "metadata": {},
   "outputs": [],
   "source": [
    "scg = SimpleColumnMapperGenerator(df=df,\n",
    "                                   observed='+',\n",
    "                                   excluded='-',\n",
    "                                  hpo_cr=hpo_cr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "2c585ebc",
   "metadata": {},
   "outputs": [],
   "source": [
    "column_mapper_d = scg.try_mapping_columns()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "a1f215be",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_171705/3642770465.py:1: DeprecationWarning: Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display\n",
      "  from IPython.core.display import display, HTML\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table style=\"border: 2px solid black;\">\n",
       "\n",
       "<tr>\n",
       "            <th>Result</th>\n",
       "            <th>Columns</th>\n",
       "        </tr>\n",
       "      \n",
       "<tr><td>Mapped</td><td>Syndactyly; Cardiac anomaly; Hydronephrosis; Vesicoureteral reflux; Asthma; Conductive hearing impairment; hypernasal speech; Dislocated radial head; Scoliosis; Hip pain; Short stature; Long face; Narrow face; proptosis; Coarse facies; Low-set ears; Limited elbow extension; Finger joint hypermobility; Clinodactyly of the 5th finger; Pes planus; Slender metacarpals; Increased vertebral height; Velopharyngeal insufficiency; Hip dislocation; Patellar dislocation; Narrow forehead; Upslanted palpebral fissure; High palate; Pectus excavatum; Tapered finger; Prominent forehead; Depressed nasal bridge; Micrognathia; Cutaneous syndactyly; Macrotia; Narrow chest; Pulmonary arterial hypertension; Oligodactyly; Tricuspid regurgitation; Platyspondyly; Intrauterine growth retardation; Motor delay; Failure to thrive; Trigonocephaly; Frontal bossing; Sparse hair; Pectus carinatum; Wormian bones; Osteopenia; Delayed skeletal maturation; Inguinal hernia; Ventricular septal defect; Brachycephaly; Anonychia; Strabismus; Low anterior hairline; Epicanthus</td></tr>\n",
       "<tr><td>Unmapped</td><td><q>index</q>; <q>DNA</q>; <q>Protein</q>; <q>Sex</q>; <q>Age at last follow-up</q>; <q>Weight</q>; <q>Height</q>; <q>Consanguinity</q>; <q>Fetal ultrasound</q>; <q>Gestation age</q>; <q>Birth weight</q>; <q>Birth length</q>; <q>Spine anomaly</q>; <q>Metaphyseal anomaly</q>; <q>Epiphyseal anomaly</q>; <q>Brachydactyly/clinodactyly/camptodactyly</q>; <q>Intellectual disability/developmental delay</q>; <q>Zygomatic hypoplasia</q>; <q>Posteriorly rotated ear</q>; <q>Cupped ear </q>; <q>patient_id</q></td></tr>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(scg.to_html()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "23b8f381",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now get the unmapped columns and try option mappers\n",
    "unmapped_columns = scg.get_unmapped_columns()\n",
    "omit_columns = set(column_mapper_d.keys())\n",
    "omit_columns.update(['index','DNA','Protein','Age at last follow-up','Consanguinity'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "f12f8eb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "auto_results = OptionColumnMapper.autoformat(df=df, concept_recognizer=hpo_cr, omit_columns=omit_columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "d248129a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "index_d = {'1A': 'PLACEHOLDER',\n",
      " '1B': 'PLACEHOLDER',\n",
      " 'Hoxha': 'PLACEHOLDER',\n",
      " 'Choucair': 'PLACEHOLDER'}\n",
      "indexMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=index_d)\n",
      "indexMapper.preview_column(df['index'])\n",
      "column_mapper_d['index'] = indexMapper\n",
      "\n",
      "dna_d = {'c.[450A>T]; [893A>G]': 'PLACEHOLDER',\n",
      " 'c.[464C>T]; [893A>C]': 'PLACEHOLDER',\n",
      " 'c.[401A>G]; [895T>C]': 'PLACEHOLDER',\n",
      " 'c.[464C>T]; [62C>A]': 'PLACEHOLDER',\n",
      " 'c.[514C>T]; [514C>T]': 'PLACEHOLDER',\n",
      " 'c.[730C>T]; [730C>T]': 'PLACEHOLDER',\n",
      " 'c.[582+1G>A]; [582+1G>A]': 'PLACEHOLDER',\n",
      " 'c.[ 352A>T]; [352A>T]': 'PLACEHOLDER',\n",
      " 'g.[8783887_9068578del]; [8783887_9068578del]': 'PLACEHOLDER'}\n",
      "dnaMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=dna_d)\n",
      "dnaMapper.preview_column(df['DNA'])\n",
      "column_mapper_d['DNA'] = dnaMapper\n",
      "\n",
      "protein_d = {'p.[Glu150Asp]; [Asp298Gly]': 'PLACEHOLDER',\n",
      " 'p.[Pro155Leu]; [Asp298Ala]': 'PLACEHOLDER',\n",
      " 'p.[Asp134Gly]; [Ser299Pro]': 'PLACEHOLDER',\n",
      " 'p.[Pro155Leu]; [Ser21∗]': 'PLACEHOLDER',\n",
      " 'p.[Gln172∗]; [Gln172∗]': 'PLACEHOLDER',\n",
      " 'p.[Gln244∗]; [Gln244∗]': 'PLACEHOLDER',\n",
      " 'p.[?]; [?]': 'PLACEHOLDER',\n",
      " 'p.[Lys118∗]; [Lys118∗]': 'PLACEHOLDER',\n",
      " 'whole gene deletion': 'PLACEHOLDER'}\n",
      "proteinMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=protein_d)\n",
      "proteinMapper.preview_column(df['Protein'])\n",
      "column_mapper_d['Protein'] = proteinMapper\n",
      "\n",
      "age_at_last_follow-up_d = {'28 years': 'PLACEHOLDER',\n",
      " '26 years': 'PLACEHOLDER',\n",
      " '3.5\\xa0months (died)': 'PLACEHOLDER',\n",
      " '2 years (died)': 'PLACEHOLDER',\n",
      " '8 years 9\\xa0months': 'PLACEHOLDER',\n",
      " '13 years 3\\xa0months': 'PLACEHOLDER',\n",
      " '7 years': 'PLACEHOLDER',\n",
      " '5 years 6\\xa0months': 'PLACEHOLDER'}\n",
      "age_at_last_follow-upMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=age_at_last_follow-up_d)\n",
      "age_at_last_follow-upMapper.preview_column(df['Age at last follow-up'])\n",
      "column_mapper_d['Age at last follow-up'] = age_at_last_follow-upMapper\n",
      "\n",
      "weight_d = {'24\\xa0kg (−5 SD)': 'PLACEHOLDER',\n",
      " '26\\xa0kg (−5 SD)': 'PLACEHOLDER',\n",
      " 'nan': 'PLACEHOLDER',\n",
      " '3.3\\xa0kg (- 4 SD)': 'PLACEHOLDER',\n",
      " 'failure to thrive': 'Failure to thrive',\n",
      " '22\\xa0kg (8th centile)': 'PLACEHOLDER',\n",
      " '62\\xa0kg (85th centile)': 'PLACEHOLDER',\n",
      " '27.6\\xa0kg (50th centile)': 'PLACEHOLDER',\n",
      " 'normal': 'PLACEHOLDER'}\n",
      "weightMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=weight_d)\n",
      "weightMapper.preview_column(df['Weight'])\n",
      "column_mapper_d['Weight'] = weightMapper\n",
      "\n",
      "height_d = {'112\\xa0cm (−8 SD)': 'PLACEHOLDER',\n",
      " '128\\xa0cm (−7 SD)': 'PLACEHOLDER',\n",
      " 'nan': 'PLACEHOLDER',\n",
      " '50.3\\xa0cm (−5 SD)': 'PLACEHOLDER',\n",
      " 'short stature': 'Short stature',\n",
      " '130.8\\xa0cm (46th centile)': 'PLACEHOLDER',\n",
      " '155\\xa0cm (25th centile)': 'PLACEHOLDER',\n",
      " '130\\xa0cm (90th centile)': 'PLACEHOLDER',\n",
      " 'normal': 'PLACEHOLDER',\n",
      " '105\\xa0cm (<3rd centile)': 'PLACEHOLDER'}\n",
      "heightMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=height_d)\n",
      "heightMapper.preview_column(df['Height'])\n",
      "column_mapper_d['Height'] = heightMapper\n",
      "\n",
      "consanguinity_d = {}\n",
      "consanguinityMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=consanguinity_d)\n",
      "consanguinityMapper.preview_column(df['Consanguinity'])\n",
      "column_mapper_d['Consanguinity'] = consanguinityMapper\n",
      "\n",
      "fetal_ultrasound_d = {'hydronephrosis': 'Hydronephrosis',\n",
      " 'short limbs': 'Limb undergrowth',\n",
      " 'nan': 'PLACEHOLDER',\n",
      " 'severe IUGR': 'PLACEHOLDER',\n",
      " 'unremarkable': 'PLACEHOLDER'}\n",
      "fetal_ultrasoundMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=fetal_ultrasound_d)\n",
      "fetal_ultrasoundMapper.preview_column(df['Fetal ultrasound'])\n",
      "column_mapper_d['Fetal ultrasound'] = fetal_ultrasoundMapper\n",
      "\n",
      "gestation_age_d = {'term': 'PLACEHOLDER',\n",
      " 'prematurity': 'PLACEHOLDER'}\n",
      "gestation_ageMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=gestation_age_d)\n",
      "gestation_ageMapper.preview_column(df['Gestation age'])\n",
      "column_mapper_d['Gestation age'] = gestation_ageMapper\n",
      "\n",
      "birth_weight_d = {'nan': 'PLACEHOLDER',\n",
      " '2180\\xa0g (−3.2 SD)': 'PLACEHOLDER',\n",
      " '2': 'PLACEHOLDER',\n",
      " '000\\xa0g (−3.3 SD)': 'PLACEHOLDER',\n",
      " '760 g': 'PLACEHOLDER',\n",
      " '810\\xa0g (−1.6 SD)': 'PLACEHOLDER',\n",
      " '600\\xa0g (−2 SD)': 'PLACEHOLDER',\n",
      " '4': 'PLACEHOLDER',\n",
      " '010\\xa0g (91st centile)': 'PLACEHOLDER',\n",
      " '3': 'PLACEHOLDER',\n",
      " '000\\xa0g (50th centile)': 'PLACEHOLDER'}\n",
      "birth_weightMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=birth_weight_d)\n",
      "birth_weightMapper.preview_column(df['Birth weight'])\n",
      "column_mapper_d['Birth weight'] = birth_weightMapper\n",
      "\n",
      "birth_length_d = {'nan': 'PLACEHOLDER',\n",
      " '40\\xa0cm (−5 SD)': 'PLACEHOLDER',\n",
      " '41.7\\xa0cm (−4.3 SD)': 'PLACEHOLDER',\n",
      " '50\\xa0cm at 6\\xa0weeks (−2.1 SD)': 'PLACEHOLDER',\n",
      " '49\\xa0cm (−1.3 SD)': 'PLACEHOLDER',\n",
      " 'normal': 'PLACEHOLDER',\n",
      " '49.5\\xa0cm (43rd centile)': 'PLACEHOLDER',\n",
      " '50\\xa0cm (−1.3 SD)': 'PLACEHOLDER'}\n",
      "birth_lengthMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=birth_length_d)\n",
      "birth_lengthMapper.preview_column(df['Birth length'])\n",
      "column_mapper_d['Birth length'] = birth_lengthMapper\n",
      "\n",
      "spine_anomaly_d = {'nan': 'PLACEHOLDER'}\n",
      "spine_anomalyMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=spine_anomaly_d)\n",
      "spine_anomalyMapper.preview_column(df['Spine anomaly'])\n",
      "column_mapper_d['Spine anomaly'] = spine_anomalyMapper\n",
      "\n",
      "metaphyseal_anomaly_d = {'nan': 'PLACEHOLDER'}\n",
      "metaphyseal_anomalyMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=metaphyseal_anomaly_d)\n",
      "metaphyseal_anomalyMapper.preview_column(df['Metaphyseal anomaly'])\n",
      "column_mapper_d['Metaphyseal anomaly'] = metaphyseal_anomalyMapper\n",
      "\n",
      "epiphyseal_anomaly_d = {'+ (wrists)': 'PLACEHOLDER'}\n",
      "epiphyseal_anomalyMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=epiphyseal_anomaly_d)\n",
      "epiphyseal_anomalyMapper.preview_column(df['Epiphyseal anomaly'])\n",
      "column_mapper_d['Epiphyseal anomaly'] = epiphyseal_anomalyMapper\n",
      "\n",
      "syndactyly_d = {}\n",
      "syndactylyMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=syndactyly_d)\n",
      "syndactylyMapper.preview_column(df['Syndactyly'])\n",
      "column_mapper_d['Syndactyly'] = syndactylyMapper\n",
      "\n",
      "brachydactyly/clinodactyly/camptodactyly_d = {}\n",
      "brachydactyly/clinodactyly/camptodactylyMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=brachydactyly/clinodactyly/camptodactyly_d)\n",
      "brachydactyly/clinodactyly/camptodactylyMapper.preview_column(df['Brachydactyly/clinodactyly/camptodactyly'])\n",
      "column_mapper_d['Brachydactyly/clinodactyly/camptodactyly'] = brachydactyly/clinodactyly/camptodactylyMapper\n",
      "\n",
      "cardiac_anomaly_d = {}\n",
      "cardiac_anomalyMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=cardiac_anomaly_d)\n",
      "cardiac_anomalyMapper.preview_column(df['Cardiac anomaly'])\n",
      "column_mapper_d['Cardiac anomaly'] = cardiac_anomalyMapper\n",
      "\n",
      "hydronephrosis_d = {'nan': 'PLACEHOLDER'}\n",
      "hydronephrosisMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=hydronephrosis_d)\n",
      "hydronephrosisMapper.preview_column(df['Hydronephrosis'])\n",
      "column_mapper_d['Hydronephrosis'] = hydronephrosisMapper\n",
      "\n",
      "intellectual_disability/developmental_delay_d = {'nan': 'PLACEHOLDER',\n",
      " 'Motor delay': 'Motor delay',\n",
      " 'Delayed speech and language development': 'Delayed speech and language development',\n",
      " 'generalized hypotonia': 'Generalized hypotonia',\n",
      " 'Global developmental delay': 'Global developmental delay',\n",
      " 'Autism': 'Autism',\n",
      " 'Intellectual disability mild': 'Intellectual disability',\n",
      " 'Intellectual disability': 'Intellectual disability'}\n",
      "intellectual_disability/developmental_delayMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=intellectual_disability/developmental_delay_d)\n",
      "intellectual_disability/developmental_delayMapper.preview_column(df['Intellectual disability/developmental delay'])\n",
      "column_mapper_d['Intellectual disability/developmental delay'] = intellectual_disability/developmental_delayMapper\n",
      "\n",
      "vesicoureteral_reflux_d = {'nan': 'PLACEHOLDER'}\n",
      "vesicoureteral_refluxMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=vesicoureteral_reflux_d)\n",
      "vesicoureteral_refluxMapper.preview_column(df['Vesicoureteral reflux'])\n",
      "column_mapper_d['Vesicoureteral reflux'] = vesicoureteral_refluxMapper\n",
      "\n",
      "hydronephrosis_d = {'nan': 'PLACEHOLDER'}\n",
      "hydronephrosisMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=hydronephrosis_d)\n",
      "hydronephrosisMapper.preview_column(df['Hydronephrosis'])\n",
      "column_mapper_d['Hydronephrosis'] = hydronephrosisMapper\n",
      "\n",
      "asthma_d = {'nan': 'PLACEHOLDER'}\n",
      "asthmaMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=asthma_d)\n",
      "asthmaMapper.preview_column(df['Asthma'])\n",
      "column_mapper_d['Asthma'] = asthmaMapper\n",
      "\n",
      "conductive_hearing_impairment_d = {'nan': 'PLACEHOLDER'}\n",
      "conductive_hearing_impairmentMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=conductive_hearing_impairment_d)\n",
      "conductive_hearing_impairmentMapper.preview_column(df['Conductive hearing impairment'])\n",
      "column_mapper_d['Conductive hearing impairment'] = conductive_hearing_impairmentMapper\n",
      "\n",
      "hypernasal_speech_d = {'nan': 'PLACEHOLDER'}\n",
      "hypernasal_speechMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=hypernasal_speech_d)\n",
      "hypernasal_speechMapper.preview_column(df['hypernasal speech'])\n",
      "column_mapper_d['hypernasal speech'] = hypernasal_speechMapper\n",
      "\n",
      "dislocated_radial_head_d = {'nan': 'PLACEHOLDER'}\n",
      "dislocated_radial_headMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=dislocated_radial_head_d)\n",
      "dislocated_radial_headMapper.preview_column(df['Dislocated radial head'])\n",
      "column_mapper_d['Dislocated radial head'] = dislocated_radial_headMapper\n",
      "\n",
      "scoliosis_d = {'nan': 'PLACEHOLDER'}\n",
      "scoliosisMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=scoliosis_d)\n",
      "scoliosisMapper.preview_column(df['Scoliosis'])\n",
      "column_mapper_d['Scoliosis'] = scoliosisMapper\n",
      "\n",
      "hip_pain_d = {'nan': 'PLACEHOLDER'}\n",
      "hip_painMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=hip_pain_d)\n",
      "hip_painMapper.preview_column(df['Hip pain'])\n",
      "column_mapper_d['Hip pain'] = hip_painMapper\n",
      "\n",
      "short_stature_d = {'nan': 'PLACEHOLDER'}\n",
      "short_statureMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=short_stature_d)\n",
      "short_statureMapper.preview_column(df['Short stature'])\n",
      "column_mapper_d['Short stature'] = short_statureMapper\n",
      "\n",
      "long_face_d = {'nan': 'PLACEHOLDER'}\n",
      "long_faceMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=long_face_d)\n",
      "long_faceMapper.preview_column(df['Long face'])\n",
      "column_mapper_d['Long face'] = long_faceMapper\n",
      "\n",
      "narrow_face_d = {'nan': 'PLACEHOLDER'}\n",
      "narrow_faceMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=narrow_face_d)\n",
      "narrow_faceMapper.preview_column(df['Narrow face'])\n",
      "column_mapper_d['Narrow face'] = narrow_faceMapper\n",
      "\n",
      "proptosis_d = {'nan': 'PLACEHOLDER'}\n",
      "proptosisMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=proptosis_d)\n",
      "proptosisMapper.preview_column(df['proptosis'])\n",
      "column_mapper_d['proptosis'] = proptosisMapper\n",
      "\n",
      "zygomatic_hypoplasia_d = {'nan': 'PLACEHOLDER'}\n",
      "zygomatic_hypoplasiaMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=zygomatic_hypoplasia_d)\n",
      "zygomatic_hypoplasiaMapper.preview_column(df['Zygomatic hypoplasia'])\n",
      "column_mapper_d['Zygomatic hypoplasia'] = zygomatic_hypoplasiaMapper\n",
      "\n",
      "coarse_facies_d = {'nan': 'PLACEHOLDER'}\n",
      "coarse_faciesMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=coarse_facies_d)\n",
      "coarse_faciesMapper.preview_column(df['Coarse facies'])\n",
      "column_mapper_d['Coarse facies'] = coarse_faciesMapper\n",
      "\n",
      "low-set_ears_d = {'nan': 'PLACEHOLDER'}\n",
      "low-set_earsMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=low-set_ears_d)\n",
      "low-set_earsMapper.preview_column(df['Low-set ears'])\n",
      "column_mapper_d['Low-set ears'] = low-set_earsMapper\n",
      "\n",
      "limited_elbow_extension_d = {'nan': 'PLACEHOLDER'}\n",
      "limited_elbow_extensionMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=limited_elbow_extension_d)\n",
      "limited_elbow_extensionMapper.preview_column(df['Limited elbow extension'])\n",
      "column_mapper_d['Limited elbow extension'] = limited_elbow_extensionMapper\n",
      "\n",
      "finger_joint_hypermobility_d = {'nan': 'PLACEHOLDER'}\n",
      "finger_joint_hypermobilityMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=finger_joint_hypermobility_d)\n",
      "finger_joint_hypermobilityMapper.preview_column(df['Finger joint hypermobility'])\n",
      "column_mapper_d['Finger joint hypermobility'] = finger_joint_hypermobilityMapper\n",
      "\n",
      "clinodactyly_of_the_5th_finger_d = {'nan': 'PLACEHOLDER'}\n",
      "clinodactyly_of_the_5th_fingerMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=clinodactyly_of_the_5th_finger_d)\n",
      "clinodactyly_of_the_5th_fingerMapper.preview_column(df['Clinodactyly of the 5th finger'])\n",
      "column_mapper_d['Clinodactyly of the 5th finger'] = clinodactyly_of_the_5th_fingerMapper\n",
      "\n",
      "pes_planus_d = {'nan': 'PLACEHOLDER'}\n",
      "pes_planusMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=pes_planus_d)\n",
      "pes_planusMapper.preview_column(df['Pes planus'])\n",
      "column_mapper_d['Pes planus'] = pes_planusMapper\n",
      "\n",
      "slender_metacarpals_d = {'nan': 'PLACEHOLDER'}\n",
      "slender_metacarpalsMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=slender_metacarpals_d)\n",
      "slender_metacarpalsMapper.preview_column(df['Slender metacarpals'])\n",
      "column_mapper_d['Slender metacarpals'] = slender_metacarpalsMapper\n",
      "\n",
      "increased_vertebral_height_d = {'nan': 'PLACEHOLDER'}\n",
      "increased_vertebral_heightMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=increased_vertebral_height_d)\n",
      "increased_vertebral_heightMapper.preview_column(df['Increased vertebral height'])\n",
      "column_mapper_d['Increased vertebral height'] = increased_vertebral_heightMapper\n",
      "\n",
      "velopharyngeal_insufficiency_d = {'nan': 'PLACEHOLDER'}\n",
      "velopharyngeal_insufficiencyMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=velopharyngeal_insufficiency_d)\n",
      "velopharyngeal_insufficiencyMapper.preview_column(df['Velopharyngeal insufficiency'])\n",
      "column_mapper_d['Velopharyngeal insufficiency'] = velopharyngeal_insufficiencyMapper\n",
      "\n",
      "hip_dislocation_d = {'nan': 'PLACEHOLDER'}\n",
      "hip_dislocationMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=hip_dislocation_d)\n",
      "hip_dislocationMapper.preview_column(df['Hip dislocation'])\n",
      "column_mapper_d['Hip dislocation'] = hip_dislocationMapper\n",
      "\n",
      "patellar_dislocation_d = {'nan': 'PLACEHOLDER'}\n",
      "patellar_dislocationMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=patellar_dislocation_d)\n",
      "patellar_dislocationMapper.preview_column(df['Patellar dislocation'])\n",
      "column_mapper_d['Patellar dislocation'] = patellar_dislocationMapper\n",
      "\n",
      "narrow_forehead_d = {'nan': 'PLACEHOLDER'}\n",
      "narrow_foreheadMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=narrow_forehead_d)\n",
      "narrow_foreheadMapper.preview_column(df['Narrow forehead'])\n",
      "column_mapper_d['Narrow forehead'] = narrow_foreheadMapper\n",
      "\n",
      "upslanted_palpebral_fissure_d = {'nan': 'PLACEHOLDER'}\n",
      "upslanted_palpebral_fissureMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=upslanted_palpebral_fissure_d)\n",
      "upslanted_palpebral_fissureMapper.preview_column(df['Upslanted palpebral fissure'])\n",
      "column_mapper_d['Upslanted palpebral fissure'] = upslanted_palpebral_fissureMapper\n",
      "\n",
      "high_palate_d = {'nan': 'PLACEHOLDER'}\n",
      "high_palateMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=high_palate_d)\n",
      "high_palateMapper.preview_column(df['High palate'])\n",
      "column_mapper_d['High palate'] = high_palateMapper\n",
      "\n",
      "pectus_excavatum_d = {'nan': 'PLACEHOLDER'}\n",
      "pectus_excavatumMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=pectus_excavatum_d)\n",
      "pectus_excavatumMapper.preview_column(df['Pectus excavatum'])\n",
      "column_mapper_d['Pectus excavatum'] = pectus_excavatumMapper\n",
      "\n",
      "tapered_finger_d = {'nan': 'PLACEHOLDER'}\n",
      "tapered_fingerMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=tapered_finger_d)\n",
      "tapered_fingerMapper.preview_column(df['Tapered finger'])\n",
      "column_mapper_d['Tapered finger'] = tapered_fingerMapper\n",
      "\n",
      "prominent_forehead_d = {'nan': 'PLACEHOLDER'}\n",
      "prominent_foreheadMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=prominent_forehead_d)\n",
      "prominent_foreheadMapper.preview_column(df['Prominent forehead'])\n",
      "column_mapper_d['Prominent forehead'] = prominent_foreheadMapper\n",
      "\n",
      "depressed_nasal_bridge_d = {'nan': 'PLACEHOLDER'}\n",
      "depressed_nasal_bridgeMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=depressed_nasal_bridge_d)\n",
      "depressed_nasal_bridgeMapper.preview_column(df['Depressed nasal bridge'])\n",
      "column_mapper_d['Depressed nasal bridge'] = depressed_nasal_bridgeMapper\n",
      "\n",
      "micrognathia_d = {'nan': 'PLACEHOLDER'}\n",
      "micrognathiaMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=micrognathia_d)\n",
      "micrognathiaMapper.preview_column(df['Micrognathia'])\n",
      "column_mapper_d['Micrognathia'] = micrognathiaMapper\n",
      "\n",
      "cutaneous_syndactyly_d = {'nan': 'PLACEHOLDER'}\n",
      "cutaneous_syndactylyMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=cutaneous_syndactyly_d)\n",
      "cutaneous_syndactylyMapper.preview_column(df['Cutaneous syndactyly'])\n",
      "column_mapper_d['Cutaneous syndactyly'] = cutaneous_syndactylyMapper\n",
      "\n",
      "macrotia_d = {'nan': 'PLACEHOLDER'}\n",
      "macrotiaMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=macrotia_d)\n",
      "macrotiaMapper.preview_column(df['Macrotia'])\n",
      "column_mapper_d['Macrotia'] = macrotiaMapper\n",
      "\n",
      "narrow_chest_d = {'nan': 'PLACEHOLDER'}\n",
      "narrow_chestMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=narrow_chest_d)\n",
      "narrow_chestMapper.preview_column(df['Narrow chest'])\n",
      "column_mapper_d['Narrow chest'] = narrow_chestMapper\n",
      "\n",
      "pulmonary_arterial_hypertension_d = {'nan': 'PLACEHOLDER'}\n",
      "pulmonary_arterial_hypertensionMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=pulmonary_arterial_hypertension_d)\n",
      "pulmonary_arterial_hypertensionMapper.preview_column(df['Pulmonary arterial hypertension'])\n",
      "column_mapper_d['Pulmonary arterial hypertension'] = pulmonary_arterial_hypertensionMapper\n",
      "\n",
      "oligodactyly_d = {'nan': 'PLACEHOLDER'}\n",
      "oligodactylyMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=oligodactyly_d)\n",
      "oligodactylyMapper.preview_column(df['Oligodactyly'])\n",
      "column_mapper_d['Oligodactyly'] = oligodactylyMapper\n",
      "\n",
      "tricuspid_regurgitation_d = {'nan': 'PLACEHOLDER'}\n",
      "tricuspid_regurgitationMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=tricuspid_regurgitation_d)\n",
      "tricuspid_regurgitationMapper.preview_column(df['Tricuspid regurgitation'])\n",
      "column_mapper_d['Tricuspid regurgitation'] = tricuspid_regurgitationMapper\n",
      "\n",
      "platyspondyly_d = {'nan': 'PLACEHOLDER'}\n",
      "platyspondylyMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=platyspondyly_d)\n",
      "platyspondylyMapper.preview_column(df['Platyspondyly'])\n",
      "column_mapper_d['Platyspondyly'] = platyspondylyMapper\n",
      "\n",
      "intrauterine_growth_retardation_d = {'nan': 'PLACEHOLDER'}\n",
      "intrauterine_growth_retardationMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=intrauterine_growth_retardation_d)\n",
      "intrauterine_growth_retardationMapper.preview_column(df['Intrauterine growth retardation'])\n",
      "column_mapper_d['Intrauterine growth retardation'] = intrauterine_growth_retardationMapper\n",
      "\n",
      "motor_delay_d = {'nan': 'PLACEHOLDER'}\n",
      "motor_delayMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=motor_delay_d)\n",
      "motor_delayMapper.preview_column(df['Motor delay'])\n",
      "column_mapper_d['Motor delay'] = motor_delayMapper\n",
      "\n",
      "failure_to_thrive_d = {'nan': 'PLACEHOLDER'}\n",
      "failure_to_thriveMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=failure_to_thrive_d)\n",
      "failure_to_thriveMapper.preview_column(df['Failure to thrive'])\n",
      "column_mapper_d['Failure to thrive'] = failure_to_thriveMapper\n",
      "\n",
      "trigonocephaly_d = {'nan': 'PLACEHOLDER'}\n",
      "trigonocephalyMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=trigonocephaly_d)\n",
      "trigonocephalyMapper.preview_column(df['Trigonocephaly'])\n",
      "column_mapper_d['Trigonocephaly'] = trigonocephalyMapper\n",
      "\n",
      "frontal_bossing_d = {'nan': 'PLACEHOLDER'}\n",
      "frontal_bossingMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=frontal_bossing_d)\n",
      "frontal_bossingMapper.preview_column(df['Frontal bossing'])\n",
      "column_mapper_d['Frontal bossing'] = frontal_bossingMapper\n",
      "\n",
      "sparse_hair_d = {'nan': 'PLACEHOLDER'}\n",
      "sparse_hairMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=sparse_hair_d)\n",
      "sparse_hairMapper.preview_column(df['Sparse hair'])\n",
      "column_mapper_d['Sparse hair'] = sparse_hairMapper\n",
      "\n",
      "micrognathia_d = {'nan': 'PLACEHOLDER'}\n",
      "micrognathiaMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=micrognathia_d)\n",
      "micrognathiaMapper.preview_column(df['Micrognathia'])\n",
      "column_mapper_d['Micrognathia'] = micrognathiaMapper\n",
      "\n",
      "posteriorly_rotated_ear_d = {'nan': 'PLACEHOLDER'}\n",
      "posteriorly_rotated_earMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=posteriorly_rotated_ear_d)\n",
      "posteriorly_rotated_earMapper.preview_column(df['Posteriorly rotated ear'])\n",
      "column_mapper_d['Posteriorly rotated ear'] = posteriorly_rotated_earMapper\n",
      "\n",
      "pectus_carinatum_d = {'nan': 'PLACEHOLDER'}\n",
      "pectus_carinatumMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=pectus_carinatum_d)\n",
      "pectus_carinatumMapper.preview_column(df['Pectus carinatum'])\n",
      "column_mapper_d['Pectus carinatum'] = pectus_carinatumMapper\n",
      "\n",
      "wormian_bones_d = {'nan': 'PLACEHOLDER'}\n",
      "wormian_bonesMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=wormian_bones_d)\n",
      "wormian_bonesMapper.preview_column(df['Wormian bones'])\n",
      "column_mapper_d['Wormian bones'] = wormian_bonesMapper\n",
      "\n",
      "osteopenia_d = {'nan': 'PLACEHOLDER'}\n",
      "osteopeniaMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=osteopenia_d)\n",
      "osteopeniaMapper.preview_column(df['Osteopenia'])\n",
      "column_mapper_d['Osteopenia'] = osteopeniaMapper\n",
      "\n",
      "delayed_skeletal_maturation_d = {'nan': 'PLACEHOLDER'}\n",
      "delayed_skeletal_maturationMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=delayed_skeletal_maturation_d)\n",
      "delayed_skeletal_maturationMapper.preview_column(df['Delayed skeletal maturation'])\n",
      "column_mapper_d['Delayed skeletal maturation'] = delayed_skeletal_maturationMapper\n",
      "\n",
      "inguinal_hernia_d = {'nan': 'PLACEHOLDER'}\n",
      "inguinal_herniaMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=inguinal_hernia_d)\n",
      "inguinal_herniaMapper.preview_column(df['Inguinal hernia'])\n",
      "column_mapper_d['Inguinal hernia'] = inguinal_herniaMapper\n",
      "\n",
      "ventricular_septal_defect_d = {'nan': 'PLACEHOLDER'}\n",
      "ventricular_septal_defectMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=ventricular_septal_defect_d)\n",
      "ventricular_septal_defectMapper.preview_column(df['Ventricular septal defect'])\n",
      "column_mapper_d['Ventricular septal defect'] = ventricular_septal_defectMapper\n",
      "\n",
      "brachycephaly_d = {'nan': 'PLACEHOLDER'}\n",
      "brachycephalyMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=brachycephaly_d)\n",
      "brachycephalyMapper.preview_column(df['Brachycephaly'])\n",
      "column_mapper_d['Brachycephaly'] = brachycephalyMapper\n",
      "\n",
      "cupped_ear__d = {'nan': 'PLACEHOLDER'}\n",
      "cupped_ear_Mapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=cupped_ear__d)\n",
      "cupped_ear_Mapper.preview_column(df['Cupped ear '])\n",
      "column_mapper_d['Cupped ear '] = cupped_ear_Mapper\n",
      "\n",
      "anonychia_d = {'nan': 'PLACEHOLDER'}\n",
      "anonychiaMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=anonychia_d)\n",
      "anonychiaMapper.preview_column(df['Anonychia'])\n",
      "column_mapper_d['Anonychia'] = anonychiaMapper\n",
      "\n",
      "strabismus_d = {'nan': 'PLACEHOLDER'}\n",
      "strabismusMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=strabismus_d)\n",
      "strabismusMapper.preview_column(df['Strabismus'])\n",
      "column_mapper_d['Strabismus'] = strabismusMapper\n",
      "\n",
      "low_anterior_hairline_d = {'nan': 'PLACEHOLDER'}\n",
      "low_anterior_hairlineMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=low_anterior_hairline_d)\n",
      "low_anterior_hairlineMapper.preview_column(df['Low anterior hairline'])\n",
      "column_mapper_d['Low anterior hairline'] = low_anterior_hairlineMapper\n",
      "\n",
      "epicanthus_d = {'nan': 'PLACEHOLDER'}\n",
      "epicanthusMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=epicanthus_d)\n",
      "epicanthusMapper.preview_column(df['Epicanthus'])\n",
      "column_mapper_d['Epicanthus'] = epicanthusMapper\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(auto_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "9ede0bd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "weight_d = {'24\\xa0kg (−5 SD)': 'Decreased body weight',\n",
    " '26\\xa0kg (−5 SD)': 'Decreased body weight',\n",
    " '3.3\\xa0kg (- 4 SD)': 'Decreased body weight',\n",
    " 'failure to thrive': 'Failure to thrive'}\n",
    "excluded_d = {\n",
    "    '22\\xa0kg (8th centile)': 'Decreased body weight',\n",
    "    '62\\xa0kg (85th centile)': 'Decreased body weight',\n",
    "    '27.6\\xa0kg (50th centile)': 'Decreased body weight',\n",
    "    'normal': 'Failure to thrive',\n",
    "}\n",
    "weightMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=weight_d,\n",
    "                                 excluded_d=excluded_d)\n",
    "weightMapper.preview_column(df['Weight'])\n",
    "column_mapper_d['Weight'] = weightMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "a54cf9ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "height_d = {'112\\xa0cm (−8 SD)': 'Short stature',\n",
    " '128\\xa0cm (−7 SD)': 'Short stature',\n",
    " '50.3\\xa0cm (−5 SD)': 'Short stature',\n",
    " 'short stature': 'Short stature',\n",
    " \n",
    " '105\\xa0cm (<3rd centile)': 'Short stature'}\n",
    "\n",
    "excluded_d = {\n",
    "    '130.8\\xa0cm (46th centile)': 'Short stature',\n",
    " '155\\xa0cm (25th centile)': 'Short stature',\n",
    " '130\\xa0cm (90th centile)': 'Short stature',\n",
    "}\n",
    "heightMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=height_d,\n",
    "                                excluded_d=excluded_d)\n",
    "heightMapper.preview_column(df['Height'])\n",
    "column_mapper_d['Height'] = heightMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "5b0aae33",
   "metadata": {},
   "outputs": [],
   "source": [
    "fetal_ultrasound_d = {'hydronephrosis': 'Hydronephrosis',\n",
    " 'short limbs': 'Limb undergrowth',\n",
    " 'severe IUGR': 'Intrauterine growth retardation',\n",
    " }\n",
    "excluded = {\n",
    "    'unremarkable': 'Intrauterine growth retardation'\n",
    "}\n",
    "fetal_ultrasoundMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=fetal_ultrasound_d)\n",
    "fetal_ultrasoundMapper.preview_column(df['Fetal ultrasound'])\n",
    "column_mapper_d['Fetal ultrasound'] = fetal_ultrasoundMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "7c592383",
   "metadata": {},
   "outputs": [],
   "source": [
    "birth_weight_d = {\n",
    " '2180\\xa0g (−3.2 SD)': 'Small for gestational age',\n",
    " '000\\xa0g (−3.3 SD)': 'Small for gestational age',}\n",
    "birth_weightMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=birth_weight_d)\n",
    "birth_weightMapper.preview_column(df['Birth weight'])\n",
    "column_mapper_d['Birth weight'] = birth_weightMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "20695fa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Omitting these because we manually curated detailed phenotypes and added them to the input table\n",
    "#spine_anomaly_d \n",
    "#metaphyseal_anomaly_d = {'nan': 'PLACEHOLDER'}\n",
    "#epiphyseal_anomaly_d = {'+ (wrists)': 'PLACEHOLDER'}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "3025ad6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_gdd_d = {\n",
    " 'Motor delay': 'Motor delay',\n",
    " 'Delayed speech and language development': 'Delayed speech and language development',\n",
    " 'generalized hypotonia': 'Generalized hypotonia',\n",
    " 'Global developmental delay': 'Global developmental delay',\n",
    " 'Autism': 'Autism',\n",
    " 'Intellectual disability mild': 'Intellectual disability, mild',}\n",
    "id_gddMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=id_gdd_d)\n",
    "id_gddMapper.preview_column(df['Intellectual disability/developmental delay'])\n",
    "column_mapper_d['Intellectual disability/developmental delay'] = id_gddMapper"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91805b6e",
   "metadata": {},
   "source": [
    "<H2>Variants</H2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "889ad993",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\"c.893A>C\"\n",
      "\"c.450A>T\"\n",
      "\"c.464C>T\"\n",
      "\"g.8783887_9068578del\"\n",
      "\"c.401A>G\"\n",
      "\"c.514C>T\"\n",
      "\"c.352A>T\"\n",
      "\"c.893A>G\"\n",
      "\"c.730C>T\"\n",
      "\"c.62C>A\"\n",
      "\"c.582+1G>A\"\n",
      "\"c.895T>C\"\n"
     ]
    }
   ],
   "source": [
    "variant_set = set()\n",
    "for v_string in df[\"DNA\"]:\n",
    "    fields = v_string.split(\";\")\n",
    "    if len(fields) != 2:\n",
    "        raise ValueError(f\"Malformed variant line {v_string}\")\n",
    "    for var in fields:\n",
    "        variant_str = var.strip()\n",
    "        variant_str = re.sub(r\"(c.)?\\[\",\"\", variant_str)\n",
    "        variant_str = variant_str.replace(\"]\", \"\").strip()\n",
    "        if \"8783887\" in variant_str:  # this is the structural variant\n",
    "            variant_set.add(\"g.8783887_9068578del\")\n",
    "        else:\n",
    "            variant_str = \"c.\" + variant_str\n",
    "            variant_set.add(variant_str)\n",
    "for v in variant_set:\n",
    "    print(f\"\\\"{v}\\\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "b955a529",
   "metadata": {},
   "outputs": [],
   "source": [
    "validator = VariantValidator(genome_build='hg38', transcript=\"NM_153332.4\")\n",
    "validated_var_d = defaultdict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "12eea9ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating c.893A>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_153332.4%3Ac.893A>C/NM_153332.4?content-type=application%2Fjson\n",
      "Validating c.450A>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_153332.4%3Ac.450A>T/NM_153332.4?content-type=application%2Fjson\n",
      "Validating c.464C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_153332.4%3Ac.464C>T/NM_153332.4?content-type=application%2Fjson\n",
      "Validating g.8783887_9068578del\n",
      "Validating c.401A>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_153332.4%3Ac.401A>G/NM_153332.4?content-type=application%2Fjson\n",
      "Validating c.514C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_153332.4%3Ac.514C>T/NM_153332.4?content-type=application%2Fjson\n",
      "Validating c.352A>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_153332.4%3Ac.352A>T/NM_153332.4?content-type=application%2Fjson\n",
      "Validating c.893A>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_153332.4%3Ac.893A>G/NM_153332.4?content-type=application%2Fjson\n",
      "Validating c.730C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_153332.4%3Ac.730C>T/NM_153332.4?content-type=application%2Fjson\n",
      "Validating c.62C>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_153332.4%3Ac.62C>A/NM_153332.4?content-type=application%2Fjson\n",
      "Validating c.582+1G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_153332.4%3Ac.582+1G>A/NM_153332.4?content-type=application%2Fjson\n",
      "Validating c.895T>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_153332.4%3Ac.895T>C/NM_153332.4?content-type=application%2Fjson\n",
      "We got 12 variant objects\n"
     ]
    }
   ],
   "source": [
    "for var in variant_set:\n",
    "    print(f\"Validating {var}\")\n",
    "    if var == 'g.8783887_9068578del':\n",
    "        sv = StructuralVariant.chromosomal_deletion(cell_contents='Deletion exons 1-4',\n",
    "                 gene_symbol=\"ERI1\",\n",
    "                 gene_id=\"HGNC:23994\",\n",
    "                 genotype=\"homozygous\")\n",
    "        validated_var_d[var] = sv\n",
    "    else:\n",
    "        var_object = validator.encode_hgvs(hgvs=var)\n",
    "        validated_var_d[var] = var_object\n",
    "print(f\"We got {len(validated_var_d)} variant objects\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "034417a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "ageMapper = AgeColumnMapper.by_year('Age at last follow-up')\n",
    "#ageMapper.preview_column(df['Age at last follow-up'])\n",
    "sexMapper = SexColumnMapper(male_symbol='M', female_symbol='F', column_name='Sex')\n",
    "#sexMapper.preview_column(df['Sex'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc6c2845",
   "metadata": {},
   "source": [
    "<h2>Disease diagnosis</h2>\n",
    "<p>Diseases related to ERI1 are currenttly not represented in OMIM. For this reason, we represent the diagnosis as preliminary below. The authors write:  SEMD was present in the five individuals with at least one missense variant (Table 1). In contrast, three individuals with ERI1 null mutations and the Eri1 KO mice showed a much milder skeletal phenotype without any evidence for SEMD, consistent with the two individuals reported previously, who had homozygous a 284 kb deletion and p.Lys118∗. Notably, of the five individuals with SEMD, three died within 2 years after birth, suggesting missense variants lead to a poor prognosis.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "8105d0f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pmid=\"PMID:37352860\"\n",
    "encoder = CohortEncoder(df=df, hpo_cr=hpo_cr, column_mapper_d=column_mapper_d, \n",
    "                        individual_column_name=\"patient_id\", agemapper=ageMapper, sexmapper=sexMapper,\n",
    "                       metadata=metadata,\n",
    "                       pmid=pmid)\n",
    "encoder.set_disease(disease_id='Preliminary:Preliminary', label='ERI1-related disease')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "528b4519",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Error: cell_contents argument (Individual\nHydronephrosis    +\nHydronephrosis    +\nName: 0, dtype: object) must be string but was <class 'pandas.core.series.Series'> -- coerced to string",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[148], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m individuals \u001b[38;5;241m=\u001b[39m \u001b[43mencoder\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_individuals\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/psvenv/lib/python3.10/site-packages/pyphetools/creation/cohort_encoder.py:167\u001b[0m, in \u001b[0;36mCohortEncoder.get_individuals\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    165\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(cell_contents, \u001b[38;5;28mfloat\u001b[39m) \u001b[38;5;129;01mand\u001b[39;00m isnan(cell_contents):\n\u001b[1;32m    166\u001b[0m         \u001b[38;5;28;01mcontinue\u001b[39;00m\n\u001b[0;32m--> 167\u001b[0m     terms \u001b[38;5;241m=\u001b[39m \u001b[43mcolumn_mapper\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmap_cell\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcell_contents\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    168\u001b[0m     hpo_terms\u001b[38;5;241m.\u001b[39mextend(terms)\n\u001b[1;32m    169\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m variant_colname \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/psvenv/lib/python3.10/site-packages/pyphetools/creation/simple_column_mapper.py:60\u001b[0m, in \u001b[0;36mSimpleColumnMapper.map_cell\u001b[0;34m(self, cell_contents)\u001b[0m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mmap_cell\u001b[39m(\u001b[38;5;28mself\u001b[39m, cell_contents) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m List[HpTerm]:\n\u001b[1;32m     59\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(cell_contents, \u001b[38;5;28mstr\u001b[39m):\n\u001b[0;32m---> 60\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m     61\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mError: cell_contents argument (\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcell_contents\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m) must be string but was \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(cell_contents)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m -- coerced to string\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     62\u001b[0m     contents \u001b[38;5;241m=\u001b[39m cell_contents\u001b[38;5;241m.\u001b[39mstrip()\n\u001b[1;32m     63\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m contents \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_observed:\n",
      "\u001b[0;31mValueError\u001b[0m: Error: cell_contents argument (Individual\nHydronephrosis    +\nHydronephrosis    +\nName: 0, dtype: object) must be string but was <class 'pandas.core.series.Series'> -- coerced to string"
     ]
    }
   ],
   "source": [
    "individuals = encoder.get_individuals()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43084079",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "psvenv",
   "language": "python",
   "name": "psvenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
