{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a39f3c65",
   "metadata": {},
   "source": [
    "<h1>Creation of phenopackets from tabular data (individuals in columns)</h1>\n",
    "<p>We will process <a href=\"https://pubmed.ncbi.nlm.nih.gov/28258187/\" target=\"__blank\">Vandervore, et al. (2017) Bi-allelic variants in COL3A1 encoding the ligand to GPR56 are associated with cobblestone-like cortical malformation, white matter changes and cerebellar cysts</a></p>\n",
    "<p>pyphetools provides a convenient way of extracting HPO terms from typical tables presented in supplemental material. Typical tables can have the individuals in columns or rows. In this case, we extract data from TABLE.</p>\n",
    "<p>Please see the <a href=\"https://monarch-initiative.github.io/pyphetools/\" target=\"__blank\">phephetools documentation</a> on the software.</p>\n",
    "<p>Users can work on one column at a time and then generate a collection of <a href=\"https://pubmed.ncbi.nlm.nih.gov/35705716/\" target=\"__blank\">GA4GH phenopackets</a> to represent each patient included in the original supplemental material. These phenopackets can then be used for a variety of downstream applications.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f0602c0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import phenopackets as php\n",
    "from google.protobuf.json_format import MessageToDict, MessageToJson\n",
    "from google.protobuf.json_format import Parse, ParseDict\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_colwidth', None) # show entire column contents, important!\n",
    "from collections import defaultdict\n",
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "from pyphetools.creation import *\n",
    "from pyphetools.creation.simple_column_mapper import try_mapping_columns, get_separate_hpos_from_df\n",
    "# last tested with pyphetools version 0.4.5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8606e7eb",
   "metadata": {},
   "source": [
    "<h2>Importing HPO data</h2>\n",
    "<p>pyphetools uses the Human Phenotype Ontology (HPO) to encode phenotypic features. The recommended way of doing this is to ingest the hp.json file using HpoParser, which in turn creates an HpoConceptRecognizer object. </p>\n",
    "<p>The HpoParser can accept a hpo_json_file argument if you want to use a specific file. If the argument is not passed, it will download the latext hp.json file from the HPO GitHub site and store it in a new subdirectory called hpo_data. It will not download the file if the file is already downloaded.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5a7789fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = HpoParser()\n",
    "hpo_cr = parser.get_hpo_concept_recognizer()\n",
    "hpo_version = parser.get_version()\n",
    "metadata = MetaData(created_by=\"ORCID:0000-0002-5648-2155\")\n",
    "metadata.default_versions_with_hpo(version=hpo_version)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5102722a",
   "metadata": {},
   "source": [
    "<h2>Importing the supplemental table</h2>\n",
    "<p>Here, we use the pandas library to import this file (note that the Python package called openpyxl must be installed to read Excel files with pandas, although the library does not need to be imported in this notebook). pyphetools expects a pandas DataFrame as input, and users can choose any input format available for pandas include CSV, TSV, and Excel, or can use any other method to transform their input data into a Pandas DataFrame before using pyphetools.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f17c3924",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel('input/PMID_28258187.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "06ec7dbd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Clinical features</th>\n",
       "      <th>Patient 1 (11.3 this manuscript)</th>\n",
       "      <th>Patient 2 (11.4 this manuscript)</th>\n",
       "      <th>Patient 3 (Plancke et al, 2009)</th>\n",
       "      <th>Patient 4 (Jergensen et al, 2014)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sex</td>\n",
       "      <td>female</td>\n",
       "      <td>male</td>\n",
       "      <td>female</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Age at examination(years)</td>\n",
       "      <td>7</td>\n",
       "      <td>3.5</td>\n",
       "      <td>10</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mutation in COL3A1</td>\n",
       "      <td>c.145C&gt;G</td>\n",
       "      <td>c.145C&gt;G</td>\n",
       "      <td>c.479dupT</td>\n",
       "      <td>c.1786C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Second mutation in COL3A1</td>\n",
       "      <td>c.145C&gt;G</td>\n",
       "      <td>c.145C&gt;G</td>\n",
       "      <td>c.479dupT</td>\n",
       "      <td>c.3851G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Villefranche criteria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Clinical features Patient 1 (11.3 this manuscript)  \\\n",
       "0                         Sex                           female   \n",
       "1   Age at examination(years)                                7   \n",
       "2          Mutation in COL3A1                         c.145C>G   \n",
       "3  Second mutation in COL3A1                          c.145C>G   \n",
       "4       Villefranche criteria                              NaN   \n",
       "\n",
       "  Patient 2 (11.4 this manuscript) Patient 3 (Plancke et al, 2009)  \\\n",
       "0                             male                          female   \n",
       "1                              3.5                              10   \n",
       "2                         c.145C>G                       c.479dupT   \n",
       "3                         c.145C>G                       c.479dupT   \n",
       "4                              NaN                             NaN   \n",
       "\n",
       "  Patient 4 (Jergensen et al, 2014)  \n",
       "0                            female  \n",
       "1                                19  \n",
       "2                         c.1786C>T  \n",
       "3                         c.3851G>A  \n",
       "4                               NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7f7bad9",
   "metadata": {},
   "source": [
    "<h1>Converting to row-based format</h1>\n",
    "<p>To use pyphetools, we need to have the individuals represented as rows (one row per individual) and have the items of interest be encoded as column names. The required transformations for doing this may be different for different input data, but often we will want to transpose the table (using the pandas <tt>transpose</tt> function) and set the column names of the new table to the zero-th row. After this, we drop the zero-th row (otherwise, it will be interpreted as an individual by the pyphetools code).</p>\n",
    "<p>After this step is completed, the remaining steps to create phenopackets are the same as in the \n",
    "    <a href=\"http://localhost:8888/notebooks/notebooks/Create%20phenopackets%20from%20tabular%20data%20with%20individuals%20in%20rows.ipynb\" target=\"__blank\">row-based notebook</a>.</p>\n",
    "    \n",
    "Furthermore, for this specific case, there is a Count features row that we want dropped, so we filter out any row that does not have Patient in the first column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c85c793f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Clinical features</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age at examination(years)</th>\n",
       "      <th>Mutation in COL3A1</th>\n",
       "      <th>Second mutation in COL3A1</th>\n",
       "      <th>Villefranche criteria</th>\n",
       "      <th>Major features</th>\n",
       "      <th>Minor features</th>\n",
       "      <th>Additional features</th>\n",
       "      <th>Congenital anomalies</th>\n",
       "      <th>Neurological examination</th>\n",
       "      <th>...</th>\n",
       "      <th>Corpus callosum</th>\n",
       "      <th>Hippocampus</th>\n",
       "      <th>Cerebellum</th>\n",
       "      <th>Cortex</th>\n",
       "      <th>White matter_5</th>\n",
       "      <th>Vermis</th>\n",
       "      <th>Post fossa</th>\n",
       "      <th>Pituitary</th>\n",
       "      <th>Arachnoid cysts</th>\n",
       "      <th>Vessels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Patient 1 (11.3 this manuscript)</th>\n",
       "      <td>female</td>\n",
       "      <td>7</td>\n",
       "      <td>c.145C&gt;G</td>\n",
       "      <td>c.145C&gt;G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Global developmental delay, walks without support, uses a few words</td>\n",
       "      <td>...</td>\n",
       "      <td>Present, elongated and mildly thickened</td>\n",
       "      <td>Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Dysplastic cerebellar cortex, multiple cortical cysts superior&gt;inferior</td>\n",
       "      <td>No hypoplasia, multifocal lesions in cerebellar white matter</td>\n",
       "      <td>Vermis hypoplasia cysts</td>\n",
       "      <td>Mega cisterna magna</td>\n",
       "      <td>Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intracranial segment of the A carotis interna is normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient 2 (11.4 this manuscript)</th>\n",
       "      <td>male</td>\n",
       "      <td>3.5</td>\n",
       "      <td>c.145C&gt;G</td>\n",
       "      <td>c.145C&gt;G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Global developmental delay. sits independently, no words</td>\n",
       "      <td>...</td>\n",
       "      <td>Present elongated and mildly thickened</td>\n",
       "      <td>Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Dysplastic cerebellar cortex, multiple cortical cysts superior&gt;inferior</td>\n",
       "      <td>No hypoplasia multifocal lesions in cerebellar white matter</td>\n",
       "      <td>Vermis hypoplasia cysts</td>\n",
       "      <td>Mega cisterna magna</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Bilateral temporal pole arachnoidal cysts</td>\n",
       "      <td>Intracranial segment of the A carotis interna is normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient 3 (Plancke et al, 2009)</th>\n",
       "      <td>female</td>\n",
       "      <td>10</td>\n",
       "      <td>c.479dupT</td>\n",
       "      <td>c.479dupT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Easy bruising thin translucent skin, intestinal/arterial tissue fragility. characteristic facial appearance</td>\n",
       "      <td>Early-onset varicose veins, small joint hypermobility</td>\n",
       "      <td>Pulmonary valve stenosis, pronounced atrophic scars, multiple gingival recessions</td>\n",
       "      <td>Talipes equinovarus</td>\n",
       "      <td>Delayed motor milestones, normal language development</td>\n",
       "      <td>...</td>\n",
       "      <td>Present, elongated</td>\n",
       "      <td>Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cortical cysts</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Bilateral temporal pole arachnoidal cysts</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient 4 (Jergensen et al, 2014)</th>\n",
       "      <td>female</td>\n",
       "      <td>19</td>\n",
       "      <td>c.1786C&gt;T</td>\n",
       "      <td>c.3851G&gt;A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Easy bruising, thin translucent skin, arterial dissections, characteristic facial appearance</td>\n",
       "      <td>Small joint hypermobility, tendon rupture, first-degree relative with vascular Ehlers- Danlos syndrome</td>\n",
       "      <td>Long, slender fingers</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Normal development and neurological examination</td>\n",
       "      <td>...</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Few cortical cysts superior cerebellar lobe</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Mild atrophy</td>\n",
       "      <td>Mega cistema magna</td>\n",
       "      <td>Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Dilatation left A carotis interna, stenosis right A carotis interna</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Clinical features                     Sex Age at examination(years)  \\\n",
       "Patient 1 (11.3 this manuscript)   female                         7   \n",
       "Patient 2 (11.4 this manuscript)     male                       3.5   \n",
       "Patient 3 (Plancke et al, 2009)    female                        10   \n",
       "Patient 4 (Jergensen et al, 2014)  female                        19   \n",
       "\n",
       "Clinical features                 Mutation in COL3A1  \\\n",
       "Patient 1 (11.3 this manuscript)            c.145C>G   \n",
       "Patient 2 (11.4 this manuscript)            c.145C>G   \n",
       "Patient 3 (Plancke et al, 2009)            c.479dupT   \n",
       "Patient 4 (Jergensen et al, 2014)          c.1786C>T   \n",
       "\n",
       "Clinical features                 Second mutation in COL3A1   \\\n",
       "Patient 1 (11.3 this manuscript)                    c.145C>G   \n",
       "Patient 2 (11.4 this manuscript)                    c.145C>G   \n",
       "Patient 3 (Plancke et al, 2009)                    c.479dupT   \n",
       "Patient 4 (Jergensen et al, 2014)                  c.3851G>A   \n",
       "\n",
       "Clinical features                 Villefranche criteria  \\\n",
       "Patient 1 (11.3 this manuscript)                    NaN   \n",
       "Patient 2 (11.4 this manuscript)                    NaN   \n",
       "Patient 3 (Plancke et al, 2009)                     NaN   \n",
       "Patient 4 (Jergensen et al, 2014)                   NaN   \n",
       "\n",
       "Clinical features                                                                                                               Major features  \\\n",
       "Patient 1 (11.3 this manuscript)                                                                                                             -   \n",
       "Patient 2 (11.4 this manuscript)                                                                                                             -   \n",
       "Patient 3 (Plancke et al, 2009)    Easy bruising thin translucent skin, intestinal/arterial tissue fragility. characteristic facial appearance   \n",
       "Patient 4 (Jergensen et al, 2014)                 Easy bruising, thin translucent skin, arterial dissections, characteristic facial appearance   \n",
       "\n",
       "Clinical features                                                                                                          Minor features  \\\n",
       "Patient 1 (11.3 this manuscript)                                                                                                        -   \n",
       "Patient 2 (11.4 this manuscript)                                                                                                        -   \n",
       "Patient 3 (Plancke et al, 2009)                                                     Early-onset varicose veins, small joint hypermobility   \n",
       "Patient 4 (Jergensen et al, 2014)  Small joint hypermobility, tendon rupture, first-degree relative with vascular Ehlers- Danlos syndrome   \n",
       "\n",
       "Clinical features                                                                                Additional features  \\\n",
       "Patient 1 (11.3 this manuscript)                                                                                   -   \n",
       "Patient 2 (11.4 this manuscript)                                                                                   -   \n",
       "Patient 3 (Plancke et al, 2009)    Pulmonary valve stenosis, pronounced atrophic scars, multiple gingival recessions   \n",
       "Patient 4 (Jergensen et al, 2014)                                                              Long, slender fingers   \n",
       "\n",
       "Clinical features                 Congenital anomalies  \\\n",
       "Patient 1 (11.3 this manuscript)                     -   \n",
       "Patient 2 (11.4 this manuscript)                     -   \n",
       "Patient 3 (Plancke et al, 2009)    Talipes equinovarus   \n",
       "Patient 4 (Jergensen et al, 2014)                  NaN   \n",
       "\n",
       "Clinical features                                                             Neurological examination  \\\n",
       "Patient 1 (11.3 this manuscript)   Global developmental delay, walks without support, uses a few words   \n",
       "Patient 2 (11.4 this manuscript)              Global developmental delay. sits independently, no words   \n",
       "Patient 3 (Plancke et al, 2009)                  Delayed motor milestones, normal language development   \n",
       "Patient 4 (Jergensen et al, 2014)                      Normal development and neurological examination   \n",
       "\n",
       "Clinical features                  ...  \\\n",
       "Patient 1 (11.3 this manuscript)   ...   \n",
       "Patient 2 (11.4 this manuscript)   ...   \n",
       "Patient 3 (Plancke et al, 2009)    ...   \n",
       "Patient 4 (Jergensen et al, 2014)  ...   \n",
       "\n",
       "Clinical features                                          Corpus callosum  \\\n",
       "Patient 1 (11.3 this manuscript)   Present, elongated and mildly thickened   \n",
       "Patient 2 (11.4 this manuscript)    Present elongated and mildly thickened   \n",
       "Patient 3 (Plancke et al, 2009)                         Present, elongated   \n",
       "Patient 4 (Jergensen et al, 2014)                                   Normal   \n",
       "\n",
       "Clinical features                 Hippocampus Cerebellum  \\\n",
       "Patient 1 (11.3 this manuscript)       Normal        NaN   \n",
       "Patient 2 (11.4 this manuscript)       Normal        NaN   \n",
       "Patient 3 (Plancke et al, 2009)        Normal        NaN   \n",
       "Patient 4 (Jergensen et al, 2014)      Normal        NaN   \n",
       "\n",
       "Clinical features                                                                                   Cortex  \\\n",
       "Patient 1 (11.3 this manuscript)   Dysplastic cerebellar cortex, multiple cortical cysts superior>inferior   \n",
       "Patient 2 (11.4 this manuscript)   Dysplastic cerebellar cortex, multiple cortical cysts superior>inferior   \n",
       "Patient 3 (Plancke et al, 2009)                                                             Cortical cysts   \n",
       "Patient 4 (Jergensen et al, 2014)                              Few cortical cysts superior cerebellar lobe   \n",
       "\n",
       "Clinical features                                                                White matter_5  \\\n",
       "Patient 1 (11.3 this manuscript)   No hypoplasia, multifocal lesions in cerebellar white matter   \n",
       "Patient 2 (11.4 this manuscript)    No hypoplasia multifocal lesions in cerebellar white matter   \n",
       "Patient 3 (Plancke et al, 2009)                                                          Normal   \n",
       "Patient 4 (Jergensen et al, 2014)                                                        Normal   \n",
       "\n",
       "Clinical features                                   Vermis  \\\n",
       "Patient 1 (11.3 this manuscript)   Vermis hypoplasia cysts   \n",
       "Patient 2 (11.4 this manuscript)   Vermis hypoplasia cysts   \n",
       "Patient 3 (Plancke et al, 2009)                     Normal   \n",
       "Patient 4 (Jergensen et al, 2014)             Mild atrophy   \n",
       "\n",
       "Clinical features                           Post fossa Pituitary  \\\n",
       "Patient 1 (11.3 this manuscript)   Mega cisterna magna    Normal   \n",
       "Patient 2 (11.4 this manuscript)   Mega cisterna magna    Normal   \n",
       "Patient 3 (Plancke et al, 2009)                 Normal    Normal   \n",
       "Patient 4 (Jergensen et al, 2014)   Mega cistema magna    Normal   \n",
       "\n",
       "Clinical features                                            Arachnoid cysts  \\\n",
       "Patient 1 (11.3 this manuscript)                                         NaN   \n",
       "Patient 2 (11.4 this manuscript)   Bilateral temporal pole arachnoidal cysts   \n",
       "Patient 3 (Plancke et al, 2009)    Bilateral temporal pole arachnoidal cysts   \n",
       "Patient 4 (Jergensen et al, 2014)                                        NaN   \n",
       "\n",
       "Clinical features                                                                              Vessels  \n",
       "Patient 1 (11.3 this manuscript)               Intracranial segment of the A carotis interna is normal  \n",
       "Patient 2 (11.4 this manuscript)               Intracranial segment of the A carotis interna is normal  \n",
       "Patient 3 (Plancke et al, 2009)                                                                    NaN  \n",
       "Patient 4 (Jergensen et al, 2014)  Dilatation left A carotis interna, stenosis right A carotis interna  \n",
       "\n",
       "[4 rows x 37 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft = df.transpose()\n",
    "dft.columns = dft.iloc[0]\n",
    "dft.drop(dft.index[0], inplace=True)\n",
    "dft.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e013666d",
   "metadata": {},
   "source": [
    "Some column names might include spaces in front or after, and a couple of columns are subheadings and only contain NaNs, so lets correct that:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5d604600",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Clinical features</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age at examination(years)</th>\n",
       "      <th>Mutation in COL3A1</th>\n",
       "      <th>Second mutation in COL3A1</th>\n",
       "      <th>Major features</th>\n",
       "      <th>Minor features</th>\n",
       "      <th>Additional features</th>\n",
       "      <th>Congenital anomalies</th>\n",
       "      <th>Neurological examination</th>\n",
       "      <th>Head circumference</th>\n",
       "      <th>...</th>\n",
       "      <th>Corpus callosum</th>\n",
       "      <th>Hippocampus</th>\n",
       "      <th>Cortex</th>\n",
       "      <th>White matter_5</th>\n",
       "      <th>Vermis</th>\n",
       "      <th>Post fossa</th>\n",
       "      <th>Pituitary</th>\n",
       "      <th>Arachnoid cysts</th>\n",
       "      <th>Vessels</th>\n",
       "      <th>patient_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Patient 1 (11.3 this manuscript)</th>\n",
       "      <td>female</td>\n",
       "      <td>7</td>\n",
       "      <td>c.145C&gt;G</td>\n",
       "      <td>c.145C&gt;G</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Global developmental delay, walks without support, uses a few words</td>\n",
       "      <td>90th centile</td>\n",
       "      <td>...</td>\n",
       "      <td>Present, elongated and mildly thickened</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Dysplastic cerebellar cortex, multiple cortical cysts superior&gt;inferior</td>\n",
       "      <td>No hypoplasia, multifocal lesions in cerebellar white matter</td>\n",
       "      <td>Vermis hypoplasia cysts</td>\n",
       "      <td>Mega cisterna magna</td>\n",
       "      <td>Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intracranial segment of the A carotis interna is normal</td>\n",
       "      <td>Patient 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient 2 (11.4 this manuscript)</th>\n",
       "      <td>male</td>\n",
       "      <td>3.5</td>\n",
       "      <td>c.145C&gt;G</td>\n",
       "      <td>c.145C&gt;G</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Global developmental delay. sits independently, no words</td>\n",
       "      <td>&gt;97th centile</td>\n",
       "      <td>...</td>\n",
       "      <td>Present elongated and mildly thickened</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Dysplastic cerebellar cortex, multiple cortical cysts superior&gt;inferior</td>\n",
       "      <td>No hypoplasia multifocal lesions in cerebellar white matter</td>\n",
       "      <td>Vermis hypoplasia cysts</td>\n",
       "      <td>Mega cisterna magna</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Bilateral temporal pole arachnoidal cysts</td>\n",
       "      <td>Intracranial segment of the A carotis interna is normal</td>\n",
       "      <td>Patient 2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Clinical features                    Sex Age at examination(years)  \\\n",
       "Patient 1 (11.3 this manuscript)  female                         7   \n",
       "Patient 2 (11.4 this manuscript)    male                       3.5   \n",
       "\n",
       "Clinical features                Mutation in COL3A1 Second mutation in COL3A1  \\\n",
       "Patient 1 (11.3 this manuscript)           c.145C>G                  c.145C>G   \n",
       "Patient 2 (11.4 this manuscript)           c.145C>G                  c.145C>G   \n",
       "\n",
       "Clinical features                Major features Minor features  \\\n",
       "Patient 1 (11.3 this manuscript)              -              -   \n",
       "Patient 2 (11.4 this manuscript)              -              -   \n",
       "\n",
       "Clinical features                Additional features Congenital anomalies  \\\n",
       "Patient 1 (11.3 this manuscript)                   -                    -   \n",
       "Patient 2 (11.4 this manuscript)                   -                    -   \n",
       "\n",
       "Clinical features                                                            Neurological examination  \\\n",
       "Patient 1 (11.3 this manuscript)  Global developmental delay, walks without support, uses a few words   \n",
       "Patient 2 (11.4 this manuscript)             Global developmental delay. sits independently, no words   \n",
       "\n",
       "Clinical features                Head circumference  ...  \\\n",
       "Patient 1 (11.3 this manuscript)       90th centile  ...   \n",
       "Patient 2 (11.4 this manuscript)      >97th centile  ...   \n",
       "\n",
       "Clinical features                                         Corpus callosum  \\\n",
       "Patient 1 (11.3 this manuscript)  Present, elongated and mildly thickened   \n",
       "Patient 2 (11.4 this manuscript)   Present elongated and mildly thickened   \n",
       "\n",
       "Clinical features                Hippocampus  \\\n",
       "Patient 1 (11.3 this manuscript)      Normal   \n",
       "Patient 2 (11.4 this manuscript)      Normal   \n",
       "\n",
       "Clinical features                                                                                  Cortex  \\\n",
       "Patient 1 (11.3 this manuscript)  Dysplastic cerebellar cortex, multiple cortical cysts superior>inferior   \n",
       "Patient 2 (11.4 this manuscript)  Dysplastic cerebellar cortex, multiple cortical cysts superior>inferior   \n",
       "\n",
       "Clinical features                                                               White matter_5  \\\n",
       "Patient 1 (11.3 this manuscript)  No hypoplasia, multifocal lesions in cerebellar white matter   \n",
       "Patient 2 (11.4 this manuscript)   No hypoplasia multifocal lesions in cerebellar white matter   \n",
       "\n",
       "Clinical features                                  Vermis  \\\n",
       "Patient 1 (11.3 this manuscript)  Vermis hypoplasia cysts   \n",
       "Patient 2 (11.4 this manuscript)  Vermis hypoplasia cysts   \n",
       "\n",
       "Clinical features                          Post fossa Pituitary  \\\n",
       "Patient 1 (11.3 this manuscript)  Mega cisterna magna    Normal   \n",
       "Patient 2 (11.4 this manuscript)  Mega cisterna magna    Normal   \n",
       "\n",
       "Clinical features                                           Arachnoid cysts  \\\n",
       "Patient 1 (11.3 this manuscript)                                        NaN   \n",
       "Patient 2 (11.4 this manuscript)  Bilateral temporal pole arachnoidal cysts   \n",
       "\n",
       "Clinical features                                                                 Vessels  \\\n",
       "Patient 1 (11.3 this manuscript)  Intracranial segment of the A carotis interna is normal   \n",
       "Patient 2 (11.4 this manuscript)  Intracranial segment of the A carotis interna is normal   \n",
       "\n",
       "Clinical features                 patient_id  \n",
       "Patient 1 (11.3 this manuscript)  Patient 1   \n",
       "Patient 2 (11.4 this manuscript)  Patient 2   \n",
       "\n",
       "[2 rows x 35 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "dft.columns = dft.columns.str.strip()\n",
    "dft = dft.dropna(axis=1, how='all')\n",
    "dft['patient_id'] = dft.index\n",
    "# simplify the name of the id column to remove e.g.,(11.3 this manuscript) f\n",
    "dft['patient_id'] = dft['patient_id'].transform(lambda x: re.sub(r\"\\(.*?\\)\", \"\", x) )\n",
    "dft.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8314d6bd",
   "metadata": {},
   "source": [
    "<h2>Column mappers</h2>\n",
    "<p>Please see the notebook \"Create phenopackets from tabular data with individuals in rows\" for explanations. In the following cell we create a dictionary for the ColumnMappers. Note that the code is identical except that we use the df.loc function to get the corresponding row data</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2305a0d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "hpo_cr = parser.get_hpo_concept_recognizer()\n",
    "column_mapper_d = defaultdict(ColumnMapper)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ecb20c7",
   "metadata": {},
   "source": [
    "Lets try to get code autoformatted so that we can easily copy-paste and change it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fa281c54",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "age_at_examination(years)_d = {'3.5': 'PLACEHOLDER',\n",
      " '10': 'PLACEHOLDER',\n",
      " '19': 'PLACEHOLDER'}\n",
      "age_at_examination(years)Mapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=age_at_examination(years)_d)\n",
      "age_at_examination(years)Mapper.preview_column(df['Age at examination(years)']))\n",
      "column_mapper_d['Age at examination(years)'] = age_at_examination(years)Mapper\n",
      "\n",
      "major_features_d = {'Easy bruising thin translucent skin': 'Dermal translucency',\n",
      " 'intestinal/arterial tissue fragility. characteristic facial appearance': 'PLACEHOLDER',\n",
      " 'Easy bruising': 'Bruising susceptibility',\n",
      " 'thin translucent skin': 'Dermal translucency',\n",
      " 'arterial dissections': 'Arterial dissection',\n",
      " 'characteristic facial appearance': 'PLACEHOLDER'}\n",
      "major_featuresMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=major_features_d)\n",
      "major_featuresMapper.preview_column(df['Major features']))\n",
      "column_mapper_d['Major features'] = major_featuresMapper\n",
      "\n",
      "minor_features_d = {'Early-onset varicose veins': 'Varicose veins',\n",
      " 'small joint hypermobility': 'Joint hypermobility',\n",
      " 'Small joint hypermobility': 'Joint hypermobility',\n",
      " 'tendon rupture': 'Tendon rupture',\n",
      " 'first-degree relative with vascular Ehlers- Danlos syndrome': 'PLACEHOLDER'}\n",
      "minor_featuresMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=minor_features_d)\n",
      "minor_featuresMapper.preview_column(df['Minor features']))\n",
      "column_mapper_d['Minor features'] = minor_featuresMapper\n",
      "\n",
      "additional_features_d = {'Pulmonary valve stenosis': 'Pulmonic stenosis',\n",
      " 'pronounced atrophic scars': 'Atrophic scars',\n",
      " 'multiple gingival recessions': 'Gingival recession',\n",
      " 'Long': 'PLACEHOLDER',\n",
      " 'slender fingers': 'Slender finger'}\n",
      "additional_featuresMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=additional_features_d)\n",
      "additional_featuresMapper.preview_column(df['Additional features']))\n",
      "column_mapper_d['Additional features'] = additional_featuresMapper\n",
      "\n",
      "congenital_anomalies_d = {'Talipes equinovarus': 'Talipes equinovarus',\n",
      " 'nan': 'PLACEHOLDER'}\n",
      "congenital_anomaliesMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=congenital_anomalies_d)\n",
      "congenital_anomaliesMapper.preview_column(df['Congenital anomalies']))\n",
      "column_mapper_d['Congenital anomalies'] = congenital_anomaliesMapper\n",
      "\n",
      "neurological_examination_d = {'Global developmental delay': 'Global developmental delay',\n",
      " 'walks without support': 'PLACEHOLDER',\n",
      " 'uses a few words': 'PLACEHOLDER',\n",
      " 'Global developmental delay. sits independently': 'Global developmental delay',\n",
      " 'no words': 'PLACEHOLDER',\n",
      " 'Delayed motor milestones': 'Motor delay',\n",
      " 'normal language development': 'PLACEHOLDER',\n",
      " 'Normal development and neurological examination': 'PLACEHOLDER'}\n",
      "neurological_examinationMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=neurological_examination_d)\n",
      "neurological_examinationMapper.preview_column(df['Neurological examination']))\n",
      "column_mapper_d['Neurological examination'] = neurological_examinationMapper\n",
      "\n",
      "head_circumference_d = {'90th centile': 'PLACEHOLDER',\n",
      " '>97th centile': 'PLACEHOLDER',\n",
      " '75th centile': 'PLACEHOLDER',\n",
      " 'nan': 'PLACEHOLDER'}\n",
      "head_circumferenceMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=head_circumference_d)\n",
      "head_circumferenceMapper.preview_column(df['Head circumference']))\n",
      "column_mapper_d['Head circumference'] = head_circumferenceMapper\n",
      "\n",
      "epilepsy/onset_d = {'Spasms/5 years': 'PLACEHOLDER',\n",
      " 'Spasms/26 months': 'PLACEHOLDER',\n",
      " 'Absence seizures/unknown': 'Generalized non-motor (absence) seizure'}\n",
      "epilepsy/onsetMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=epilepsy/onset_d)\n",
      "epilepsy/onsetMapper.preview_column(df['Epilepsy/onset']))\n",
      "column_mapper_d['Epilepsy/onset'] = epilepsy/onsetMapper\n",
      "\n",
      "age_at_mri_d = {'34 months': 'PLACEHOLDER',\n",
      " '6 months': 'PLACEHOLDER',\n",
      " '15 months': 'PLACEHOLDER',\n",
      " '19 years': 'PLACEHOLDER'}\n",
      "age_at_mriMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=age_at_mri_d)\n",
      "age_at_mriMapper.preview_column(df['Age at MRI']))\n",
      "column_mapper_d['Age at MRI'] = age_at_mriMapper\n",
      "\n",
      "gyral_pattern_d = {'Diffuse thickened cobblestone cortex with relative sparing of the temporal lobes': 'PLACEHOLDER',\n",
      " 'Diffuse thickened cobblestone cortex with relative sparing of the temporal poles': 'PLACEHOLDER',\n",
      " 'Frontal cobblestone cortex': 'PLACEHOLDER',\n",
      " 'parietal polymicrogyria': 'Polymicrogyria',\n",
      " 'relative sparing of the temporal lobes': 'PLACEHOLDER',\n",
      " 'Bilateral frontal polymicrogyria including cingulate gyri': 'Frontal polymicrogyria'}\n",
      "gyral_patternMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=gyral_pattern_d)\n",
      "gyral_patternMapper.preview_column(df['Gyral pattern']))\n",
      "column_mapper_d['Gyral pattern'] = gyral_patternMapper\n",
      "\n",
      "gradient_d = {'A>P': 'PLACEHOLDER',\n",
      " 'No gradient': 'PLACEHOLDER'}\n",
      "gradientMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=gradient_d)\n",
      "gradientMapper.preview_column(df['Gradient']))\n",
      "column_mapper_d['Gradient'] = gradientMapper\n",
      "\n",
      "periventricular_region_d = {'Capping of the frontal>posterior horns of the lateral ventricles': 'PLACEHOLDER',\n",
      " 'Capping of the posterior homs of the lateral ventricles': 'PLACEHOLDER',\n",
      " 'Normal': 'PLACEHOLDER'}\n",
      "periventricular_regionMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=periventricular_region_d)\n",
      "periventricular_regionMapper.preview_column(df['Periventricular region']))\n",
      "column_mapper_d['Periventricular region'] = periventricular_regionMapper\n",
      "\n",
      "white_matter_d = {'Globale reduction of white matter': 'PLACEHOLDER',\n",
      " 'Diffuse hypomyelination': 'PLACEHOLDER',\n",
      " 'Global reduction of white matter': 'PLACEHOLDER',\n",
      " 'Normal': 'PLACEHOLDER'}\n",
      "white_matterMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=white_matter_d)\n",
      "white_matterMapper.preview_column(df['White matter']))\n",
      "column_mapper_d['White matter'] = white_matterMapper\n",
      "\n",
      "white_matter_2_d = {'Prominent perivascular spaces': 'Dilation of Virchow-Robin spaces',\n",
      " 'Prominent perivascular spaces bilateral frontal': 'Dilation of Virchow-Robin spaces'}\n",
      "white_matter_2Mapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=white_matter_2_d)\n",
      "white_matter_2Mapper.preview_column(df['White matter_2']))\n",
      "column_mapper_d['White matter_2'] = white_matter_2Mapper\n",
      "\n",
      "white_matter_3_d = {'Stripe sign/tigroid pattern with involvement of U-fibres': 'PLACEHOLDER'}\n",
      "white_matter_3Mapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=white_matter_3_d)\n",
      "white_matter_3Mapper.preview_column(df['White matter_3']))\n",
      "column_mapper_d['White matter_3'] = white_matter_3Mapper\n",
      "\n",
      "white_matter_4_d = {'Frontal nodular heterotopia (beads)': 'Gray matter heterotopia',\n",
      " 'perisylvian and occipital band heterotopia': 'Gray matter heterotopia'}\n",
      "white_matter_4Mapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=white_matter_4_d)\n",
      "white_matter_4Mapper.preview_column(df['White matter_4']))\n",
      "column_mapper_d['White matter_4'] = white_matter_4Mapper\n",
      "\n",
      "calcification_d = {'nan': 'PLACEHOLDER',\n",
      " 'No indication for calcifications on MRI gradient echo': 'PLACEHOLDER',\n",
      " 'No calcifications on CT': 'PLACEHOLDER'}\n",
      "calcificationMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=calcification_d)\n",
      "calcificationMapper.preview_column(df['Calcification']))\n",
      "column_mapper_d['Calcification'] = calcificationMapper\n",
      "\n",
      "lateral_ventricles_d = {'Ventriculomegaly': 'Ventriculomegaly',\n",
      " 'Mild enlargement': 'PLACEHOLDER',\n",
      " 'Normal': 'PLACEHOLDER'}\n",
      "lateral_ventriclesMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=lateral_ventricles_d)\n",
      "lateral_ventriclesMapper.preview_column(df['Lateral ventricles']))\n",
      "column_mapper_d['Lateral ventricles'] = lateral_ventriclesMapper\n",
      "\n",
      "third_ventricle_d = {'Ventriculomegaly': 'Ventriculomegaly',\n",
      " 'Mild enlargement': 'PLACEHOLDER',\n",
      " 'Normal': 'PLACEHOLDER'}\n",
      "third_ventricleMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=third_ventricle_d)\n",
      "third_ventricleMapper.preview_column(df['Third ventricle']))\n",
      "column_mapper_d['Third ventricle'] = third_ventricleMapper\n",
      "\n",
      "fourth_venticle_d = {'Normal': 'PLACEHOLDER'}\n",
      "fourth_venticleMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=fourth_venticle_d)\n",
      "fourth_venticleMapper.preview_column(df['Fourth venticle']))\n",
      "column_mapper_d['Fourth venticle'] = fourth_venticleMapper\n",
      "\n",
      "brainstem_d = {'Hypoplastic': 'PLACEHOLDER',\n",
      " 'Mildly hypoplastic': 'PLACEHOLDER',\n",
      " 'Hypoplasia of the pons': 'Hypoplasia of the pons',\n",
      " 'Normal': 'PLACEHOLDER'}\n",
      "brainstemMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=brainstem_d)\n",
      "brainstemMapper.preview_column(df['Brainstem']))\n",
      "column_mapper_d['Brainstem'] = brainstemMapper\n",
      "\n",
      "basal_ganglia_d = {'Thalamus normal putamen/globus pallidus small': 'PLACEHOLDER',\n",
      " 'accentuated Virchow- Robin spaces': 'PLACEHOLDER',\n",
      " 'Normal volume and signal intensity': 'PLACEHOLDER',\n",
      " 'accentuated Virchow-Robin spaces': 'PLACEHOLDER',\n",
      " 'Normal': 'PLACEHOLDER'}\n",
      "basal_gangliaMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=basal_ganglia_d)\n",
      "basal_gangliaMapper.preview_column(df['Basal ganglia']))\n",
      "column_mapper_d['Basal ganglia'] = basal_gangliaMapper\n",
      "\n",
      "corpus_callosum_d = {'Present': 'PLACEHOLDER',\n",
      " 'elongated and mildly thickened': 'PLACEHOLDER',\n",
      " 'Present elongated and mildly thickened': 'PLACEHOLDER',\n",
      " 'elongated': 'PLACEHOLDER',\n",
      " 'Normal': 'PLACEHOLDER'}\n",
      "corpus_callosumMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=corpus_callosum_d)\n",
      "corpus_callosumMapper.preview_column(df['Corpus callosum']))\n",
      "column_mapper_d['Corpus callosum'] = corpus_callosumMapper\n",
      "\n",
      "hippocampus_d = {'Normal': 'PLACEHOLDER'}\n",
      "hippocampusMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=hippocampus_d)\n",
      "hippocampusMapper.preview_column(df['Hippocampus']))\n",
      "column_mapper_d['Hippocampus'] = hippocampusMapper\n",
      "\n",
      "cortex_d = {'Dysplastic cerebellar cortex': 'PLACEHOLDER',\n",
      " 'multiple cortical cysts superior>inferior': 'Renal cortical cysts',\n",
      " 'Cortical cysts': 'Renal cortical cysts',\n",
      " 'Few cortical cysts superior cerebellar lobe': 'Renal cortical cysts'}\n",
      "cortexMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=cortex_d)\n",
      "cortexMapper.preview_column(df['Cortex']))\n",
      "column_mapper_d['Cortex'] = cortexMapper\n",
      "\n",
      "white_matter_5_d = {'No hypoplasia': 'PLACEHOLDER',\n",
      " 'multifocal lesions in cerebellar white matter': 'PLACEHOLDER',\n",
      " 'No hypoplasia multifocal lesions in cerebellar white matter': 'PLACEHOLDER',\n",
      " 'Normal': 'PLACEHOLDER'}\n",
      "white_matter_5Mapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=white_matter_5_d)\n",
      "white_matter_5Mapper.preview_column(df['White matter_5']))\n",
      "column_mapper_d['White matter_5'] = white_matter_5Mapper\n",
      "\n",
      "vermis_d = {'Vermis hypoplasia cysts': 'PLACEHOLDER',\n",
      " 'Normal': 'PLACEHOLDER',\n",
      " 'Mild atrophy': 'PLACEHOLDER'}\n",
      "vermisMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=vermis_d)\n",
      "vermisMapper.preview_column(df['Vermis']))\n",
      "column_mapper_d['Vermis'] = vermisMapper\n",
      "\n",
      "post_fossa_d = {'Mega cisterna magna': 'Enlarged cisterna magna',\n",
      " 'Normal': 'PLACEHOLDER',\n",
      " 'Mega cistema magna': 'PLACEHOLDER'}\n",
      "post_fossaMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=post_fossa_d)\n",
      "post_fossaMapper.preview_column(df['Post fossa']))\n",
      "column_mapper_d['Post fossa'] = post_fossaMapper\n",
      "\n",
      "pituitary_d = {'Normal': 'PLACEHOLDER'}\n",
      "pituitaryMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=pituitary_d)\n",
      "pituitaryMapper.preview_column(df['Pituitary']))\n",
      "column_mapper_d['Pituitary'] = pituitaryMapper\n",
      "\n",
      "arachnoid_cysts_d = {'nan': 'PLACEHOLDER',\n",
      " 'Bilateral temporal pole arachnoidal cysts': 'PLACEHOLDER'}\n",
      "arachnoid_cystsMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=arachnoid_cysts_d)\n",
      "arachnoid_cystsMapper.preview_column(df['Arachnoid cysts']))\n",
      "column_mapper_d['Arachnoid cysts'] = arachnoid_cystsMapper\n",
      "\n",
      "vessels_d = {'Intracranial segment of the A carotis interna is normal': 'Acrania',\n",
      " 'nan': 'PLACEHOLDER',\n",
      " 'Dilatation left A carotis interna': 'PLACEHOLDER',\n",
      " 'stenosis right A carotis interna': 'PLACEHOLDER'}\n",
      "vesselsMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=vessels_d)\n",
      "vesselsMapper.preview_column(df['Vessels']))\n",
      "column_mapper_d['Vessels'] = vesselsMapper\n",
      "\n"
     ]
    }
   ],
   "source": [
    "output = OptionColumnMapper.autoformat(df=dft, concept_recognizer=hpo_cr)\n",
    "print(output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "88e52430",
   "metadata": {},
   "outputs": [],
   "source": [
    "                                         \n",
    "major_features = {'Easy bruising thin translucent skin': 'Bruising susceptibility',\n",
    " 'arterial tissue fragility': 'Abnormal arterial physiology',\n",
    " 'Easy bruising': 'Bruising susceptibility',\n",
    " 'thin translucent skin': 'Dermal translucency',\n",
    " 'arterial dissections': 'Arterial dissection',}\n",
    "major_featuresMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=major_features)\n",
    "#print(major_featuresMapper.preview_column(dft['Major features']))\n",
    "column_mapper_d['Major features'] = major_featuresMapper\n",
    "\n",
    "cortex = {'Dysplastic cerebellar cortex': 'Abnormal cerebellar cortex morphology'}\n",
    "cortexMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=cortex)\n",
    "#print(cortexMapper.preview_column(dft['Cortex']))\n",
    "column_mapper_d['Cortex'] = cortexMapper\n",
    "\n",
    "minor_features = {'Early-onset varicose veins': 'Varicose veins',\n",
    " 'small joint hypermobility': 'Joint hypermobility',\n",
    " 'tendon rupture': 'Tendon rupture'}\n",
    "minor_featuresMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=minor_features)\n",
    "#print(minor_featuresMapper.preview_column(dft['Minor features']))\n",
    "column_mapper_d['Minor features'] = minor_featuresMapper\n",
    "\n",
    "additional_features = {'Pulmonary valve stenosis': 'Pulmonic stenosis',\n",
    " 'pronounced atrophic scars': 'Atrophic scars',\n",
    " 'multiple gingival recessions': 'Gingival recession',\n",
    " 'slender fingers': 'Slender finger'}\n",
    "additional_featuresMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=additional_features)\n",
    "#print(major_featuresMapper.preview_column(dft['Additional features']))\n",
    "column_mapper_d['Additional features'] = additional_featuresMapper\n",
    "\n",
    "\n",
    "congenital_anomalies = {'Talipes equinovarus': 'Talipes equinovarus'}\n",
    "congenital_anomaliesMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=congenital_anomalies)\n",
    "#print(congenital_anomaliesMapper.preview_column(dft['Congenital anomalies']))\n",
    "column_mapper_d['Congenital anomalies'] = congenital_anomaliesMapper\n",
    "\n",
    "neurological_examination = {'Global developmental delay': 'Global developmental delay',\n",
    " 'no words': 'Absent speech',\n",
    " 'Delayed motor milestones': 'Motor delay'}\n",
    "neurological_examinationMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=neurological_examination)\n",
    "#print(neurological_examinationMapper.preview_column(dft['Neurological examination']))\n",
    "column_mapper_d['Neurological examination'] = neurological_examinationMapper\n",
    "\n",
    "head_circumference = {'>97th centile': 'Macrocephaly'}\n",
    "head_circumferenceMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=head_circumference)\n",
    "#print(head_circumferenceMapper.preview_column(dft['Head circumference']))\n",
    "column_mapper_d['Head circumference'] = head_circumferenceMapper\n",
    "\n",
    "epilepsy_onset = {'Spasms/5 years': 'Seizure',\n",
    " 'Spasms/26 months': 'Seizure',\n",
    " 'Absence seizures/unknown': 'Typical absence seizure'}\n",
    "epilepsy_onsetMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=epilepsy_onset)\n",
    "#print(epilepsy_onsetMapper.preview_column(dft['Epilepsy/onset']))\n",
    "column_mapper_d['Epilepsy/onset'] = epilepsy_onsetMapper\n",
    "\n",
    "\n",
    "gyral_pattern = {'Diffuse thickened cobblestone cortex with relative sparing of the temporal lobes': 'Dysgyria with thickened cortex',\n",
    " 'Diffuse thickened cobblestone cortex with relative sparing of the temporal poles': 'Dysgyria with thickened cortex',\n",
    " 'Frontal cobblestone cortex': 'Dysgyria with thickened cortex',\n",
    " 'parietal polymicrogyria': 'Polymicrogyria',\n",
    " 'Bilateral frontal polymicrogyria including cingulate gyri': 'Polymicrogyria'}\n",
    "gyral_patternMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=gyral_pattern)\n",
    "#print(gyral_patternMapper.preview_column(dft['Gyral pattern']))\n",
    "column_mapper_d['Gyral pattern'] = gyral_patternMapper\n",
    "\n",
    "\n",
    "white_matter = {'Globale reduction of white matter': 'Hypointensity of cerebral white matter on MRI',\n",
    " 'Diffuse hypomyelination': 'Cerebral hypomyelination'}\n",
    "white_matterMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=white_matter)\n",
    "#print(white_matterMapper.preview_column(dft['White matter']))\n",
    "column_mapper_d['White matter'] = white_matterMapper\n",
    "\n",
    "white_matter_2 = {'Prominent perivascular spaces': 'Dilation of Virchow-Robin spaces',\n",
    " 'Prominent perivascular spaces bilateral frontal': 'Dilation of Virchow-Robin spaces'}\n",
    "white_matter_2Mapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=white_matter_2)\n",
    "#print(white_matter_2Mapper .preview_column(dft['White matter_2']))\n",
    "column_mapper_d['White matter_2'] = white_matter_2Mapper\n",
    "\n",
    "white_matter_3 = {'Frontal nodular heterotopia (beads)': 'Gray matter heterotopia',\n",
    " 'perisylvian and occipital band heterotopia': 'Gray matter heterotopia'}\n",
    "white_matter_3Mapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=white_matter_3)\n",
    "#print(white_matter_3Mapper.preview_column(dft['White matter_3']))\n",
    "column_mapper_d['White matter_3'] = white_matter_3Mapper\n",
    "\n",
    "\n",
    "lateral_ventricles = {'Ventriculomegaly': 'Lateral ventricle dilatation',\n",
    " 'Mild enlargement': 'Lateral ventricle dilatation'}\n",
    "lateral_ventriclesMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=lateral_ventricles)\n",
    "#print(lateral_ventriclesMapper.preview_column(dft['Lateral ventricles']))\n",
    "column_mapper_d['Lateral ventricles'] = lateral_ventriclesMapper\n",
    "\n",
    "third_ventricle = {'Ventriculomegaly': 'Dilated third ventricle',\n",
    " 'Mild enlargement': 'Dilated third ventricle'}\n",
    "third_ventricleMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=third_ventricle)\n",
    "#print(third_ventricleMapper.preview_column(dft['Third ventricle']))\n",
    "column_mapper_d['Third ventricle'] = third_ventricleMapper\n",
    "\n",
    "brainstem = {'Hypoplastic': 'Abnormal brainstem morphology',\n",
    " 'Mildly hypoplastic': 'Abnormal brainstem morphology',\n",
    " 'Hypoplasia of the pons': 'Hypoplasia of the pons'}\n",
    "brainstemMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=brainstem)\n",
    "#print(brainstemMapper.preview_column(dft['Brainstem']))\n",
    "column_mapper_d['Brainstem'] = brainstemMapper\n",
    "\n",
    "basal_ganglia = {'Thalamus normal putamen/globus pallidus small': 'Abnormal globus pallidus morphology',\n",
    " 'accentuated Virchow- Robin spaces': 'Dilation of Virchow-Robin spaces'}\n",
    "basal_gangliaMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=basal_ganglia)\n",
    "#print(basal_gangliaMapper.preview_column(dft['Basal ganglia']))\n",
    "column_mapper_d['Basal ganglia'] = basal_gangliaMapper\n",
    "\n",
    "corpus_callosum = {'elongated and mildly thickened': 'Abnormal length of corpus callosum',\n",
    " 'Present elongated and mildly thickened': 'Abnormal length of corpus callosum',\n",
    " 'elongated': 'Abnormal length of corpus callosum'}\n",
    "corpus_callosumMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=corpus_callosum)\n",
    "#print(corpus_callosumMapper.preview_column(dft['Corpus callosum']))\n",
    "column_mapper_d['Corpus callosum'] = corpus_callosumMapper\n",
    "\n",
    "vermis = {'Vermis hypoplasia cysts': 'Cerebellar vermis hypoplasia',\n",
    " 'Mild atrophy': 'Cerebellar vermis hypoplasia'}\n",
    "vermisMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=vermis)\n",
    "#print(vermisMapper.preview_column(dft['Vermis']))\n",
    "column_mapper_d['Vermis'] = vermisMapper\n",
    "\n",
    "post_fossa = {'Mega cisterna magna': 'Enlarged cisterna magna',\n",
    " 'Mega cistema magna': 'Enlarged cisterna magna'}\n",
    "post_fossaMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=post_fossa)\n",
    "#print(post_fossaMapper.preview_column(dft['Post fossa']))\n",
    "column_mapper_d['Post fossa'] = post_fossaMapper\n",
    "\n",
    "vessels = { 'Dilatation left A carotis interna': 'Carotid artery dilatation',\n",
    " 'stenosis right A carotis interna': 'Carotid artery stenosis'}\n",
    "vesselsMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=vessels)\n",
    "#print(vesselsMapper.preview_column(dft['Vessels']))\n",
    "column_mapper_d['Vessels'] = vesselsMapper\n",
    "\n",
    "arachnoid_cysts = {'Bilateral temporal pole arachnoidal cysts': 'Arachnoid cyst'}\n",
    "arachnoid_cystsMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=arachnoid_cysts)\n",
    "#print(arachnoid_cystsMapper.preview_column(dft['Arachnoid cysts']))\n",
    "column_mapper_d['Arachnoid cysts'] = arachnoid_cystsMapper\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e53617d",
   "metadata": {},
   "source": [
    "<h2>Variant Data</h2>\n",
    "<p>The variant data (HGVS< transcript) is listed in the Variant (hg19, NM_015133.4) column.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "da4d5706",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = 'hg38'\n",
    "default_genotype = 'homozygous'\n",
    "transcript='NM_000090.3'\n",
    "varMapper = VariantColumnMapper(assembly=genome,column_name='Mutation in COL3A1', \n",
    "                                transcript=transcript, default_genotype=default_genotype)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1af2052",
   "metadata": {},
   "source": [
    "<h1>Demographic data</h1>\n",
    "<p>pyphetools can be used to capture information about age, sex, and individual identifiers. This information is stored in a map of \"IndividualMapper\" objects. Special treatment may be required for the indifiers, which may be used as the column names or row index.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3e64dc08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original column contents</th>\n",
       "      <th>age</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>P7Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3.5</td>\n",
       "      <td>P3Y6M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>P10Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>P19Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  original column contents    age\n",
       "0                        7    P7Y\n",
       "1                      3.5  P3Y6M\n",
       "2                       10   P10Y\n",
       "3                       19   P19Y"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ageMapper = AgeColumnMapper.by_year('Age at examination(years)')\n",
    "ageMapper.preview_column(dft['Age at examination(years)'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "71f664cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original column contents</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>female</td>\n",
       "      <td>FEMALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>male</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>female</td>\n",
       "      <td>FEMALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>female</td>\n",
       "      <td>FEMALE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  original column contents     sex\n",
       "0                   female  FEMALE\n",
       "1                     male    MALE\n",
       "2                   female  FEMALE\n",
       "3                   female  FEMALE"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sexMapper = SexColumnMapper(male_symbol='male', female_symbol='female', column_name='Sex')\n",
    "sexMapper.preview_column(dft['Sex'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f6581a8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pmid = \"PMID:28258187\"\n",
    "encoder = CohortEncoder(df=dft, hpo_cr=hpo_cr, column_mapper_d=column_mapper_d, \n",
    "                        individual_column_name=\"patient_id\", agemapper=ageMapper, sexmapper=sexMapper,\n",
    "                       variant_mapper=varMapper, metadata=metadata,\n",
    "                       pmid=pmid)\n",
    "encoder.set_disease(disease_id='618343', label='Polymicrogyria with or without vascular-type EDS')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "fd367ed6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000090.3%3Ac.145C>G/NM_000090.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000090.3%3Ac.145C>G/NM_000090.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000090.3%3Ac.479dupT/NM_000090.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000090.3%3Ac.1786C>T/NM_000090.3?content-type=application%2Fjson\n"
     ]
    }
   ],
   "source": [
    "individuals = encoder.get_individuals()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5d044b78",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"id\": \"Patient 1 (11.3 this manuscript)\",\n",
      "  \"subject\": {\n",
      "    \"id\": \"Patient 1 (11.3 this manuscript)\",\n",
      "    \"timeAtLastEncounter\": {\n",
      "      \"age\": {\n",
      "        \"iso8601duration\": \"P7Y\"\n",
      "      }\n",
      "    },\n",
      "    \"sex\": \"FEMALE\"\n",
      "  },\n",
      "  \"phenotypicFeatures\": [\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0031422\",\n",
      "        \"label\": \"Abnormal cerebellar cortex morphology\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0001263\",\n",
      "        \"label\": \"Global developmental delay\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0032400\",\n",
      "        \"label\": \"Dysgyria with thickened cortex\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0007103\",\n",
      "        \"label\": \"Hypointensity of cerebral white matter on MRI\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0012520\",\n",
      "        \"label\": \"Dilation of Virchow-Robin spaces\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0006956\",\n",
      "        \"label\": \"Lateral ventricle dilatation\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0007082\",\n",
      "        \"label\": \"Dilated third ventricle\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0002363\",\n",
      "        \"label\": \"Abnormal brainstem morphology\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0012520\",\n",
      "        \"label\": \"Dilation of Virchow-Robin spaces\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0200011\",\n",
      "        \"label\": \"Abnormal length of corpus callosum\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0200011\",\n",
      "        \"label\": \"Abnormal length of corpus callosum\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0001320\",\n",
      "        \"label\": \"Cerebellar vermis hypoplasia\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0002280\",\n",
      "        \"label\": \"Enlarged cisterna magna\"\n",
      "      }\n",
      "    }\n",
      "  ],\n",
      "  \"interpretations\": [\n",
      "    {\n",
      "      \"id\": \"Patient 1 (11.3 this manuscript)\",\n",
      "      \"progressStatus\": \"SOLVED\",\n",
      "      \"diagnosis\": {\n",
      "        \"disease\": {\n",
      "          \"id\": \"618343\",\n",
      "          \"label\": \"Polymicrogyria with or without vascular-type EDS\"\n",
      "        },\n",
      "        \"genomicInterpretations\": [\n",
      "          {\n",
      "            \"subjectOrBiosampleId\": \"Patient 1 (11.3 this manuscript)\",\n",
      "            \"interpretationStatus\": \"CAUSATIVE\",\n",
      "            \"variantInterpretation\": {\n",
      "              \"variationDescriptor\": {\n",
      "                \"id\": \"var_sjxIHAvuAnFrnRQHPMzCuyrOi\",\n",
      "                \"geneContext\": {\n",
      "                  \"valueId\": \"HGNC:2201\",\n",
      "                  \"symbol\": \"COL3A1\"\n",
      "                },\n",
      "                \"expressions\": [\n",
      "                  {\n",
      "                    \"syntax\": \"hgvs.c\",\n",
      "                    \"value\": \"NM_000090.3:c.145C>G\"\n",
      "                  },\n",
      "                  {\n",
      "                    \"syntax\": \"hgvs.g\",\n",
      "                    \"value\": \"NC_000002.12:g.188984825C>G\"\n",
      "                  }\n",
      "                ],\n",
      "                \"vcfRecord\": {\n",
      "                  \"genomeAssembly\": \"hg38\",\n",
      "                  \"chrom\": \"chr2\",\n",
      "                  \"pos\": \"188984825\",\n",
      "                  \"ref\": \"C\",\n",
      "                  \"alt\": \"G\"\n",
      "                },\n",
      "                \"moleculeContext\": \"genomic\",\n",
      "                \"allelicState\": {\n",
      "                  \"id\": \"GENO:0000136\",\n",
      "                  \"label\": \"homozygous\"\n",
      "                }\n",
      "              }\n",
      "            }\n",
      "          }\n",
      "        ]\n",
      "      }\n",
      "    }\n",
      "  ],\n",
      "  \"metaData\": {\n",
      "    \"created\": \"2023-07-02T12:55:32.092464208Z\",\n",
      "    \"createdBy\": \"ORCID:0000-0002-5648-2155\",\n",
      "    \"resources\": [\n",
      "      {\n",
      "        \"id\": \"geno\",\n",
      "        \"name\": \"Genotype Ontology\",\n",
      "        \"url\": \"http://purl.obolibrary.org/obo/geno.owl\",\n",
      "        \"version\": \"2022-03-05\",\n",
      "        \"namespacePrefix\": \"GENO\",\n",
      "        \"iriPrefix\": \"http://purl.obolibrary.org/obo/GENO_\"\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"hgnc\",\n",
      "        \"name\": \"HUGO Gene Nomenclature Committee\",\n",
      "        \"url\": \"https://www.genenames.org\",\n",
      "        \"version\": \"06/01/23\",\n",
      "        \"namespacePrefix\": \"HGNC\",\n",
      "        \"iriPrefix\": \"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/\"\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"omim\",\n",
      "        \"name\": \"An Online Catalog of Human Genes and Genetic Disorders\",\n",
      "        \"url\": \"https://www.omim.org\",\n",
      "        \"version\": \"January 4, 2023\",\n",
      "        \"namespacePrefix\": \"OMIM\",\n",
      "        \"iriPrefix\": \"https://www.omim.org/entry/\"\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"so\",\n",
      "        \"name\": \"Sequence types and features ontology\",\n",
      "        \"url\": \"http://purl.obolibrary.org/obo/so.obo\",\n",
      "        \"version\": \"2021-11-22\",\n",
      "        \"namespacePrefix\": \"SO\",\n",
      "        \"iriPrefix\": \"http://purl.obolibrary.org/obo/SO_\"\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"hp\",\n",
      "        \"name\": \"human phenotype ontology\",\n",
      "        \"url\": \"http://purl.obolibrary.org/obo/hp.owl\",\n",
      "        \"version\": \"2023-06-17\",\n",
      "        \"namespacePrefix\": \"HP\",\n",
      "        \"iriPrefix\": \"http://purl.obolibrary.org/obo/HP_\"\n",
      "      }\n",
      "    ],\n",
      "    \"phenopacketSchemaVersion\": \"2.0\"\n",
      "  }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "i1 = individuals[0]\n",
    "phenopacket1 = i1.to_ga4gh_phenopacket(metadata=metadata.to_ga4gh())\n",
    "json_string = MessageToJson(phenopacket1)\n",
    "print(json_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "23f1094f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000090.3%3Ac.145C>G/NM_000090.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000090.3%3Ac.145C>G/NM_000090.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000090.3%3Ac.479dupT/NM_000090.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000090.3%3Ac.1786C>T/NM_000090.3?content-type=application%2Fjson\n",
      "Wrote 4 phenopackets to ../../phenopackets/COL3A1/\n"
     ]
    }
   ],
   "source": [
    "output_directory = \"../../phenopackets/COL3A1/\"\n",
    "encoder.output_phenopackets(outdir=output_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cebb02c6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
