{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a39f3c65",
   "metadata": {},
   "source": [
    "<h1>Creation of phenopackets from tabular data (individuals in columns)</h1>\n",
    "<p>We will process <a href=\"https://pubmed.ncbi.nlm.nih.gov/15007000/\" target=\"__blank\">Boutouyrie, et al. (2004) Increased carotid wall stress in vascular Ehlers-Danlos syndrome</a> (PMID:15007000).</p>\n",
    "<p>Note that this publication does not report variants.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f0602c0e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pyphetools version 0.6.3\n"
     ]
    }
   ],
   "source": [
    "import phenopackets as php\n",
    "from google.protobuf.json_format import MessageToDict, MessageToJson\n",
    "from google.protobuf.json_format import Parse, ParseDict\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_colwidth', None) # show entire column contents, important!\n",
    "from collections import defaultdict\n",
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "from pyphetools.creation import *\n",
    "from pyphetools.visualization import *\n",
    "import pyphetools\n",
    "print(f\"pyphetools version {pyphetools.__version__}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8606e7eb",
   "metadata": {},
   "source": [
    "<h2>Importing HPO data</h2>\n",
    "<p>pyphetools uses the Human Phenotype Ontology (HPO) to encode phenotypic features. The recommended way of doing this is to ingest the hp.json file using HpoParser, which in turn creates an HpoConceptRecognizer object. </p>\n",
    "<p>The HpoParser can accept a hpo_json_file argument if you want to use a specific file. If the argument is not passed, it will download the latext hp.json file from the HPO GitHub site and store it in a new subdirectory called hpo_data. It will not download the file if the file is already downloaded.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5a7789fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = HpoParser()\n",
    "hpo_cr = parser.get_hpo_concept_recognizer()\n",
    "hpo_version = parser.get_version()\n",
    "PMID = \"PMID:15007000\"\n",
    "title = \"Increased carotid wall stress in vascular Ehlers-Danlos syndrome\"\n",
    "metadata = MetaData(created_by=\"ORCID:0000-0002-5648-2155\", pmid=PMID, pubmed_title=title)\n",
    "metadata.default_versions_with_hpo(version=hpo_version)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5102722a",
   "metadata": {},
   "source": [
    "<h2>Importing the supplemental table</h2>\n",
    "<p>Here, we use the pandas library to import this file (note that the Python package called openpyxl must be installed to read Excel files with pandas, although the library does not need to be imported in this notebook). pyphetools expects a pandas DataFrame as input, and users can choose any input format available for pandas include CSV, TSV, and Excel, or can use any other method to transform their input data into a Pandas DataFrame before using pyphetools.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f17c3924",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel('input/PMID_15007000.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e013666d",
   "metadata": {},
   "source": [
    "Some column names might include spaces in front or after, and a couple of columns are subheadings and only contain NaNs, so lets correct that:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5d604600",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patients</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Age at Echo Examination, y</th>\n",
       "      <th>Age at First Complication, y</th>\n",
       "      <th>Arterial Dissection or Rupture</th>\n",
       "      <th>Organ Rupture</th>\n",
       "      <th>Family History of the Disease</th>\n",
       "      <th>Acrogeria</th>\n",
       "      <th>Excessive Bruising/ Hematoma</th>\n",
       "      <th>Thin Translucent Skin</th>\n",
       "      <th>patient_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C.G.</td>\n",
       "      <td>F</td>\n",
       "      <td>36</td>\n",
       "      <td>28.0</td>\n",
       "      <td>Carotid-cavernous fistula, iliac artery dissection</td>\n",
       "      <td>Uterus</td>\n",
       "      <td>Yes (son)</td>\n",
       "      <td>++</td>\n",
       "      <td>+</td>\n",
       "      <td>++</td>\n",
       "      <td>CG_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I.A.</td>\n",
       "      <td>F</td>\n",
       "      <td>31</td>\n",
       "      <td>30.0</td>\n",
       "      <td>AV fistula tibial artery, stroke</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes (numerous)</td>\n",
       "      <td>+++</td>\n",
       "      <td>+++</td>\n",
       "      <td>+</td>\n",
       "      <td>IA_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A.G.</td>\n",
       "      <td>M</td>\n",
       "      <td>26</td>\n",
       "      <td>24.0</td>\n",
       "      <td>Left vertebrobasilar dissection</td>\n",
       "      <td>Colonic perforation, colectomy</td>\n",
       "      <td>No</td>\n",
       "      <td>+++</td>\n",
       "      <td>+++</td>\n",
       "      <td>+</td>\n",
       "      <td>AG_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>J.S.</td>\n",
       "      <td>M</td>\n",
       "      <td>29</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Total colectomy</td>\n",
       "      <td>Yes (mother)</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>JS_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>V.C.</td>\n",
       "      <td>F</td>\n",
       "      <td>22</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2 Colonic perforations</td>\n",
       "      <td>No</td>\n",
       "      <td>+++</td>\n",
       "      <td>++</td>\n",
       "      <td>++</td>\n",
       "      <td>VC_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>V. P.</td>\n",
       "      <td>F</td>\n",
       "      <td>31</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Uterus</td>\n",
       "      <td>No</td>\n",
       "      <td>+++</td>\n",
       "      <td>++</td>\n",
       "      <td>+++</td>\n",
       "      <td>VP_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>L.B.</td>\n",
       "      <td>M</td>\n",
       "      <td>14</td>\n",
       "      <td>13.0</td>\n",
       "      <td>Voluminous hematoma</td>\n",
       "      <td>0</td>\n",
       "      <td>No</td>\n",
       "      <td>+++</td>\n",
       "      <td>+++</td>\n",
       "      <td>++</td>\n",
       "      <td>LB_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>E.Z.</td>\n",
       "      <td>F</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>No</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>EZ_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>R.S.</td>\n",
       "      <td>F</td>\n",
       "      <td>51</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Splenic artery dissection</td>\n",
       "      <td>0</td>\n",
       "      <td>No</td>\n",
       "      <td>+++</td>\n",
       "      <td>+++</td>\n",
       "      <td>+</td>\n",
       "      <td>RS_9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>J.C.</td>\n",
       "      <td>F</td>\n",
       "      <td>31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Varicose</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes (father, aunt)</td>\n",
       "      <td>+</td>\n",
       "      <td>++</td>\n",
       "      <td>++</td>\n",
       "      <td>JC_10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>P.C.</td>\n",
       "      <td>F</td>\n",
       "      <td>46</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Iliac artery dissection</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes (brother)</td>\n",
       "      <td>+</td>\n",
       "      <td>++</td>\n",
       "      <td>++</td>\n",
       "      <td>PC_11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A.M.</td>\n",
       "      <td>F</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes (mother)</td>\n",
       "      <td>++</td>\n",
       "      <td>+</td>\n",
       "      <td>++</td>\n",
       "      <td>AM_12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A.M.</td>\n",
       "      <td>F</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes (mother)</td>\n",
       "      <td>++</td>\n",
       "      <td>+</td>\n",
       "      <td>++</td>\n",
       "      <td>AM_13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>V.L.</td>\n",
       "      <td>F</td>\n",
       "      <td>26</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes (mother, sister)</td>\n",
       "      <td>+</td>\n",
       "      <td>++</td>\n",
       "      <td>++</td>\n",
       "      <td>VL_14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>C.L.</td>\n",
       "      <td>F</td>\n",
       "      <td>32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes (mother, sister)</td>\n",
       "      <td>+++</td>\n",
       "      <td>+</td>\n",
       "      <td>+++</td>\n",
       "      <td>CL_15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>M.D.</td>\n",
       "      <td>F</td>\n",
       "      <td>45</td>\n",
       "      <td>37.0</td>\n",
       "      <td>Mesenteric artery aneurysm, renal artery dissection, iliac artery dissection</td>\n",
       "      <td>Colectomy</td>\n",
       "      <td>Yes (mother, uncle)</td>\n",
       "      <td>+++</td>\n",
       "      <td>+++</td>\n",
       "      <td>+++</td>\n",
       "      <td>MD_16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Patients Gender  ...  Thin Translucent Skin  patient_id\n",
       "0      C.G.      F  ...                     ++        CG_1\n",
       "1      I.A.      F  ...                      +        IA_2\n",
       "2      A.G.      M  ...                      +        AG_3\n",
       "3      J.S.      M  ...                      +        JS_4\n",
       "4      V.C.      F  ...                     ++        VC_5\n",
       "5     V. P.      F  ...                    +++        VP_6\n",
       "6      L.B.      M  ...                     ++        LB_7\n",
       "7      E.Z.      F  ...                      +        EZ_8\n",
       "8      R.S.      F  ...                      +        RS_9\n",
       "9      J.C.      F  ...                     ++       JC_10\n",
       "10     P.C.      F  ...                     ++       PC_11\n",
       "11     A.M.      F  ...                     ++       AM_12\n",
       "12     A.M.      F  ...                     ++       AM_13\n",
       "13     V.L.      F  ...                     ++       VL_14\n",
       "14     C.L.      F  ...                    +++       CL_15\n",
       "15     M.D.      F  ...                    +++       MD_16\n",
       "\n",
       "[16 rows x 11 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "# We add an index to the id because two patients have the same initials\n",
    "indx = 0\n",
    "def clean_id(patient_id):\n",
    "    pat_id = re.sub(r\"[\\s\\.]\", \"\", str(patient_id))\n",
    "    global indx\n",
    "    indx += 1\n",
    "    return f\"{pat_id}_{indx}\"\n",
    "df.columns = df.columns.str.strip()\n",
    "df = df.dropna(axis=1, how='all')\n",
    "df['patient_id'] = df['Patients'].transform(lambda x: clean_id(x) )\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8314d6bd",
   "metadata": {},
   "source": [
    "<h2>Column mappers</h2>\n",
    "<p>Please see the <a href=\"https://monarch-initiative.github.io/pyphetools/\" target=\"__blank\">phephetools documentation</a> on the software. In the following cell we create a dictionary for the ColumnMappers. Note that the code is identical except that we use the df.loc function to get the corresponding row data. For this notebook, we will not model severity, and so we first make the table easier to parse by replace all +++ and ++ by a single +.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2305a0d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Patients', 'Gender', 'Age at Echo Examination, y', 'Age at First Complication, y', 'Arterial Dissection or Rupture', 'Organ Rupture', 'Family History of the Disease', 'Acrogeria', 'Excessive Bruising/ Hematoma', 'Thin Translucent Skin', 'patient_id']\n"
     ]
    }
   ],
   "source": [
    "hpo_cr = parser.get_hpo_concept_recognizer()\n",
    "df = df.replace('+++', '+').replace('++', '+').replace(0, '-')\n",
    "generator = SimpleColumnMapperGenerator(df=df, observed='+', excluded='-', hpo_cr=hpo_cr)\n",
    "column_mapper_d = generator.try_mapping_columns()\n",
    "print(generator.get_unmapped_columns())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "12a8ebe7",
   "metadata": {},
   "outputs": [],
   "source": [
    "AcrogeriaMapper = SimpleColumnMapper(hpo_id='HP:0000978',\n",
    "    hpo_label='Bruising susceptibility',\n",
    "    observed='+',\n",
    "    excluded='-')\n",
    "AcrogeriaMapper.preview_column(df['Excessive Bruising/ Hematoma'])\n",
    "column_mapper_d['Excessive Bruising/ Hematoma'] = AcrogeriaMapper\n",
    "\n",
    "thinskinMapper = SimpleColumnMapper(hpo_id='HP:0010648',\n",
    "    hpo_label='Dermal translucency',\n",
    "    observed='+',\n",
    "    excluded='-')\n",
    "thinskinMapper.preview_column(df['Thin Translucent Skin'])\n",
    "column_mapper_d['Thin Translucent Skin'] = thinskinMapper\n",
    "\n",
    "excessive_bruisingMapper = SimpleColumnMapper(hpo_id='HP:0000978',\n",
    "    hpo_label='Bruising susceptibility',\n",
    "    observed='+',\n",
    "    excluded='-')\n",
    "excessive_bruisingMapper.preview_column(df['Excessive Bruising/ Hematoma'])\n",
    "column_mapper_d['Excessive Bruising/ Hematoma'] = excessive_bruisingMapper"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ecb20c7",
   "metadata": {},
   "source": [
    "<h2>Autoformating code for the OptionColumnMapper</h2>\n",
    "<p>Lets try to get code autoformatted so that we can easily copy-paste and change it. Some of the columns can be created in this way,</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "cc006113",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "age_at_echo_examination_y_d = {'36': 'PLACEHOLDER',\n",
      " '31': 'PLACEHOLDER',\n",
      " '26': 'PLACEHOLDER',\n",
      " '29': 'PLACEHOLDER',\n",
      " '22': 'PLACEHOLDER',\n",
      " '14': 'PLACEHOLDER',\n",
      " '19': 'PLACEHOLDER',\n",
      " '51': 'PLACEHOLDER',\n",
      " '46': 'PLACEHOLDER',\n",
      " '17': 'PLACEHOLDER',\n",
      " '32': 'PLACEHOLDER',\n",
      " '45': 'PLACEHOLDER'}\n",
      "age_at_echo_examination_yMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=age_at_echo_examination_y_d)\n",
      "age_at_echo_examination_yMapper.preview_column(df['Age at Echo Examination, y'])\n",
      "column_mapper_d['Age at Echo Examination, y'] = age_at_echo_examination_yMapper\n",
      "\n",
      "age_at_first_complication_y_d = {'28.0': 'PLACEHOLDER',\n",
      " '30.0': 'PLACEHOLDER',\n",
      " '24.0': 'PLACEHOLDER',\n",
      " '13.0': 'PLACEHOLDER',\n",
      " '19.0': 'PLACEHOLDER',\n",
      " 'nan': 'PLACEHOLDER',\n",
      " '50.0': 'PLACEHOLDER',\n",
      " '37.0': 'PLACEHOLDER'}\n",
      "age_at_first_complication_yMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=age_at_first_complication_y_d)\n",
      "age_at_first_complication_yMapper.preview_column(df['Age at First Complication, y'])\n",
      "column_mapper_d['Age at First Complication, y'] = age_at_first_complication_yMapper\n",
      "\n",
      "arterial_dissection_or_rupture_d = {'Carotid-cavernous fistula': 'PLACEHOLDER',\n",
      " 'iliac artery dissection': 'PLACEHOLDER',\n",
      " 'AV fistula tibial artery': 'PLACEHOLDER',\n",
      " 'stroke': 'Stroke',\n",
      " 'Left vertebrobasilar dissection': 'PLACEHOLDER',\n",
      " 'Voluminous hematoma': 'PLACEHOLDER',\n",
      " 'Splenic artery dissection': 'PLACEHOLDER',\n",
      " 'Varicose': 'PLACEHOLDER',\n",
      " 'Iliac artery dissection': 'PLACEHOLDER',\n",
      " 'Mesenteric artery aneurysm': 'Dilatation of mesenteric artery',\n",
      " 'renal artery dissection': 'PLACEHOLDER'}\n",
      "arterial_dissection_or_ruptureMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=arterial_dissection_or_rupture_d)\n",
      "arterial_dissection_or_ruptureMapper.preview_column(df['Arterial Dissection or Rupture'])\n",
      "column_mapper_d['Arterial Dissection or Rupture'] = arterial_dissection_or_ruptureMapper\n",
      "\n",
      "organ_rupture_d = {'Uterus': 'PLACEHOLDER',\n",
      " 'Colonic perforation': 'PLACEHOLDER',\n",
      " 'colectomy': 'PLACEHOLDER',\n",
      " 'Total colectomy': 'PLACEHOLDER',\n",
      " '2 Colonic perforations': 'PLACEHOLDER',\n",
      " 'Colectomy': 'PLACEHOLDER'}\n",
      "organ_ruptureMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=organ_rupture_d)\n",
      "organ_ruptureMapper.preview_column(df['Organ Rupture'])\n",
      "column_mapper_d['Organ Rupture'] = organ_ruptureMapper\n",
      "\n",
      "acrogeria_d = {}\n",
      "acrogeriaMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=acrogeria_d)\n",
      "acrogeriaMapper.preview_column(df['Acrogeria'])\n",
      "column_mapper_d['Acrogeria'] = acrogeriaMapper\n",
      "\n",
      "excessive_bruising/_hematoma_d = {}\n",
      "excessive_bruising/_hematomaMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=excessive_bruising/_hematoma_d)\n",
      "excessive_bruising/_hematomaMapper.preview_column(df['Excessive Bruising/ Hematoma'])\n",
      "column_mapper_d['Excessive Bruising/ Hematoma'] = excessive_bruising/_hematomaMapper\n",
      "\n",
      "thin_translucent_skin_d = {}\n",
      "thin_translucent_skinMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=thin_translucent_skin_d)\n",
      "thin_translucent_skinMapper.preview_column(df['Thin Translucent Skin'])\n",
      "column_mapper_d['Thin Translucent Skin'] = thin_translucent_skinMapper\n",
      "\n"
     ]
    }
   ],
   "source": [
    "output = OptionColumnMapper.autoformat(df=df, concept_recognizer=hpo_cr)\n",
    "print(output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "88e52430",
   "metadata": {},
   "outputs": [],
   "source": [
    "arterial_dissection_or_rupture = {'Carotid-cavernous fistula': 'Carotid cavernous fistula',\n",
    " 'iliac artery dissection': 'Arterial dissection',\n",
    " 'AV fistula tibial artery': 'Arteriovenous fistula',\n",
    " 'stroke': 'Stroke',\n",
    " 'Left vertebrobasilar dissection': 'Arterial dissection',\n",
    " 'Voluminous hematoma': 'Spontaneous hematomas',\n",
    " 'Splenic artery dissection': 'Arterial dissection',\n",
    " 'Varicose': 'Varicose veins',\n",
    " 'Iliac artery dissection': 'Arterial dissection',\n",
    " 'Mesenteric artery aneurysm': 'Vascular dilatation',\n",
    " 'renal artery dissection': 'Arterial dissection'}\n",
    "arterial_dissection_or_ruptureMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=arterial_dissection_or_rupture)\n",
    "#print(arterial_dissection_or_ruptureMapper.preview_column(df['Arterial Dissection or Rupture']))\n",
    "column_mapper_d['Arterial Dissection or Rupture'] = arterial_dissection_or_ruptureMapper\n",
    "\n",
    "organ_rupture = {'Uterus': 'Uterine rupture',\n",
    " 'Colonic perforation': 'Colon perforation',\n",
    " 'colectomy': 'Colon perforation',\n",
    " 'Total colectomy': 'Colon perforation',\n",
    " '2 Colonic perforations': 'Colon perforation',\n",
    " 'Colectomy': 'Colon perforation'}\n",
    "organ_ruptureMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=organ_rupture)\n",
    "#print(organ_ruptureMapper.preview_column(df['Organ Rupture']))\n",
    "column_mapper_d['Organ Rupture'] = organ_ruptureMapper"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1af2052",
   "metadata": {},
   "source": [
    "<h1>Demographic data</h1>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3e64dc08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original column contents</th>\n",
       "      <th>age</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36</td>\n",
       "      <td>P36Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>31</td>\n",
       "      <td>P31Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   original column contents   age\n",
       "0                        36  P36Y\n",
       "1                        31  P31Y"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ageMapper = AgeColumnMapper.by_year('Age at Echo Examination, y')\n",
    "ageMapper.preview_column(df['Age at Echo Examination, y']).head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "71f664cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original column contents</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>F</td>\n",
       "      <td>FEMALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>F</td>\n",
       "      <td>FEMALE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  original column contents     sex\n",
       "0                        F  FEMALE\n",
       "1                        F  FEMALE"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sexMapper = SexColumnMapper(male_symbol='M', female_symbol='F', column_name='Gender')\n",
    "sexMapper.preview_column(df['Gender']).head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f6581a8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "encoder = CohortEncoder(df=df, \n",
    "                        hpo_cr=hpo_cr, \n",
    "                        column_mapper_d=column_mapper_d, \n",
    "                        individual_column_name=\"patient_id\", \n",
    "                        agemapper=ageMapper, \n",
    "                        sexmapper=sexMapper,\n",
    "                        metadata=metadata,  \n",
    "                        pmid=PMID)\n",
    "encoder.set_disease(disease_id='OMIM:130050', label='Ehlers-Danlos syndrome, vascular type')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fd367ed6",
   "metadata": {},
   "outputs": [],
   "source": [
    "individuals = encoder.get_individuals()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5d044b78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"id\": \"PMID_15007000_CG_1\",\n",
      "  \"subject\": {\n",
      "    \"id\": \"CG_1\",\n",
      "    \"timeAtLastEncounter\": {\n",
      "      \"age\": {\n",
      "        \"iso8601duration\": \"P36Y\"\n",
      "      }\n",
      "    },\n",
      "    \"sex\": \"FEMALE\"\n",
      "  },\n",
      "  \"phenotypicFeatures\": [\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0005294\",\n",
      "        \"label\": \"Arterial dissection\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0031157\",\n",
      "        \"label\": \"Carotid cavernous fistula\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0100718\",\n",
      "        \"label\": \"Uterine rupture\"\n",
      "      }\n",
      "    }\n",
      "  ],\n",
      "  \"metaData\": {\n",
      "    \"created\": \"2023-10-01T01:27:59.727237224Z\",\n",
      "    \"createdBy\": \"ORCID:0000-0002-5648-2155\",\n",
      "    \"resources\": [\n",
      "      {\n",
      "        \"id\": \"geno\",\n",
      "        \"name\": \"Genotype Ontology\",\n",
      "        \"url\": \"http://purl.obolibrary.org/obo/geno.owl\",\n",
      "        \"version\": \"2022-03-05\",\n",
      "        \"namespacePrefix\": \"GENO\",\n",
      "        \"iriPrefix\": \"http://purl.obolibrary.org/obo/GENO_\"\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"hgnc\",\n",
      "        \"name\": \"HUGO Gene Nomenclature Committee\",\n",
      "        \"url\": \"https://www.genenames.org\",\n",
      "        \"version\": \"06/01/23\",\n",
      "        \"namespacePrefix\": \"HGNC\",\n",
      "        \"iriPrefix\": \"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/\"\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"omim\",\n",
      "        \"name\": \"An Online Catalog of Human Genes and Genetic Disorders\",\n",
      "        \"url\": \"https://www.omim.org\",\n",
      "        \"version\": \"January 4, 2023\",\n",
      "        \"namespacePrefix\": \"OMIM\",\n",
      "        \"iriPrefix\": \"https://www.omim.org/entry/\"\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"so\",\n",
      "        \"name\": \"Sequence types and features ontology\",\n",
      "        \"url\": \"http://purl.obolibrary.org/obo/so.obo\",\n",
      "        \"version\": \"2021-11-22\",\n",
      "        \"namespacePrefix\": \"SO\",\n",
      "        \"iriPrefix\": \"http://purl.obolibrary.org/obo/SO_\"\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"hp\",\n",
      "        \"name\": \"human phenotype ontology\",\n",
      "        \"url\": \"http://purl.obolibrary.org/obo/hp.owl\",\n",
      "        \"version\": \"2023-09-01\",\n",
      "        \"namespacePrefix\": \"HP\",\n",
      "        \"iriPrefix\": \"http://purl.obolibrary.org/obo/HP_\"\n",
      "      }\n",
      "    ],\n",
      "    \"phenopacketSchemaVersion\": \"2.0\",\n",
      "    \"externalReferences\": [\n",
      "      {\n",
      "        \"id\": \"PMID:15007000\",\n",
      "        \"reference\": \"https://pubmed.ncbi.nlm.nih.gov/15007000\",\n",
      "        \"description\": \"Increased carotid wall stress in vascular Ehlers-Danlos syndrome\"\n",
      "      }\n",
      "    ]\n",
      "  }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "i1 = individuals[0]\n",
    "phenopacket1 = i1.to_ga4gh_phenopacket(metadata=metadata.to_ga4gh())\n",
    "json_string = MessageToJson(phenopacket1)\n",
    "print(json_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "23f1094f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 16 phenopackets to ../../phenopackets/COL3A1/\n"
     ]
    }
   ],
   "source": [
    "output_directory = \"../../phenopackets/COL3A1/\"\n",
    "encoder.output_phenopackets(outdir=output_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20c67614",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
