{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "834725fa",
   "metadata": {},
   "source": [
    "<h1>GLI3: Demurger et al 2015</h1>\n",
    "<p>Extract the clinical data from <a href=\"https://pubmed.ncbi.nlm.nih.gov/24736735/\"target=\"__blank\">Démurger F, et al. (2015) New insights into genotype-phenotype correlation for GLI3 mutations. Eur J Hum Genet ;23(1):92-102. PMID:24736735</a>.<p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "da99853a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import phenopackets as php\n",
    "from google.protobuf.json_format import MessageToDict, MessageToJson\n",
    "from google.protobuf.json_format import Parse, ParseDict\n",
    "import pandas as pd\n",
    "import math\n",
    "from csv import DictReader\n",
    "pd.set_option('display.max_colwidth', None) # show entire column contents, important!\n",
    "from collections import defaultdict\n",
    "import re\n",
    "from pyphetools.creation import *\n",
    "from pyphetools.output import PhenopacketTable\n",
    "# last tested with pyphetools version 0.3.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "10151cd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = HpoParser()\n",
    "hpo_cr = parser.get_hpo_concept_recognizer()\n",
    "hpo_version = parser.get_version()\n",
    "metadata = MetaData(created_by=\"ORCID:0000-0002-0736-9199\")\n",
    "metadata.default_versions_with_hpo(version=hpo_version)\n",
    "pmid=\"PMID:29198722\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "3aba8016",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N</th>\n",
       "      <th>cDNA alteration</th>\n",
       "      <th>Predicted protein alteration</th>\n",
       "      <th>Inheritance</th>\n",
       "      <th>Postaxial PD</th>\n",
       "      <th>Preaxial PD</th>\n",
       "      <th>Broad thumbs or halluces</th>\n",
       "      <th>Syndactyly</th>\n",
       "      <th>Macrocephaly</th>\n",
       "      <th>Widely spaced eyes</th>\n",
       "      <th>MRI Findings</th>\n",
       "      <th>Developmental delay</th>\n",
       "      <th>Additional findings</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G029</td>\n",
       "      <td>327del</td>\n",
       "      <td>Phe109Leufs*50</td>\n",
       "      <td>F</td>\n",
       "      <td>–</td>\n",
       "      <td>FB</td>\n",
       "      <td></td>\n",
       "      <td>+</td>\n",
       "      <td>–</td>\n",
       "      <td>–</td>\n",
       "      <td>–</td>\n",
       "      <td>–</td>\n",
       "      <td>Precocious puberty, scaphocephaly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G070</td>\n",
       "      <td>427G&gt;T</td>\n",
       "      <td>Glu143*</td>\n",
       "      <td>F</td>\n",
       "      <td>HB</td>\n",
       "      <td>FL</td>\n",
       "      <td></td>\n",
       "      <td>–</td>\n",
       "      <td>+</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>–</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mother</td>\n",
       "      <td>427G&gt;T</td>\n",
       "      <td>Glu143*</td>\n",
       "      <td>F</td>\n",
       "      <td>–</td>\n",
       "      <td>–</td>\n",
       "      <td></td>\n",
       "      <td>–</td>\n",
       "      <td>–</td>\n",
       "      <td>–</td>\n",
       "      <td></td>\n",
       "      <td>–</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G118</td>\n",
       "      <td>444C&gt;A</td>\n",
       "      <td>Tyr148*</td>\n",
       "      <td>F</td>\n",
       "      <td>–</td>\n",
       "      <td>FB</td>\n",
       "      <td>BT</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>–</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G13684</td>\n",
       "      <td>444C&gt;A</td>\n",
       "      <td>Tyr148*</td>\n",
       "      <td>F</td>\n",
       "      <td>–</td>\n",
       "      <td>FB</td>\n",
       "      <td></td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td></td>\n",
       "      <td>–</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         N cDNA alteration Predicted protein alteration Inheritance   \n",
       "0    G029           327del               Phe109Leufs*50           F  \\\n",
       "1    G070           427G>T                      Glu143*           F   \n",
       "2   Mother          427G>T                      Glu143*           F   \n",
       "3    G118           444C>A                      Tyr148*           F   \n",
       "4  G13684           444C>A                      Tyr148*           F   \n",
       "\n",
       "  Postaxial PD Preaxial PD Broad thumbs or halluces Syndactyly Macrocephaly   \n",
       "0            –          FB                                   +            –  \\\n",
       "1           HB          FL                                   –            +   \n",
       "2            –           –                                   –            –   \n",
       "3            –          FB                       BT          +            +   \n",
       "4            –          FB                                   +            +   \n",
       "\n",
       "  Widely spaced eyes MRI Findings Developmental delay   \n",
       "0                  –            –                   –  \\\n",
       "1                                                   –   \n",
       "2                  –                                –   \n",
       "3                                                   –   \n",
       "4                  +                                –   \n",
       "\n",
       "                 Additional findings  \n",
       "0  Precocious puberty, scaphocephaly  \n",
       "1                                     \n",
       "2                                     \n",
       "3                                     \n",
       "4                                     "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = pd.read_csv(\"input/demurger_table_1.csv\", delimiter=\"\\t\")\n",
    "df1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "c6771fb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "column_mapper_d = defaultdict(ColumnMapper)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "b9879922",
   "metadata": {},
   "outputs": [],
   "source": [
    "postaxial_d = {'HB': 'Postaxial hand polydactyly',\n",
    "              'FB': 'Postaxial foot polydactyly',}\n",
    "postaxialMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=postaxial_d, \n",
    "                                     negative_symbol=\"–\", negative_label='Postaxial polydactyly')\n",
    "postaxialMapper.preview_column(df1[\"Postaxial PD\"])\n",
    "column_mapper_d[\"Postaxial PD\"] = postaxialMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "668aa32c",
   "metadata": {},
   "outputs": [],
   "source": [
    "preaxial_d = {'HB': 'Preaxial hand polydactyly',\n",
    "              'FB': 'Preaxial foot polydactyly',\n",
    "              'FL': 'Preaxial foot polydactyly',}\n",
    "preaxialMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=preaxial_d,\n",
    "                                   negative_symbol=\"–\", negative_label='Preaxial polydactyly')\n",
    "preaxialMapper.preview_column(df1[\"Preaxial PD\"])\n",
    "column_mapper_d[\"Preaxial PD\"] = preaxialMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "88c8fe0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "thumb_d = {\"BT\": \"Broad thumb\", \n",
    "          \"BH\": \"Broad hallux\",\n",
    "          \"+\": [ \"Broad thumb\", \"Broad hallux\"]}\n",
    "thumbMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=thumb_d,\n",
    "                                negative_symbol=\"–\", negative_label=\"Broad thumb\")\n",
    "thumbMapper.preview_column(df1[\"Broad thumbs or halluces\"])\n",
    "column_mapper_d[\"Broad thumbs or halluces\"] = thumbMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "2c581365",
   "metadata": {},
   "outputs": [],
   "source": [
    "syndMapper = SimpleColumnMapper(hpo_id=\"HP:0001159\", hpo_label=\"Syndactyly\", observed=\"+\", excluded=\"–\")\n",
    "syndMapper.preview_column(df1[\"Syndactyly\"])\n",
    "column_mapper_d[\"Syndactyly\"] = syndMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "02d64bd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "macMapper = SimpleColumnMapper(hpo_id=\"HP:0000256\", hpo_label=\"Macrocephaly\", observed=\"+\", excluded=\"–\")\n",
    "macMapper.preview_column(df1[\"Macrocephaly\"])\n",
    "column_mapper_d[\"Macrocephaly\"] = macMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "35321910",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Widely spaced eyes  Hypertelorism HP:0000316\n",
    "htMapper = SimpleColumnMapper(hpo_id=\"HP:0000316\", hpo_label=\"Hypertelorism\", observed=\"+\", excluded=\"–\")\n",
    "htMapper.preview_column(df1[\"Widely spaced eyes\"])\n",
    "column_mapper_d[\"Widely spaced eyes\"] = htMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "458b06e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MRI Findings\n",
    "# CCA, corpus callosum agenesis; CCH, corpus callosum hypoplasia;\n",
    "# pCCA, partial CCA;  VD, ventricular dilatation. #: Index cases are indicated in bold and related are below.\n",
    "mri_d = {'CCH': 'Hypoplasia of the corpus callosum',\n",
    "         'CCA': 'Agenesis of corpus callosum',\n",
    "         'pCCA': 'Partial agenesis of the corpus callosum',\n",
    "         'VD': 'Ventriculomegaly'}\n",
    "mriMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=mri_d)\n",
    "mriMapper.preview_column(df1[\"MRI Findings\"])\n",
    "column_mapper_d[\"MRI Findings\"] = mriMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "5f9bf14e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dd_d = {'+': 'Global developmental delay',\n",
    "         'Mild': 'Mild global developmental delay',\n",
    "         'Bilateral inguinal hernia': 'Inguinal hernia',\n",
    "         'strabismus': 'Strabismus',\n",
    "       \"Cataract\": \"Cataract\",\n",
    "       \"Seizures\":\"Seizure\",\n",
    "       \"horseshoe kidney\": \"Horseshoe kidney\",\n",
    "       \"Trigonocephaly\": \"Trigonocephaly\",\n",
    "       \"macrosomia\": \"Macrosomia\",\n",
    "       \"vermis dysgenesis\": \"Dysgenesis of the cerebellar vermis\"}\n",
    "ddMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=dd_d,\n",
    "                              negative_symbol=\"–\", negative_label=\"Global developmental delay\")\n",
    "ddMapper.preview_column(df1[\"Developmental delay\"])\n",
    "column_mapper_d[\"Developmental delay\"] = ddMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "249b2860",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Precocious puberty, scaphocephaly', '\\xa0', 'Delta phalanx',\n",
       "       'Atrial septal defect', 'Umbilical hernia',\n",
       "       'Bifid distal phalanx, BW= 4150', 'Cerebral prematurity sequelae',\n",
       "       'Delta metacarpal, BW=4880', 'BW=4740',\n",
       "       'Hypoplastic cerebellum, microretrognathism',\n",
       "       'Bilateral keratoconus, umbilical and bilateral inguinal hernia',\n",
       "       'Macrosomia', 'Neurofibromatosis type 1',\n",
       "       'Brachydactyly, delta phalanx', 'Brachydactyly, speech delay',\n",
       "       'Speech delay, exomphalos', 'Umbilical hernia, anterior anus',\n",
       "       'Laryngomalacia', 'BW=4440', 'Supernumerary nipples'], dtype=object)"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1[\"Additional findings\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "4515ca18",
   "metadata": {},
   "outputs": [],
   "source": [
    "add_d = {'Precocious puberty': 'Precocious puberty',\n",
    "         'scaphocephaly': 'Scaphocephaly',\n",
    "         'Delta phalanx': 'Triangular shaped phalanges of the hand',\n",
    "         'Bifid distal phalanx': 'Partial duplication of the distal phalanges of the hand',\n",
    "       \"Hypoplastic cerebellum\": \"Cerebellar hypoplasia\",\n",
    "       \"microretrognathism\":\"Microretrognathia\",\n",
    "       \"keratoconus\": \"Keratoconus\",\n",
    "       \"umbilical\": \"Umbilical hernia\",\n",
    "         \"Umbilical hernia\": \"Umbilical hernia\",\n",
    "       \"inguinal hernia\": \"Inguinal hernia\",\n",
    "       \"Macrosomia\": \"Large for gestational age\",\n",
    "        \"Brachydactyly\":\"Brachydactyly\",\n",
    "         \"anterior anus\": \"Anteriorly placed anus\",\n",
    "         \"Laryngomalacia\":\"Laryngomalacia\",\n",
    "         \"Supernumerary nipples\": \"Supernumerary nipple\"\n",
    "        }\n",
    "addMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=add_d)\n",
    "addMapper.preview_column(df1[\"Additional findings\"])\n",
    "column_mapper_d[\"Additional findings\"] = addMapper"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7b178fa",
   "metadata": {},
   "source": [
    "<h2>GLI3 Variants</h2>\n",
    "<p>Variants are provided in table 1 according to NM_000168.6.</p>\n",
    "<p>Note that the contents of the column \"cDNA alteration\" do not have the \"c.\" required by HGVS, so we add it to all columns before proceding.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "81977833",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1['cDNA alteration'] = 'c.' + df1['cDNA alteration'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "ee1f596b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.327del/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.327del: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.427G>T/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.427G>T: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.427G>T/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.427G>T: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.444C>A/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.444C>A: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.444C>A/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.444C>A: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.444C>A/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.444C>A: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.444C>A/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.444C>A: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.518dup/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.518dup: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.679+1G>T/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.679+1G>T: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.833_843del/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.833_843del: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.868C>T/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.868C>T: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.997_998dup/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.997_998dup: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.1063_1067dup/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.1063_1067dup: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.1378del/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.1378del: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.1498C>T/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.1498C>T: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.1498C>T/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.1498C>T: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.1498C>T/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.1498C>T: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.1543_1544dup/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.1543_1544dup: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.1543_1544dup/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.1543_1544dup: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.[=/1543_1544dup]/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.[=/1543_1544dup]: string indices must be integers\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.1733G>C/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.1733G>C: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.1745del/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.1745del: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.1767del/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.1767del: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.1767del/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.1767del: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.1786C>T/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.1786C>T: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.1786C>T/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.1786C>T: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.1786C>T/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.1786C>T: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.1787A>C/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.1787A>C: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.1787A>C/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.1787A>C: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.1874G>A/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.1874G>A: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.1874G>A/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.1874G>A: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.2082_2083delinsAGAGAAGCC/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.2082_2083delinsAGAGAAGCC: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.3427_3443del/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.3427_3443del: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.3559C>T/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.3559C>T: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.3559C>T/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.3559C>T: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.3559C>T/NM_000168.6?content-type=application%2Fjson\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Not able to get variant for c.3559C>T: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.3950del/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.3950del: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.3950del/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.3950del: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.4099dup/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.4099dup: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.4324C>T/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.4324C>T: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.4408C>T/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.4408C>T: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.4431dup/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.4431dup: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.4456C>T/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.4456C>T: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.4463del/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.4463del: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.4463del/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.4463del: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.4463del/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.4463del: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.4463del/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.4463del: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.4615_4624del/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.4615_4624del: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.4615_4624del/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.4615_4624del: 'Variant' object has no attribute 'set_genotype'\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.chr7.hg19:g.(35 674 000_37 280 000)_(46 116 000_46 598 000)del/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.chr7.hg19:g.(35 674 000_37 280 000)_(46 116 000_46 598 000)del: Expecting to get a gene_variant from Variant Validator but got warning\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.chr7.hg19:g(38521704_45810267)del/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.chr7.hg19:g(38521704_45810267)del: Expecting to get a gene_variant from Variant Validator but got warning\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.rsa7p14.1(kit P179)x1/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.rsa7p14.1(kit P179)x1: Expecting to get a gene_variant from Variant Validator but got warning\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.46,XY.ish del(7)(p14.1)(RP11-816F16-)/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.46,XY.ish del(7)(p14.1)(RP11-816F16-): Expecting to get a gene_variant from Variant Validator but got warning\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.46,XX.ish del(7)(p14.1p14.1)(GLI3-)/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.46,XX.ish del(7)(p14.1p14.1)(GLI3-): Expecting to get a gene_variant from Variant Validator but got warning\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_000168.6%3Ac.46,XY.ish del(7)(p14.1)(GLI3-)[56]/7p14.1(GLI3x2)[44]/NM_000168.6?content-type=application%2Fjson\n",
      "Not able to get variant for c.46,XY.ish del(7)(p14.1)(GLI3-)[56]/7p14.1(GLI3x2)[44]: string indices must be integers\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   variant\n",
       "0      n/a\n",
       "1      n/a\n",
       "2      n/a\n",
       "3      n/a\n",
       "4      n/a\n",
       "5      n/a\n",
       "6      n/a\n",
       "7      n/a\n",
       "8      n/a\n",
       "9      n/a\n",
       "10     n/a\n",
       "11     n/a\n",
       "12     n/a\n",
       "13     n/a\n",
       "14     n/a\n",
       "15     n/a\n",
       "16     n/a\n",
       "17     n/a\n",
       "18     n/a\n",
       "19     n/a\n",
       "20     n/a\n",
       "21     n/a\n",
       "22     n/a\n",
       "23     n/a\n",
       "24     n/a\n",
       "25     n/a\n",
       "26     n/a\n",
       "27     n/a\n",
       "28     n/a\n",
       "29     n/a\n",
       "30     n/a\n",
       "31     n/a\n",
       "32     n/a\n",
       "33     n/a\n",
       "34     n/a\n",
       "35     n/a\n",
       "36     n/a\n",
       "37     n/a\n",
       "38     n/a\n",
       "39     n/a\n",
       "40     n/a\n",
       "41     n/a\n",
       "42     n/a\n",
       "43     n/a\n",
       "44     n/a\n",
       "45     n/a\n",
       "46     n/a\n",
       "47     n/a\n",
       "48     n/a\n",
       "49     n/a\n",
       "50     n/a\n",
       "51     n/a\n",
       "52     n/a\n",
       "53     n/a\n",
       "54     n/a"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript='NM_000168.6'\n",
    "genome = 'hg38'\n",
    "varMapper = VariantColumnMapper(assembly=genome,\n",
    "                                column_name='cDNA alteration', \n",
    "                                transcript=transcript, \n",
    "                                default_genotype='heterozygous')\n",
    "varMapper.preview_column(df1['cDNA alteration'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e75984f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "c2c2venv",
   "language": "python",
   "name": "c2c2venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
