{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a39f3c65",
   "metadata": {},
   "source": [
    "<h1>Creation of phenopackets from tabular data (individuals in columns)</h1>\n",
    "<p>We will process <a href=\"https://pubmed.ncbi.nlm.nih.gov/29907796/\" target=\"__blank\">Diets, et al. (2019) SMARCB1 causes severe intellectual disability and choroid plexus hyperplasia with resultant hydrocephalus</a></p>\n",
    "<p>pyphetools provides a convenient way of extracting HPO terms from typical tables presented in supplemental material. Typical tables can have the individuals in columns or rows. In this case, we extract data from TABLE. Clinical Phenotype of Individuals With MAPK8IP3 Variants, in which data from five individuals are presented in columns, with the rows representing the category of clinical data.</p>\n",
    "<p>This note shows how to work through the table and set up the pyphetools encoder. The table is available only as a PDF table in the original publication. We copied the information into an Excel file for this workbook.</p>\n",
    "<p>Users can work on one column at a time and then generate a collection of <a href=\"https://pubmed.ncbi.nlm.nih.gov/35705716/\" target=\"__blank\">GA4GH phenopackets</a> to represent each patient included in the original supplemental material. These phenopackets can then be used for a variety of downstream applications.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f0602c0e",
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [],
   "source": [
    "import phenopackets as php\n",
    "from google.protobuf.json_format import MessageToDict, MessageToJson\n",
    "from google.protobuf.json_format import Parse, ParseDict\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_colwidth', None) # show entire column contents, important!\n",
    "from collections import defaultdict\n",
    "import os\n",
    "import sys\n",
    "\n",
    "sys.path.insert(0, os.path.abspath('../../pyphetools'))\n",
    "from pyphetools import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8606e7eb",
   "metadata": {},
   "source": [
    "<h2>Importing HPO data</h2>\n",
    "<p>pyphetools uses the Human Phenotype Ontology (HPO) to encode phenotypic features. The recommended way of doing this is to ingest the hp.json file using HpoParser, which in turn creates an HpoConceptRecognizer object. </p>\n",
    "<p>The HpoParser can accept a hpo_json_file argument if you want to use a specific file. If the argument is not passed, it will download the latext hp.json file from the HPO GitHub site and store it in a new subdirectory called hpo_data. It will not download the file if the file is already downloaded.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5a7789fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = HpoParser()\n",
    "hpo_cr = parser.get_hpo_concept_recognizer()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5102722a",
   "metadata": {},
   "source": [
    "<h2>Importing the supplemental table</h2>\n",
    "<p>The Table of the Iwasawa et al (2019) paper was copied into an Excel file that is included in the data subfolder</p>\n",
    "<p>Here, we use the pandas library to import this file (note that the Python package called openpyxl must be installed to read Excel files with pandas, although the library does not need to be imported in this notebook). pyphetools expects a pandas DataFrame as input, and users can choose any input format available for pandas include CSV, TSV, and Excel, or can use any other method to transform their input data into a Pandas DataFrame before using pyphetools.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f17c3924",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel('data/PMID_29907796.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "06ec7dbd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Patient 1</th>\n",
       "      <th>Patient 2</th>\n",
       "      <th>Patient 3</th>\n",
       "      <th>Patient 4</th>\n",
       "      <th>Count features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Pathogenic variant</td>\n",
       "      <td>c.110G&gt;A</td>\n",
       "      <td>c.110G&gt;A</td>\n",
       "      <td>c.110G&gt;A</td>\n",
       "      <td>c.110G&gt;A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>p.Arg37His</td>\n",
       "      <td>p.Arg37His</td>\n",
       "      <td>p.Arg37His</td>\n",
       "      <td>p.Arg37His</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Inheritance</td>\n",
       "      <td>De novo</td>\n",
       "      <td>De novo</td>\n",
       "      <td>De novo</td>\n",
       "      <td>De novo</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Age at examination</td>\n",
       "      <td>9.5 y</td>\n",
       "      <td>5 y 8 mo</td>\n",
       "      <td>12 y</td>\n",
       "      <td>17 mo</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Development</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Intellectual disability</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>4/4 (100%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Speech delay</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>4/4 (100%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Motor delay</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>4/4 (100%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Congenital anomalies</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Congenital heart defect</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>2/4 (50%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Laryngomalacia</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>1/4 (25%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Kidney anomalies</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>3/4 (75%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Genital anomalies</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>2/4 (50%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Neurological</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Childhood hypotonia</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>4/4 (100%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Hydrocephalus</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+/- (ICP)</td>\n",
       "      <td>+</td>\n",
       "      <td>4/4 (100%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Eye and ear abnormalities</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CVI</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>3/4 (75%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Myopia</td>\n",
       "      <td>+</td>\n",
       "      <td>ND</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>3/3 (100%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Eye movement disorder</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>3/4 (75%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Other eye problems</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>4/4 (100%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Hearing loss</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>2/4 (50%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Musculoskeletal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Brachycephaly</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>2/4 (50%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Joint hypermobility</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>3/4 (75%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Hip dysplasia</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>1/4 (25%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Contractures</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>2/4 (50%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Obstructive sleep apnea</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>4/4 (100%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Precocious puberty</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1/3 (33%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>(History of) anemia</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>3/4 (75%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Thrombocytopenia</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>1/4 (25%)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Unnamed: 0   Patient 1   Patient 2   Patient 3   Patient 4  \\\n",
       "0          Pathogenic variant    c.110G>A    c.110G>A    c.110G>A    c.110G>A   \n",
       "1                         NaN  p.Arg37His  p.Arg37His  p.Arg37His  p.Arg37His   \n",
       "2                 Inheritance     De novo     De novo     De novo     De novo   \n",
       "3          Age at examination       9.5 y    5 y 8 mo        12 y       17 mo   \n",
       "4                 Development         NaN         NaN         NaN         NaN   \n",
       "5     Intellectual disability      Severe      Severe      Severe      Severe   \n",
       "6                Speech delay      Severe      Severe      Severe      Severe   \n",
       "7                 Motor delay      Severe      Severe      Severe      Severe   \n",
       "8        Congenital anomalies         NaN         NaN         NaN         NaN   \n",
       "9     Congenital heart defect           -           +           -           +   \n",
       "10             Laryngomalacia           -           -           -           +   \n",
       "11           Kidney anomalies           -           +           +           +   \n",
       "12          Genital anomalies           +           -           +           -   \n",
       "13               Neurological         NaN         NaN         NaN         NaN   \n",
       "14        Childhood hypotonia           +           +           +           +   \n",
       "15              Hydrocephalus           +           +   +/- (ICP)           +   \n",
       "16  Eye and ear abnormalities         NaN         NaN         NaN         NaN   \n",
       "17                        CVI           +           +           -           +   \n",
       "18                     Myopia           +          ND           +           +   \n",
       "19      Eye movement disorder           -           +           +           +   \n",
       "20         Other eye problems           +           +           +           +   \n",
       "21               Hearing loss           -           +           -           +   \n",
       "22            Musculoskeletal         NaN         NaN         NaN         NaN   \n",
       "23              Brachycephaly           +           -           -           +   \n",
       "24        Joint hypermobility           -           +           +           +   \n",
       "25              Hip dysplasia           -           -           -           +   \n",
       "26               Contractures           +           -           +           -   \n",
       "27                      Other         NaN         NaN         NaN         NaN   \n",
       "28    Obstructive sleep apnea           +           +           +           +   \n",
       "29         Precocious puberty           -           +           -         NaN   \n",
       "30        (History of) anemia           +           +           +           -   \n",
       "31           Thrombocytopenia           -           -           -           +   \n",
       "\n",
       "   Count features  \n",
       "0             NaN  \n",
       "1             NaN  \n",
       "2             NaN  \n",
       "3             NaN  \n",
       "4             NaN  \n",
       "5      4/4 (100%)  \n",
       "6      4/4 (100%)  \n",
       "7      4/4 (100%)  \n",
       "8             NaN  \n",
       "9       2/4 (50%)  \n",
       "10      1/4 (25%)  \n",
       "11      3/4 (75%)  \n",
       "12      2/4 (50%)  \n",
       "13            NaN  \n",
       "14     4/4 (100%)  \n",
       "15     4/4 (100%)  \n",
       "16            NaN  \n",
       "17      3/4 (75%)  \n",
       "18     3/3 (100%)  \n",
       "19      3/4 (75%)  \n",
       "20     4/4 (100%)  \n",
       "21      2/4 (50%)  \n",
       "22            NaN  \n",
       "23      2/4 (50%)  \n",
       "24      3/4 (75%)  \n",
       "25      1/4 (25%)  \n",
       "26      2/4 (50%)  \n",
       "27            NaN  \n",
       "28     4/4 (100%)  \n",
       "29      1/3 (33%)  \n",
       "30      3/4 (75%)  \n",
       "31      1/4 (25%)  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7f7bad9",
   "metadata": {},
   "source": [
    "<h1>Converting to row-based format</h1>\n",
    "<p>To use pyphetools, we need to have the individuals represented as rows (one row per individual) and have the items of interest be encoded as column names. The required transformations for doing this may be different for different input data, but often we will want to transpose the table (using the pandas <tt>transpose</tt> function) and set the column names of the new table to the zero-th row. After this, we drop the zero-th row (otherwise, it will be interpreted as an individual by the pyphetools code).</p>\n",
    "<p>After this step is completed, the remaining steps to create phenopackets are the same as in the \n",
    "    <a href=\"http://localhost:8888/notebooks/notebooks/Create%20phenopackets%20from%20tabular%20data%20with%20individuals%20in%20rows.ipynb\" target=\"__blank\">row-based notebook</a>.</p>\n",
    "    \n",
    "Furthermore, for this specific case, there is a Count features row that we want dropped, so we filter out any row that does not have Patient in the first column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c85c793f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Pathogenic variant</th>\n",
       "      <th>NaN</th>\n",
       "      <th>Inheritance</th>\n",
       "      <th>Age at examination</th>\n",
       "      <th>Development</th>\n",
       "      <th>Intellectual disability</th>\n",
       "      <th>Speech delay</th>\n",
       "      <th>Motor delay</th>\n",
       "      <th>Congenital anomalies</th>\n",
       "      <th>Congenital heart defect</th>\n",
       "      <th>...</th>\n",
       "      <th>Musculoskeletal</th>\n",
       "      <th>Brachycephaly</th>\n",
       "      <th>Joint hypermobility</th>\n",
       "      <th>Hip dysplasia</th>\n",
       "      <th>Contractures</th>\n",
       "      <th>Other</th>\n",
       "      <th>Obstructive sleep apnea</th>\n",
       "      <th>Precocious puberty</th>\n",
       "      <th>(History of) anemia</th>\n",
       "      <th>Thrombocytopenia</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Patient 1</th>\n",
       "      <td>c.110G&gt;A</td>\n",
       "      <td>p.Arg37His</td>\n",
       "      <td>De novo</td>\n",
       "      <td>9.5 y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient 2</th>\n",
       "      <td>c.110G&gt;A</td>\n",
       "      <td>p.Arg37His</td>\n",
       "      <td>De novo</td>\n",
       "      <td>5 y 8 mo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient 3</th>\n",
       "      <td>c.110G&gt;A</td>\n",
       "      <td>p.Arg37His</td>\n",
       "      <td>De novo</td>\n",
       "      <td>12 y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient 4</th>\n",
       "      <td>c.110G&gt;A</td>\n",
       "      <td>p.Arg37His</td>\n",
       "      <td>De novo</td>\n",
       "      <td>17 mo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Unnamed: 0 Pathogenic variant         NaN Inheritance Age at examination  \\\n",
       "Patient 1            c.110G>A  p.Arg37His     De novo              9.5 y   \n",
       "Patient 2            c.110G>A  p.Arg37His     De novo           5 y 8 mo   \n",
       "Patient 3            c.110G>A  p.Arg37His     De novo               12 y   \n",
       "Patient 4            c.110G>A  p.Arg37His     De novo              17 mo   \n",
       "\n",
       "Unnamed: 0 Development  Intellectual disability  Speech delay  Motor delay  \\\n",
       "Patient 1          NaN                   Severe        Severe       Severe   \n",
       "Patient 2          NaN                   Severe        Severe       Severe   \n",
       "Patient 3          NaN                   Severe        Severe       Severe   \n",
       "Patient 4          NaN                   Severe        Severe       Severe   \n",
       "\n",
       "Unnamed: 0 Congenital anomalies  Congenital heart defect  ... Musculoskeletal  \\\n",
       "Patient 1                   NaN                        -  ...             NaN   \n",
       "Patient 2                   NaN                        +  ...             NaN   \n",
       "Patient 3                   NaN                        -  ...             NaN   \n",
       "Patient 4                   NaN                        +  ...             NaN   \n",
       "\n",
       "Unnamed: 0  Brachycephaly  Joint hypermobility  Hip dysplasia  Contractures  \\\n",
       "Patient 1               +                    -              -             +   \n",
       "Patient 2               -                    +              -             -   \n",
       "Patient 3               -                    +              -             +   \n",
       "Patient 4               +                    +              +             -   \n",
       "\n",
       "Unnamed: 0 Other  Obstructive sleep apnea  Precocious puberty  \\\n",
       "Patient 1    NaN                        +                   -   \n",
       "Patient 2    NaN                        +                   +   \n",
       "Patient 3    NaN                        +                   -   \n",
       "Patient 4    NaN                        +                 NaN   \n",
       "\n",
       "Unnamed: 0  (History of) anemia  Thrombocytopenia  \n",
       "Patient 1                     +                 -  \n",
       "Patient 2                     +                 -  \n",
       "Patient 3                     +                 -  \n",
       "Patient 4                     -                 +  \n",
       "\n",
       "[4 rows x 32 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft = df.transpose()\n",
    "\n",
    "dft.columns = dft.iloc[0]\n",
    "dft.drop(dft.index[0], inplace=True)\n",
    "dft = dft[dft.index.astype(str).str.contains('Patient')]\n",
    "dft.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bd821df",
   "metadata": {},
   "source": [
    "<h2>Index vs. normal column</h2>\n",
    "<p>Another thing to look out for is whether the individuals (usually the first column) are regarded as the index of the table or as the first normal column.</p>\n",
    "<p>If this is the case, it is easiest to create a new column with the contents of the index -- this will work with the pyphetools software. An example follows -- we can now use 'patient_id' as the column name.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "768b0963",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Pathogenic variant</th>\n",
       "      <th>NaN</th>\n",
       "      <th>Inheritance</th>\n",
       "      <th>Age at examination</th>\n",
       "      <th>Development</th>\n",
       "      <th>Intellectual disability</th>\n",
       "      <th>Speech delay</th>\n",
       "      <th>Motor delay</th>\n",
       "      <th>Congenital anomalies</th>\n",
       "      <th>Congenital heart defect</th>\n",
       "      <th>...</th>\n",
       "      <th>Brachycephaly</th>\n",
       "      <th>Joint hypermobility</th>\n",
       "      <th>Hip dysplasia</th>\n",
       "      <th>Contractures</th>\n",
       "      <th>Other</th>\n",
       "      <th>Obstructive sleep apnea</th>\n",
       "      <th>Precocious puberty</th>\n",
       "      <th>(History of) anemia</th>\n",
       "      <th>Thrombocytopenia</th>\n",
       "      <th>patient_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Patient 1</th>\n",
       "      <td>c.110G&gt;A</td>\n",
       "      <td>p.Arg37His</td>\n",
       "      <td>De novo</td>\n",
       "      <td>9.5 y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>Patient 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient 2</th>\n",
       "      <td>c.110G&gt;A</td>\n",
       "      <td>p.Arg37His</td>\n",
       "      <td>De novo</td>\n",
       "      <td>5 y 8 mo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>Patient 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient 3</th>\n",
       "      <td>c.110G&gt;A</td>\n",
       "      <td>p.Arg37His</td>\n",
       "      <td>De novo</td>\n",
       "      <td>12 y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>Patient 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient 4</th>\n",
       "      <td>c.110G&gt;A</td>\n",
       "      <td>p.Arg37His</td>\n",
       "      <td>De novo</td>\n",
       "      <td>17 mo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>Patient 4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Unnamed: 0 Pathogenic variant         NaN Inheritance Age at examination  \\\n",
       "Patient 1            c.110G>A  p.Arg37His     De novo              9.5 y   \n",
       "Patient 2            c.110G>A  p.Arg37His     De novo           5 y 8 mo   \n",
       "Patient 3            c.110G>A  p.Arg37His     De novo               12 y   \n",
       "Patient 4            c.110G>A  p.Arg37His     De novo              17 mo   \n",
       "\n",
       "Unnamed: 0 Development Intellectual disability Speech delay Motor delay  \\\n",
       "Patient 1          NaN                  Severe       Severe      Severe   \n",
       "Patient 2          NaN                  Severe       Severe      Severe   \n",
       "Patient 3          NaN                  Severe       Severe      Severe   \n",
       "Patient 4          NaN                  Severe       Severe      Severe   \n",
       "\n",
       "Unnamed: 0 Congenital anomalies Congenital heart defect  ... Brachycephaly  \\\n",
       "Patient 1                   NaN                       -  ...             +   \n",
       "Patient 2                   NaN                       +  ...             -   \n",
       "Patient 3                   NaN                       -  ...             -   \n",
       "Patient 4                   NaN                       +  ...             +   \n",
       "\n",
       "Unnamed: 0 Joint hypermobility Hip dysplasia Contractures Other  \\\n",
       "Patient 1                    -             -            +   NaN   \n",
       "Patient 2                    +             -            -   NaN   \n",
       "Patient 3                    +             -            +   NaN   \n",
       "Patient 4                    +             +            -   NaN   \n",
       "\n",
       "Unnamed: 0 Obstructive sleep apnea Precocious puberty (History of) anemia  \\\n",
       "Patient 1                        +                  -                   +   \n",
       "Patient 2                        +                  +                   +   \n",
       "Patient 3                        +                  -                   +   \n",
       "Patient 4                        +                NaN                   -   \n",
       "\n",
       "Unnamed: 0 Thrombocytopenia patient_id  \n",
       "Patient 1                 -  Patient 1  \n",
       "Patient 2                 -  Patient 2  \n",
       "Patient 3                 -  Patient 3  \n",
       "Patient 4                 +  Patient 4  \n",
       "\n",
       "[4 rows x 33 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft.index\n",
    "dft['patient_id'] = dft.index\n",
    "dft.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e013666d",
   "metadata": {},
   "source": [
    "Some column names might include spaces in front or after, so lets make sure that is not the case:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "5d604600",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Pathogenic variant</th>\n",
       "      <th>NaN</th>\n",
       "      <th>Inheritance</th>\n",
       "      <th>Age at examination</th>\n",
       "      <th>Development</th>\n",
       "      <th>Intellectual disability</th>\n",
       "      <th>Speech delay</th>\n",
       "      <th>Motor delay</th>\n",
       "      <th>Congenital anomalies</th>\n",
       "      <th>Congenital heart defect</th>\n",
       "      <th>...</th>\n",
       "      <th>Brachycephaly</th>\n",
       "      <th>Joint hypermobility</th>\n",
       "      <th>Hip dysplasia</th>\n",
       "      <th>Contractures</th>\n",
       "      <th>Other</th>\n",
       "      <th>Obstructive sleep apnea</th>\n",
       "      <th>Precocious puberty</th>\n",
       "      <th>(History of) anemia</th>\n",
       "      <th>Thrombocytopenia</th>\n",
       "      <th>patient_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Patient 1</th>\n",
       "      <td>c.110G&gt;A</td>\n",
       "      <td>p.Arg37His</td>\n",
       "      <td>De novo</td>\n",
       "      <td>9.5 y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>Patient 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient 2</th>\n",
       "      <td>c.110G&gt;A</td>\n",
       "      <td>p.Arg37His</td>\n",
       "      <td>De novo</td>\n",
       "      <td>5 y 8 mo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>Patient 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient 3</th>\n",
       "      <td>c.110G&gt;A</td>\n",
       "      <td>p.Arg37His</td>\n",
       "      <td>De novo</td>\n",
       "      <td>12 y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>Patient 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient 4</th>\n",
       "      <td>c.110G&gt;A</td>\n",
       "      <td>p.Arg37His</td>\n",
       "      <td>De novo</td>\n",
       "      <td>17 mo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Severe</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>Patient 4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Unnamed: 0 Pathogenic variant         NaN Inheritance Age at examination  \\\n",
       "Patient 1            c.110G>A  p.Arg37His     De novo              9.5 y   \n",
       "Patient 2            c.110G>A  p.Arg37His     De novo           5 y 8 mo   \n",
       "Patient 3            c.110G>A  p.Arg37His     De novo               12 y   \n",
       "Patient 4            c.110G>A  p.Arg37His     De novo              17 mo   \n",
       "\n",
       "Unnamed: 0 Development Intellectual disability Speech delay Motor delay  \\\n",
       "Patient 1          NaN                  Severe       Severe      Severe   \n",
       "Patient 2          NaN                  Severe       Severe      Severe   \n",
       "Patient 3          NaN                  Severe       Severe      Severe   \n",
       "Patient 4          NaN                  Severe       Severe      Severe   \n",
       "\n",
       "Unnamed: 0 Congenital anomalies Congenital heart defect  ... Brachycephaly  \\\n",
       "Patient 1                   NaN                       -  ...             +   \n",
       "Patient 2                   NaN                       +  ...             -   \n",
       "Patient 3                   NaN                       -  ...             -   \n",
       "Patient 4                   NaN                       +  ...             +   \n",
       "\n",
       "Unnamed: 0 Joint hypermobility Hip dysplasia Contractures Other  \\\n",
       "Patient 1                    -             -            +   NaN   \n",
       "Patient 2                    +             -            -   NaN   \n",
       "Patient 3                    +             -            +   NaN   \n",
       "Patient 4                    +             +            -   NaN   \n",
       "\n",
       "Unnamed: 0 Obstructive sleep apnea Precocious puberty (History of) anemia  \\\n",
       "Patient 1                        +                  -                   +   \n",
       "Patient 2                        +                  +                   +   \n",
       "Patient 3                        +                  -                   +   \n",
       "Patient 4                        +                NaN                   -   \n",
       "\n",
       "Unnamed: 0 Thrombocytopenia patient_id  \n",
       "Patient 1                 -  Patient 1  \n",
       "Patient 2                 -  Patient 2  \n",
       "Patient 3                 -  Patient 3  \n",
       "Patient 4                 +  Patient 4  \n",
       "\n",
       "[4 rows x 33 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft.columns = dft.columns.str.strip()\n",
    "dft.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8314d6bd",
   "metadata": {},
   "source": [
    "<h2>Column mappers</h2>\n",
    "<p>Please see the notebook \"Create phenopackets from tabular data with individuals in rows\" for explanations. In the following cell we create a dictionary for the ColumnMappers. Note that the code is identical except that we use the df.loc function to get the corresponding row data</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "59d9e9e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "column_mapper_d = defaultdict(ColumnMapper)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "135f303f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def try_mapping_columns(hpo_id, observed, excluded, df):\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "8e2e313a",
   "metadata": {},
   "outputs": [],
   "source": [
    "congenital_heart_defect_mapper = SimpleColumnMapper(hpo_id='HP:0001627',\n",
    "    hpo_label='Abnormal heart morphology',\n",
    "    observed='+',\n",
    "    excluded='-')\n",
    "congenital_heart_defect_mapper.preview_column(dft['Congenital heart defect'])\n",
    "column_mapper_d['Congenital heart defect'] = congenital_heart_defect_mapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "2e022f7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "brachycephaly_mapper = SimpleColumnMapper(hpo_id='HP:0000248',\n",
    "    hpo_label='Brachycephaly',\n",
    "    observed='+',\n",
    "    excluded='-')\n",
    "brachycephaly_mapper.preview_column(dft['Brachycephaly'])\n",
    "column_mapper_d['Brachycephaly'] = brachycephaly_mapper"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4205a563",
   "metadata": {},
   "source": [
    "<h2>Other columns</h2>\n",
    "<p>The following columns are shown without detailed explanations. See \n",
    "    <a href=\"http://localhost:8888/notebooks/notebooks/Create%20phenopackets%20from%20tabular%20data%20with%20individuals%20in%20rows.ipynb\" target=\"__blank\">row-based notebook</a>\n",
    "    for more information.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "90b39bc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "severity_d = {'Severe': 'Intellectual disability, severe',\n",
    "             'Profound': 'Intellectual disability, profound'}\n",
    "idMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=severity_d)\n",
    "#idMapper.preview_column(dft['Intellectual disability'])\n",
    "column_mapper_d['Intellectual disability'] = idMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "881182cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Autistic behavior  Autistic behavior HP:0000729\n",
    "autisticMapper = SimpleColumnMapper(hpo_id='HP:0000729',\n",
    "    hpo_label='Autistic behavior',\n",
    "    observed='+',\n",
    "    excluded='−')\n",
    "#autisticMapper.preview_column(dft['Autistic behavior'])\n",
    "column_mapper_d['Autistic behavior'] = autisticMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "89834a4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Language skills\n",
    "\n",
    "# Simple two-word sentences \tSimple words \tNonverbal \t\n",
    "language_d = {'Simple two-word sentences': 'Delayed speech and language development',\n",
    "             'Simple words': 'Delayed speech and language development',\n",
    "             'Nonverbal': 'Absent speech'}\n",
    "languageMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=language_d)\n",
    "# languageMapper.preview_column(dft['Language skills'])\n",
    "column_mapper_d['Language skills'] = languageMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ea012719",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Spastic diplegia \tSpastic diplegia HP:0001264\n",
    "spasticDiplegiaMapper = SimpleColumnMapper(hpo_id='HP:0001264',\n",
    "    hpo_label='Spastic diplegia',\n",
    "    observed='+',\n",
    "    excluded='−')\n",
    "#spasticDiplegiaMapper.preview_column(dft['Spastic diplegia'])\n",
    "column_mapper_d['Spastic diplegia'] = spasticDiplegiaMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "eb0933ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gross motor skills Wheelchair bound \tWheelchair bound \tWheelchair bound \tCruising (5y)\tWalking  (5y)\n",
    "gms_d = {\n",
    "    \"Wheelchair bound\": \"Loss of ambulation\",\n",
    "    \"Cruising\": \"Delayed gross motor development\"\n",
    "}\n",
    "gmsMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=gms_d)\n",
    "# gmsMapper.preview_column(dft['Gross motor skills'])\n",
    "column_mapper_d['Gross motor skills'] = gmsMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c76760cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Infantile hypotonia Infantile muscular hypotonia HP:0008947\n",
    "hypotoniaMapper = SimpleColumnMapper(hpo_id='HP:0008947',\n",
    "    hpo_label='Infantile muscular hypotonia',\n",
    "    observed='+',\n",
    "    excluded='−')\n",
    "hypotoniaMapper.preview_column(dft['Infantile hypotonia'])\n",
    "column_mapper_d['Infantile hypotonia'] = hypotoniaMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "41d59581",
   "metadata": {},
   "outputs": [],
   "source": [
    "# History of regression Developmental regression HP:0002376\n",
    "regressionMapper = SimpleColumnMapper(hpo_id='HP:0002376',\n",
    "    hpo_label='Developmental regression',\n",
    "    observed='+',\n",
    "    excluded='−')\n",
    "# regressionMapper.preview_column(dft['History of regression'])\n",
    "column_mapper_d['History of regression'] = regressionMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "1af9d0de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Epilepsy  Seizure HP:0001250\n",
    "seizureMapper = SimpleColumnMapper(hpo_id='HP:0001250',\n",
    "    hpo_label='Seizure',\n",
    "    observed='+',\n",
    "    excluded='−')\n",
    "# seizureMapper.preview_column(dft['Epilepsy'])\n",
    "column_mapper_d['Epilepsy'] = seizureMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "df6701c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cerebral atrophy \tCerebral atrophy HP:0002059\n",
    "# Note -- record the 'Mild' version as not measured\n",
    "cerebralAtrophyMapper = SimpleColumnMapper(hpo_id='HP:0002059',\n",
    "    hpo_label='Cerebral atrophy',\n",
    "    observed='+',\n",
    "    excluded='−')\n",
    "# cerebralAtrophyMapper.preview_column(dft['Cerebral atrophy'])\n",
    "column_mapper_d['Cerebral atrophy'] = cerebralAtrophyMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b8080dd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Delayed myelination  Delayed CNS myelination HP:0002188\n",
    "delayedCNSMy = SimpleColumnMapper(hpo_id='HP:0002188',\n",
    "    hpo_label='Delayed CNS myelination',\n",
    "    observed='+',\n",
    "    excluded='−')\n",
    "#delayedCNSMy.preview_column(dft['Delayed myelination'])\n",
    "column_mapper_d['Delayed myelination'] = delayedCNSMy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "01154af3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Corpus callosum hypoplasia  Hypoplasia of the corpus callosum HP:0002079\n",
    "cchMapper = SimpleColumnMapper(hpo_id='HP:0002079',\n",
    "    hpo_label='Hypoplasia of the corpus callosum',\n",
    "    observed='++',\n",
    "    excluded='−')\n",
    "# cchMapper.preview_column(dft['Corpus callosum hypoplasia'])\n",
    "column_mapper_d['Corpus callosum hypoplasia'] = cchMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "484d4198",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Round face  - Round face HP:0000311\n",
    "roundFaceMapper = SimpleColumnMapper(hpo_id='HP:0000311',\n",
    "    hpo_label='Round face',\n",
    "    observed='+',\n",
    "    excluded='−')\n",
    "#roundFaceMapper.preview_column(dft['Round face'])\n",
    "column_mapper_d['Round face'] = roundFaceMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "882d9b19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prominent nasal bridge \tProminent nasal bridge HP:0000426\n",
    "promNasalBridgeMapper = SimpleColumnMapper(hpo_id='HP:0000426',\n",
    "    hpo_label='Prominent nasal bridge',\n",
    "    observed='+',\n",
    "    excluded='−')\n",
    "#promNasalBridgeMapper.preview_column(dft['Prominent nasal bridge'])\n",
    "column_mapper_d['Prominent nasal bridge'] = promNasalBridgeMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "eb5920af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Thin upper lip -- Thin upper lip vermilion HP:0000219\n",
    "upperLipMapper = SimpleColumnMapper(hpo_id='HP:0000219',\n",
    "    hpo_label='Thin upper lip vermilion',\n",
    "    observed='+',\n",
    "    excluded='−')\n",
    "#upperLipMapper.preview_column(dft['Thin upper lip'])\n",
    "column_mapper_d['Thin upper lip'] = upperLipMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "65c27c04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Others\n",
    "other_d = {'upper slanted palpebral fissures': 'Upslanted palpebral fissure'}\n",
    "otherMapper = CustomColumnMapper(concept_recognizer=hpo_cr, custom_map_d=other_d)\n",
    "#otherMapper.preview_column(dft['Others'])\n",
    "column_mapper_d['Others'] = otherMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "49dfb286",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Short stature  Short stature HP:0004322\n",
    "shortStatureMapper = SimpleColumnMapper(hpo_id='HP:0004322',\n",
    "    hpo_label='Short stature',\n",
    "    observed='+',\n",
    "    excluded='−')\n",
    "#shortStatureMapper.preview_column(dft['Short stature'])\n",
    "column_mapper_d['Short stature'] = shortStatureMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ff41a09a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Obesity -- Obesity HP:0001513\n",
    "obesityMapper = SimpleColumnMapper(hpo_id='HP:0004322',\n",
    "    hpo_label='Obesity',\n",
    "    observed='+',\n",
    "    excluded='−')\n",
    "#obesityMapper.preview_column(dft['Obesity'])\n",
    "column_mapper_d['Obesity'] = obesityMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "23e34032",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Precocious puberty -- Precocious puberty HP:0000826\n",
    "precociousMapper = SimpleColumnMapper(hpo_id='HP:0004322',\n",
    "    hpo_label='Precocious puberty',\n",
    "    observed='+',\n",
    "    excluded='−')\n",
    "#precociousMapper.preview_column(dft['Precocious puberty'])\n",
    "column_mapper_d['Precocious puberty'] = precociousMapper"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e53617d",
   "metadata": {},
   "source": [
    "<h2>Variant Data</h2>\n",
    "<p>The variant data (HGVS< transcript) is listed in the Variant (hg19, NM_015133.4) column.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "da4d5706",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = 'hg38'\n",
    "default_genotype = 'heterozygous'\n",
    "transcript='NM_015133.4'\n",
    "varMapper = VariantColumnMapper(assembly=genome,column_name='Variant (hg19, NM_015133.4)', \n",
    "                                transcript=transcript, genotype=default_genotype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "5e7cab22",
   "metadata": {},
   "outputs": [],
   "source": [
    "#varMapper.preview_column(dft['Variant (hg19, NM_015133.4)'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1af2052",
   "metadata": {},
   "source": [
    "<h1>Demographic data</h1>\n",
    "<p>pyphetools can be used to capture information about age, sex, and individual identifiers. This information is stored in a map of \"IndividualMapper\" objects. Special treatment may be required for the indifiers, which may be used as the column names or row index.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "3e64dc08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original column contents</th>\n",
       "      <th>age</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>29</td>\n",
       "      <td>P29Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>27</td>\n",
       "      <td>P27Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16</td>\n",
       "      <td>P16Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>P5Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   original column contents   age\n",
       "0                        29  P29Y\n",
       "1                        27  P27Y\n",
       "2                        16  P16Y\n",
       "3                         5   P5Y"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ageMapper = AgeColumnMapper.by_year('Age (yr)')\n",
    "ageMapper.preview_column(dft['Age (yr)'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "71f664cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original column contents</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Male</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Female</td>\n",
       "      <td>FEMALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Male</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Male</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Female</td>\n",
       "      <td>FEMALE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  original column contents     sex\n",
       "0                     Male    MALE\n",
       "1                   Female  FEMALE\n",
       "2                     Male    MALE\n",
       "3                     Male    MALE\n",
       "4                   Female  FEMALE"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sexMapper = SexColumnMapper(male_symbol='Male', female_symbol='Female', column_name='Sex')\n",
    "sexMapper.preview_column(dft['Sex'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "f6581a8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pmid = \"PMID: 30612693\"\n",
    "encoder = CohortEncoder(df=dft, hpo_cr=hpo_cr, column_mapper_d=column_mapper_d, \n",
    "                        individual_column_name=\"patient_id\", agemapper=ageMapper, sexmapper=sexMapper,\n",
    "                       variant_mapper=varMapper,\n",
    "                       pmid=pmid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "fd367ed6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_015133.4%3Ac.1732C>T/NM_015133.4\n",
      "size of variant_list is {len(variant_list)}\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_015133.4%3Ac.1732C>T/NM_015133.4\n",
      "size of variant_list is {len(variant_list)}\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_015133.4%3Ac.1732C>T/NM_015133.4\n",
      "size of variant_list is {len(variant_list)}\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_015133.4%3Ac.3436C>T/NM_015133.4\n",
      "size of variant_list is {len(variant_list)}\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_015133.4%3Ac.3436C>T/NM_015133.4\n",
      "size of variant_list is {len(variant_list)}\n"
     ]
    }
   ],
   "source": [
    "individuals = encoder.get_individuals()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "5d044b78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Individual, size of variants 1\n",
      "{\n",
      "  \"id\": \"Individual 1\",\n",
      "  \"subject\": {\n",
      "    \"id\": \"Individual 1\",\n",
      "    \"timeAtLastEncounter\": {\n",
      "      \"age\": {\n",
      "        \"iso8601duration\": \"P29Y\"\n",
      "      }\n",
      "    },\n",
      "    \"sex\": \"MALE\"\n",
      "  },\n",
      "  \"phenotypicFeatures\": [\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0001270\",\n",
      "        \"label\": \"Motor delay\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0032988\",\n",
      "        \"label\": \"Persistent head lag\"\n",
      "      },\n",
      "      \"excluded\": true\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0025336\",\n",
      "        \"label\": \"Delayed ability to sit\"\n",
      "      },\n",
      "      \"excluded\": true\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0031936\",\n",
      "        \"label\": \"Delayed ability to walk\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0010864\",\n",
      "        \"label\": \"Intellectual disability, severe\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0000729\",\n",
      "        \"label\": \"Autistic behavior\"\n",
      "      },\n",
      "      \"excluded\": true\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0000750\",\n",
      "        \"label\": \"Delayed speech and language development\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0001264\",\n",
      "        \"label\": \"Spastic diplegia\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0002505\",\n",
      "        \"label\": \"Loss of ambulation\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0008947\",\n",
      "        \"label\": \"Infantile muscular hypotonia\"\n",
      "      },\n",
      "      \"excluded\": true\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0002376\",\n",
      "        \"label\": \"Developmental regression\"\n",
      "      },\n",
      "      \"excluded\": true\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0001250\",\n",
      "        \"label\": \"Seizure\"\n",
      "      },\n",
      "      \"excluded\": true\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0002059\",\n",
      "        \"label\": \"Cerebral atrophy\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0002188\",\n",
      "        \"label\": \"Delayed CNS myelination\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0000311\",\n",
      "        \"label\": \"Round face\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0000426\",\n",
      "        \"label\": \"Prominent nasal bridge\"\n",
      "      },\n",
      "      \"excluded\": true\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0000219\",\n",
      "        \"label\": \"Thin upper lip vermilion\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0004322\",\n",
      "        \"label\": \"Short stature\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0004322\",\n",
      "        \"label\": \"Obesity\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0004322\",\n",
      "        \"label\": \"Precocious puberty\"\n",
      "      }\n",
      "    }\n",
      "  ],\n",
      "  \"interpretations\": [\n",
      "    {\n",
      "      \"id\": \"Individual 1\",\n",
      "      \"progressStatus\": \"SOLVED\",\n",
      "      \"diagnosis\": {\n",
      "        \"genomicInterpretations\": [\n",
      "          {\n",
      "            \"subjectOrBiosampleId\": \"Individual 1\",\n",
      "            \"interpretationStatus\": \"CAUSATIVE\",\n",
      "            \"variantInterpretation\": {\n",
      "              \"variationDescriptor\": {\n",
      "                \"geneContext\": {\n",
      "                  \"valueId\": \"HGNC:6884\",\n",
      "                  \"symbol\": \"MAPK8IP3\"\n",
      "                },\n",
      "                \"expressions\": [\n",
      "                  {\n",
      "                    \"syntax\": \"hgvs.c\",\n",
      "                    \"value\": \"NM_015133.4:c.1732C>T\"\n",
      "                  },\n",
      "                  {\n",
      "                    \"syntax\": \"hgvs.g\",\n",
      "                    \"value\": \"NC_000016.10:g.1762843C>T\"\n",
      "                  }\n",
      "                ],\n",
      "                \"vcfRecord\": {\n",
      "                  \"genomeAssembly\": \"hg38\",\n",
      "                  \"chrom\": \"chr16\",\n",
      "                  \"pos\": \"1762843\",\n",
      "                  \"ref\": \"C\",\n",
      "                  \"alt\": \"T\"\n",
      "                },\n",
      "                \"moleculeContext\": \"genomic\",\n",
      "                \"allelicState\": {\n",
      "                  \"id\": \"GENO:0000135\",\n",
      "                  \"label\": \"heterozygous\"\n",
      "                }\n",
      "              }\n",
      "            }\n",
      "          }\n",
      "        ]\n",
      "      }\n",
      "    }\n",
      "  ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "i1 = individuals[0]\n",
    "phenopacket1 = i1.to_ga4gh_phenopacket()\n",
    "json_string = MessageToJson(phenopacket1)\n",
    "print(json_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "23f1094f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Individual, size of variants 1\n",
      "Individual, size of variants 1\n",
      "Individual, size of variants 1\n",
      "Individual, size of variants 1\n",
      "Individual, size of variants 1\n",
      "Wrote 5 phenopackets to phenopackets\n"
     ]
    }
   ],
   "source": [
    "output_directory = \"phenopackets\"\n",
    "encoder.output_phenopackets(outdir=output_directory, individual_list=individuals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31305025",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
