{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a39f3c65",
   "metadata": {},
   "source": [
    "<h1>Creation of phenopackets from tabular data (individuals in columns)</h1>\n",
    "<p>We will process <a href=\"https://pubmed.ncbi.nlm.nih.gov/34101994//\" target=\"__blank\">Kordes, et al. (2021) Evidence for a low-penetrant extended phenotype of rhabdoid tumor predisposition syndrome type 1 from a kindred with gain of SMARCB1 exon 6</a> and\n",
    "    <a href=\"https://pubmed.ncbi.nlm.nih.gov/33470921/\" target=\"__blank\">Baker, et al. (2021)Epithelioid Sarcoma Arising in a Long-Term Survivor of an Atypical Teratoid/Rhabdoid Tumor in a Patient With Rhabdoid Tumor Predisposition Syndrome</a>\n",
    "here.\n",
    "We need a different approach, because there are no clinical tables in these papers</p>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f0602c0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import phenopackets as php\n",
    "from google.protobuf.json_format import MessageToDict, MessageToJson\n",
    "from google.protobuf.json_format import Parse, ParseDict\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_colwidth', None) # show entire column contents, important!\n",
    "from collections import defaultdict\n",
    "import os\n",
    "import sys\n",
    "sys.path.insert(0, os.path.abspath('../../../pyphetools'))\n",
    "sys.path.insert(0, os.path.abspath('../../pyphetools'))\n",
    "from pyphetools.creation import *\n",
    "from pyphetools.creation.simple_column_mapper import try_mapping_columns\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8606e7eb",
   "metadata": {},
   "source": [
    "<h2>Importing HPO data</h2>\n",
    "<p>pyphetools uses the Human Phenotype Ontology (HPO) to encode phenotypic features. The recommended way of doing this is to ingest the hp.json file using HpoParser, which in turn creates an HpoConceptRecognizer object. </p>\n",
    "<p>The HpoParser can accept a hpo_json_file argument if you want to use a specific file. If the argument is not passed, it will download the latext hp.json file from the HPO GitHub site and store it in a new subdirectory called hpo_data. It will not download the file if the file is already downloaded.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5a7789fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = HpoParser()\n",
    "hpo_cr = parser.get_hpo_concept_recognizer()\n",
    "hpo_version = parser.get_version()\n",
    "metadata = MetaData(created_by=\"ORCID:0000-0002-5648-2155\")\n",
    "metadata.default_versions_with_hpo(version=hpo_version)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5102722a",
   "metadata": {},
   "source": [
    "<h2>Creating and loading the table</h2>\n",
    "<p>Both papers do not have specific clinical tables, therefor, we have first created one manually, that we will now load.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f17c3924",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel('../../data/SMARCB1/PMID_34101994_and_33470921.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "06ec7dbd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>pt_1</th>\n",
       "      <th>pt_2</th>\n",
       "      <th>pt_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMID</td>\n",
       "      <td>33470921</td>\n",
       "      <td>34101994</td>\n",
       "      <td>34101994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>age</td>\n",
       "      <td>16</td>\n",
       "      <td>21</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sex</td>\n",
       "      <td>male</td>\n",
       "      <td>male</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pathogenic variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Feeding difficulties</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Lethargy</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Vomiting</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Hydrocephalus</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Neoplasm</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Atypical teratoid/rhabdoid tumor</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Flesh colored papules</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Granuloma</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Hemiparesis</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Stroke</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Leukemia</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Ulcerative colitis</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Specific learning disability</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Ependymoma</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Unnamed: 0      pt_1      pt_2      pt_3\n",
       "0                               PMID  33470921  34101994  34101994\n",
       "1                                age        16        21        57\n",
       "2                                sex      male      male      male\n",
       "3                 pathogenic variant       NaN       NaN       NaN\n",
       "4               Feeding difficulties         +       NaN       NaN\n",
       "5                           Lethargy         +       NaN       NaN\n",
       "6                           Vomiting         +       NaN       NaN\n",
       "7                      Hydrocephalus         +       NaN       NaN\n",
       "8                           Neoplasm         +         +         +\n",
       "9   Atypical teratoid/rhabdoid tumor         +         +       NaN\n",
       "10             Flesh colored papules         +       NaN       NaN\n",
       "11                         Granuloma         +       NaN       NaN\n",
       "12                       Hemiparesis         +       NaN       NaN\n",
       "13                            Stroke         +       NaN       NaN\n",
       "14                          Leukemia         -         -         +\n",
       "15                Ulcerative colitis         -         -         +\n",
       "16      Specific learning disability         -         +         -\n",
       "17                        Ependymoma         -         -         +"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7f7bad9",
   "metadata": {},
   "source": [
    "<h1>Converting to row-based format</h1>\n",
    "<p>To use pyphetools, we need to have the individuals represented as rows (one row per individual) and have the items of interest be encoded as column names. The required transformations for doing this may be different for different input data, but often we will want to transpose the table (using the pandas <tt>transpose</tt> function) and set the column names of the new table to the zero-th row. After this, we drop the zero-th row (otherwise, it will be interpreted as an individual by the pyphetools code).</p>\n",
    "<p>After this step is completed, the remaining steps to create phenopackets are the same as in the \n",
    "    <a href=\"http://localhost:8888/notebooks/notebooks/Create%20phenopackets%20from%20tabular%20data%20with%20individuals%20in%20rows.ipynb\" target=\"__blank\">row-based notebook</a>.</p>\n",
    "    \n",
    "Furthermore, for this specific case, there is a Count features row that we want dropped, so we filter out any row that does not have Patient in the first column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c85c793f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>PMID</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "      <th>pathogenic variant</th>\n",
       "      <th>Feeding difficulties</th>\n",
       "      <th>Lethargy</th>\n",
       "      <th>Vomiting</th>\n",
       "      <th>Hydrocephalus</th>\n",
       "      <th>Neoplasm</th>\n",
       "      <th>Atypical teratoid/rhabdoid tumor</th>\n",
       "      <th>Flesh colored papules</th>\n",
       "      <th>Granuloma</th>\n",
       "      <th>Hemiparesis</th>\n",
       "      <th>Stroke</th>\n",
       "      <th>Leukemia</th>\n",
       "      <th>Ulcerative colitis</th>\n",
       "      <th>Specific learning disability</th>\n",
       "      <th>Ependymoma</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pt_1</th>\n",
       "      <td>33470921</td>\n",
       "      <td>16</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pt_2</th>\n",
       "      <td>34101994</td>\n",
       "      <td>21</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pt_3</th>\n",
       "      <td>34101994</td>\n",
       "      <td>57</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Unnamed: 0      PMID age   sex pathogenic variant Feeding difficulties  \\\n",
       "pt_1        33470921  16  male                NaN                    +   \n",
       "pt_2        34101994  21  male                NaN                  NaN   \n",
       "pt_3        34101994  57  male                NaN                  NaN   \n",
       "\n",
       "Unnamed: 0 Lethargy Vomiting Hydrocephalus Neoplasm  \\\n",
       "pt_1              +        +             +        +   \n",
       "pt_2            NaN      NaN           NaN        +   \n",
       "pt_3            NaN      NaN           NaN        +   \n",
       "\n",
       "Unnamed: 0 Atypical teratoid/rhabdoid tumor Flesh colored papules Granuloma  \\\n",
       "pt_1                                      +                     +         +   \n",
       "pt_2                                      +                   NaN       NaN   \n",
       "pt_3                                    NaN                   NaN       NaN   \n",
       "\n",
       "Unnamed: 0 Hemiparesis Stroke Leukemia Ulcerative colitis  \\\n",
       "pt_1                 +      +        -                  -   \n",
       "pt_2               NaN    NaN        -                  -   \n",
       "pt_3               NaN    NaN        +                  +   \n",
       "\n",
       "Unnamed: 0 Specific learning disability Ependymoma  \n",
       "pt_1                                  -          -  \n",
       "pt_2                                  +          -  \n",
       "pt_3                                  -          +  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft = df.transpose()\n",
    "dft.columns = dft.iloc[0]\n",
    "dft.drop(dft.index[0], inplace=True)\n",
    "dft.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bd821df",
   "metadata": {},
   "source": [
    "<h2>Index vs. normal column</h2>\n",
    "<p>Another thing to look out for is whether the individuals (usually the first column) are regarded as the index of the table or as the first normal column.</p>\n",
    "<p>If this is the case, it is easiest to create a new column with the contents of the index -- this will work with the pyphetools software. An example follows -- we can now use 'patient_id' as the column name.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "768b0963",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>PMID</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "      <th>pathogenic variant</th>\n",
       "      <th>Feeding difficulties</th>\n",
       "      <th>Lethargy</th>\n",
       "      <th>Vomiting</th>\n",
       "      <th>Hydrocephalus</th>\n",
       "      <th>Neoplasm</th>\n",
       "      <th>Atypical teratoid/rhabdoid tumor</th>\n",
       "      <th>Flesh colored papules</th>\n",
       "      <th>Granuloma</th>\n",
       "      <th>Hemiparesis</th>\n",
       "      <th>Stroke</th>\n",
       "      <th>Leukemia</th>\n",
       "      <th>Ulcerative colitis</th>\n",
       "      <th>Specific learning disability</th>\n",
       "      <th>Ependymoma</th>\n",
       "      <th>patient_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pt_1</th>\n",
       "      <td>33470921</td>\n",
       "      <td>16</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>pt_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pt_2</th>\n",
       "      <td>34101994</td>\n",
       "      <td>21</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>pt_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pt_3</th>\n",
       "      <td>34101994</td>\n",
       "      <td>57</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>pt_3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Unnamed: 0      PMID age   sex pathogenic variant Feeding difficulties  \\\n",
       "pt_1        33470921  16  male                NaN                    +   \n",
       "pt_2        34101994  21  male                NaN                  NaN   \n",
       "pt_3        34101994  57  male                NaN                  NaN   \n",
       "\n",
       "Unnamed: 0 Lethargy Vomiting Hydrocephalus Neoplasm  \\\n",
       "pt_1              +        +             +        +   \n",
       "pt_2            NaN      NaN           NaN        +   \n",
       "pt_3            NaN      NaN           NaN        +   \n",
       "\n",
       "Unnamed: 0 Atypical teratoid/rhabdoid tumor Flesh colored papules Granuloma  \\\n",
       "pt_1                                      +                     +         +   \n",
       "pt_2                                      +                   NaN       NaN   \n",
       "pt_3                                    NaN                   NaN       NaN   \n",
       "\n",
       "Unnamed: 0 Hemiparesis Stroke Leukemia Ulcerative colitis  \\\n",
       "pt_1                 +      +        -                  -   \n",
       "pt_2               NaN    NaN        -                  -   \n",
       "pt_3               NaN    NaN        +                  +   \n",
       "\n",
       "Unnamed: 0 Specific learning disability Ependymoma patient_id  \n",
       "pt_1                                  -          -       pt_1  \n",
       "pt_2                                  +          -       pt_2  \n",
       "pt_3                                  -          +       pt_3  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft.index\n",
    "dft['patient_id'] = dft.index\n",
    "dft.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e013666d",
   "metadata": {},
   "source": [
    "Some column names might include spaces in front or after, and a couple of columns are subheadings and only contain NaNs, so lets correct that:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5d604600",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>PMID</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "      <th>Feeding difficulties</th>\n",
       "      <th>Lethargy</th>\n",
       "      <th>Vomiting</th>\n",
       "      <th>Hydrocephalus</th>\n",
       "      <th>Neoplasm</th>\n",
       "      <th>Atypical teratoid/rhabdoid tumor</th>\n",
       "      <th>Flesh colored papules</th>\n",
       "      <th>Granuloma</th>\n",
       "      <th>Hemiparesis</th>\n",
       "      <th>Stroke</th>\n",
       "      <th>Leukemia</th>\n",
       "      <th>Ulcerative colitis</th>\n",
       "      <th>Specific learning disability</th>\n",
       "      <th>Ependymoma</th>\n",
       "      <th>patient_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pt_1</th>\n",
       "      <td>33470921</td>\n",
       "      <td>16</td>\n",
       "      <td>male</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>pt_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pt_2</th>\n",
       "      <td>34101994</td>\n",
       "      <td>21</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>pt_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pt_3</th>\n",
       "      <td>34101994</td>\n",
       "      <td>57</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>pt_3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Unnamed: 0      PMID age   sex Feeding difficulties Lethargy Vomiting  \\\n",
       "pt_1        33470921  16  male                    +        +        +   \n",
       "pt_2        34101994  21  male                  NaN      NaN      NaN   \n",
       "pt_3        34101994  57  male                  NaN      NaN      NaN   \n",
       "\n",
       "Unnamed: 0 Hydrocephalus Neoplasm Atypical teratoid/rhabdoid tumor  \\\n",
       "pt_1                   +        +                                +   \n",
       "pt_2                 NaN        +                                +   \n",
       "pt_3                 NaN        +                              NaN   \n",
       "\n",
       "Unnamed: 0 Flesh colored papules Granuloma Hemiparesis Stroke Leukemia  \\\n",
       "pt_1                           +         +           +      +        -   \n",
       "pt_2                         NaN       NaN         NaN    NaN        -   \n",
       "pt_3                         NaN       NaN         NaN    NaN        +   \n",
       "\n",
       "Unnamed: 0 Ulcerative colitis Specific learning disability Ependymoma  \\\n",
       "pt_1                        -                            -          -   \n",
       "pt_2                        -                            +          -   \n",
       "pt_3                        +                            -          +   \n",
       "\n",
       "Unnamed: 0 patient_id  \n",
       "pt_1             pt_1  \n",
       "pt_2             pt_2  \n",
       "pt_3             pt_3  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft.columns = dft.columns.str.strip()\n",
    "dft = dft.dropna(axis=1, how='all')\n",
    "dft.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8314d6bd",
   "metadata": {},
   "source": [
    "<h2>Column mappers</h2>\n",
    "<p>Please see the notebook \"Create phenopackets from tabular data with individuals in rows\" for explanations. In the following cell we create a dictionary for the ColumnMappers. Note that the code is identical except that we use the df.loc function to get the corresponding row data</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "59d9e9e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                term        status\n",
      "0  Feeding difficulties (HP:0011968)      observed\n",
      "1  Feeding difficulties (HP:0011968)  not measured\n",
      "2  Feeding difficulties (HP:0011968)  not measured\n",
      "                    term        status\n",
      "0  Lethargy (HP:0001254)      observed\n",
      "1  Lethargy (HP:0001254)  not measured\n",
      "2  Lethargy (HP:0001254)  not measured\n",
      "                    term        status\n",
      "0  Vomiting (HP:0002013)      observed\n",
      "1  Vomiting (HP:0002013)  not measured\n",
      "2  Vomiting (HP:0002013)  not measured\n",
      "                         term        status\n",
      "0  Hydrocephalus (HP:0000238)      observed\n",
      "1  Hydrocephalus (HP:0000238)  not measured\n",
      "2  Hydrocephalus (HP:0000238)  not measured\n",
      "                    term    status\n",
      "0  Neoplasm (HP:0002664)  observed\n",
      "1  Neoplasm (HP:0002664)  observed\n",
      "2  Neoplasm (HP:0002664)  observed\n",
      "                                            term        status\n",
      "0  Atypical teratoid/rhabdoid tumor (HP:0034401)      observed\n",
      "1  Atypical teratoid/rhabdoid tumor (HP:0034401)      observed\n",
      "2  Atypical teratoid/rhabdoid tumor (HP:0034401)  not measured\n",
      "                  term        status\n",
      "0  Papule (HP:0200034)      observed\n",
      "1  Papule (HP:0200034)  not measured\n",
      "2  Papule (HP:0200034)  not measured\n",
      "                     term        status\n",
      "0  Granuloma (HP:0032252)      observed\n",
      "1  Granuloma (HP:0032252)  not measured\n",
      "2  Granuloma (HP:0032252)  not measured\n",
      "                       term        status\n",
      "0  Hemiparesis (HP:0001269)      observed\n",
      "1  Hemiparesis (HP:0001269)  not measured\n",
      "2  Hemiparesis (HP:0001269)  not measured\n",
      "                  term        status\n",
      "0  Stroke (HP:0001297)      observed\n",
      "1  Stroke (HP:0001297)  not measured\n",
      "2  Stroke (HP:0001297)  not measured\n",
      "                    term    status\n",
      "0  Leukemia (HP:0001909)  excluded\n",
      "1  Leukemia (HP:0001909)  excluded\n",
      "2  Leukemia (HP:0001909)  observed\n",
      "                              term    status\n",
      "0  Ulcerative colitis (HP:0100279)  excluded\n",
      "1  Ulcerative colitis (HP:0100279)  excluded\n",
      "2  Ulcerative colitis (HP:0100279)  observed\n",
      "                                        term    status\n",
      "0  Specific learning disability (HP:0001328)  excluded\n",
      "1  Specific learning disability (HP:0001328)  observed\n",
      "2  Specific learning disability (HP:0001328)  excluded\n",
      "                      term    status\n",
      "0  Ependymoma (HP:0002888)  excluded\n",
      "1  Ependymoma (HP:0002888)  excluded\n",
      "2  Ependymoma (HP:0002888)  observed\n"
     ]
    }
   ],
   "source": [
    "hpo_cr = parser.get_hpo_concept_recognizer()\n",
    "column_mapper_d, col_not_found = try_mapping_columns(df=dft.loc[:,:],\n",
    "                                                    observed='+',\n",
    "                                                    excluded='-',\n",
    "                                                    hpo_cr=hpo_cr,\n",
    "                                                    preview=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ac3e687",
   "metadata": {},
   "source": [
    "Lets merge the two dicts and the lists, to see which columns are not mapped yet."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "122946b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['PMID', 'age', 'sex', 'patient_id']\n"
     ]
    }
   ],
   "source": [
    "print(col_not_found)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e53617d",
   "metadata": {},
   "source": [
    "<h2>Variant Data</h2>\n",
    "<p>The variant data (HGVS< transcript) is listed in the Variant (hg19, NM_015133.4) column.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "da4d5706",
   "metadata": {},
   "outputs": [],
   "source": [
    "#genome = 'hg38'\n",
    "#default_genotype = 'heterozygous'\n",
    "#transcript='NM_003073.3'\n",
    "#varMapper = VariantColumnMapper(assembly=genome,column_name='Pathogenic variant', \n",
    "#                                transcript=transcript, genotype=default_genotype)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1af2052",
   "metadata": {},
   "source": [
    "<h1>Demographic data</h1>\n",
    "<p>pyphetools can be used to capture information about age, sex, and individual identifiers. This information is stored in a map of \"IndividualMapper\" objects. Special treatment may be required for the indifiers, which may be used as the column names or row index.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3e64dc08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original column contents</th>\n",
       "      <th>age</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16</td>\n",
       "      <td>P16Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21</td>\n",
       "      <td>P21Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>57</td>\n",
       "      <td>P57Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   original column contents   age\n",
       "0                        16  P16Y\n",
       "1                        21  P21Y\n",
       "2                        57  P57Y"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ageMapper = AgeColumnMapper.by_year('age')\n",
    "ageMapper.preview_column(dft['age'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "71f664cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original column contents</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>male</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>male</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>male</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  original column contents   sex\n",
       "0                     male  MALE\n",
       "1                     male  MALE\n",
       "2                     male  MALE"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#sex is not in columns, since it were all females in this paper\n",
    "sexMapper = SexColumnMapper(male_symbol='male', female_symbol='female', column_name='sex')\n",
    "sexMapper.preview_column(dft['sex'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f6581a8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pmid = \"PMID: 33470921\"\n",
    "encoder = CohortEncoder(df=dft.iloc[1:,:], hpo_cr=hpo_cr, column_mapper_d=column_mapper_d, \n",
    "                        individual_column_name=\"patient_id\", agemapper=ageMapper, sexmapper=sexMapper, metadata=metadata,\n",
    "                       pmid=pmid)\n",
    "encoder.set_disease(disease_id='609322', label='Rhabdoid tumor predisposition syndrome-1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fd367ed6",
   "metadata": {},
   "outputs": [],
   "source": [
    "individuals = encoder.get_individuals()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5d044b78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"id\": \"pt_2\",\n",
      "  \"subject\": {\n",
      "    \"id\": \"pt_2\",\n",
      "    \"timeAtLastEncounter\": {\n",
      "      \"age\": {\n",
      "        \"iso8601duration\": \"P21Y\"\n",
      "      }\n",
      "    },\n",
      "    \"sex\": \"MALE\"\n",
      "  },\n",
      "  \"phenotypicFeatures\": [\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0002664\",\n",
      "        \"label\": \"Neoplasm\"\n",
      "      },\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P21Y\"\n",
      "        }\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0034401\",\n",
      "        \"label\": \"Atypical teratoid/rhabdoid tumor\"\n",
      "      },\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P21Y\"\n",
      "        }\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0001909\",\n",
      "        \"label\": \"Leukemia\"\n",
      "      },\n",
      "      \"excluded\": true,\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P21Y\"\n",
      "        }\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0100279\",\n",
      "        \"label\": \"Ulcerative colitis\"\n",
      "      },\n",
      "      \"excluded\": true,\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P21Y\"\n",
      "        }\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0001328\",\n",
      "        \"label\": \"Specific learning disability\"\n",
      "      },\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P21Y\"\n",
      "        }\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0002888\",\n",
      "        \"label\": \"Ependymoma\"\n",
      "      },\n",
      "      \"excluded\": true,\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P21Y\"\n",
      "        }\n",
      "      }\n",
      "    }\n",
      "  ],\n",
      "  \"interpretations\": [\n",
      "    {\n",
      "      \"diagnosis\": {\n",
      "        \"disease\": {\n",
      "          \"id\": \"609322\",\n",
      "          \"label\": \"Rhabdoid tumor predisposition syndrome-1\"\n",
      "        }\n",
      "      }\n",
      "    }\n",
      "  ],\n",
      "  \"metaData\": {\n",
      "    \"created\": \"2023-01-10T14:09:16.272447586Z\",\n",
      "    \"createdBy\": \"ORCID:0000-0002-5648-2155\",\n",
      "    \"resources\": [\n",
      "      {\n",
      "        \"id\": \"geno\",\n",
      "        \"name\": \"Genotype Ontology\",\n",
      "        \"url\": \"http://purl.obolibrary.org/obo/geno.owl\",\n",
      "        \"version\": \"2022-03-05\",\n",
      "        \"namespacePrefix\": \"GENO\",\n",
      "        \"iriPrefix\": \"http://purl.obolibrary.org/obo/GENO_\"\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"hgnc\",\n",
      "        \"name\": \"HUGO Gene Nomenclature Committee\",\n",
      "        \"url\": \"https://www.genenames.org\",\n",
      "        \"version\": \"06/01/23\",\n",
      "        \"namespacePrefix\": \"HGNC\",\n",
      "        \"iriPrefix\": \"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/\"\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"omim\",\n",
      "        \"name\": \"An Online Catalog of Human Genes and Genetic Disorders\",\n",
      "        \"url\": \"https://www.omim.org\",\n",
      "        \"version\": \"January 4, 2023\",\n",
      "        \"namespacePrefix\": \"OMIM\",\n",
      "        \"iriPrefix\": \"https://www.omim.org/entry/\"\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"hp\",\n",
      "        \"name\": \"human phenotype ontology\",\n",
      "        \"url\": \"http://purl.obolibrary.org/obo/hp.owl\",\n",
      "        \"version\": \"2022-12-15\",\n",
      "        \"namespacePrefix\": \"HP\",\n",
      "        \"iriPrefix\": \"http://purl.obolibrary.org/obo/HP_\"\n",
      "      }\n",
      "    ],\n",
      "    \"phenopacketSchemaVersion\": \"2.0\"\n",
      "  }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "i1 = individuals[0]\n",
    "phenopacket1 = i1.to_ga4gh_phenopacket(metadata=metadata.to_ga4gh())\n",
    "json_string = MessageToJson(phenopacket1)\n",
    "print(json_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "23f1094f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 2 phenopackets to ../../phenopackets/SMARCB1/\n"
     ]
    }
   ],
   "source": [
    "output_directory = \"../../phenopackets/SMARCB1/\"\n",
    "encoder.output_phenopackets(outdir=output_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5c5febf1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"id\": \"pt_1\",\n",
      "  \"subject\": {\n",
      "    \"id\": \"pt_1\",\n",
      "    \"timeAtLastEncounter\": {\n",
      "      \"age\": {\n",
      "        \"iso8601duration\": \"P16Y\"\n",
      "      }\n",
      "    },\n",
      "    \"sex\": \"MALE\"\n",
      "  },\n",
      "  \"phenotypicFeatures\": [\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0011968\",\n",
      "        \"label\": \"Feeding difficulties\"\n",
      "      },\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P16Y\"\n",
      "        }\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0001254\",\n",
      "        \"label\": \"Lethargy\"\n",
      "      },\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P16Y\"\n",
      "        }\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0002013\",\n",
      "        \"label\": \"Vomiting\"\n",
      "      },\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P16Y\"\n",
      "        }\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0000238\",\n",
      "        \"label\": \"Hydrocephalus\"\n",
      "      },\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P16Y\"\n",
      "        }\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0002664\",\n",
      "        \"label\": \"Neoplasm\"\n",
      "      },\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P16Y\"\n",
      "        }\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0034401\",\n",
      "        \"label\": \"Atypical teratoid/rhabdoid tumor\"\n",
      "      },\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P16Y\"\n",
      "        }\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0200034\",\n",
      "        \"label\": \"Papule\"\n",
      "      },\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P16Y\"\n",
      "        }\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0032252\",\n",
      "        \"label\": \"Granuloma\"\n",
      "      },\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P16Y\"\n",
      "        }\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0001269\",\n",
      "        \"label\": \"Hemiparesis\"\n",
      "      },\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P16Y\"\n",
      "        }\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0001297\",\n",
      "        \"label\": \"Stroke\"\n",
      "      },\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P16Y\"\n",
      "        }\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0001909\",\n",
      "        \"label\": \"Leukemia\"\n",
      "      },\n",
      "      \"excluded\": true,\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P16Y\"\n",
      "        }\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0100279\",\n",
      "        \"label\": \"Ulcerative colitis\"\n",
      "      },\n",
      "      \"excluded\": true,\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P16Y\"\n",
      "        }\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0001328\",\n",
      "        \"label\": \"Specific learning disability\"\n",
      "      },\n",
      "      \"excluded\": true,\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P16Y\"\n",
      "        }\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0002888\",\n",
      "        \"label\": \"Ependymoma\"\n",
      "      },\n",
      "      \"excluded\": true,\n",
      "      \"onset\": {\n",
      "        \"age\": {\n",
      "          \"iso8601duration\": \"P16Y\"\n",
      "        }\n",
      "      }\n",
      "    }\n",
      "  ],\n",
      "  \"interpretations\": [\n",
      "    {\n",
      "      \"diagnosis\": {\n",
      "        \"disease\": {\n",
      "          \"id\": \"609322\",\n",
      "          \"label\": \"Rhabdoid tumor predisposition syndrome-1\"\n",
      "        }\n",
      "      }\n",
      "    }\n",
      "  ],\n",
      "  \"metaData\": {\n",
      "    \"created\": \"2023-01-10T14:09:16.306772232Z\",\n",
      "    \"createdBy\": \"ORCID:0000-0002-5648-2155\",\n",
      "    \"resources\": [\n",
      "      {\n",
      "        \"id\": \"geno\",\n",
      "        \"name\": \"Genotype Ontology\",\n",
      "        \"url\": \"http://purl.obolibrary.org/obo/geno.owl\",\n",
      "        \"version\": \"2022-03-05\",\n",
      "        \"namespacePrefix\": \"GENO\",\n",
      "        \"iriPrefix\": \"http://purl.obolibrary.org/obo/GENO_\"\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"hgnc\",\n",
      "        \"name\": \"HUGO Gene Nomenclature Committee\",\n",
      "        \"url\": \"https://www.genenames.org\",\n",
      "        \"version\": \"06/01/23\",\n",
      "        \"namespacePrefix\": \"HGNC\",\n",
      "        \"iriPrefix\": \"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/\"\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"omim\",\n",
      "        \"name\": \"An Online Catalog of Human Genes and Genetic Disorders\",\n",
      "        \"url\": \"https://www.omim.org\",\n",
      "        \"version\": \"January 4, 2023\",\n",
      "        \"namespacePrefix\": \"OMIM\",\n",
      "        \"iriPrefix\": \"https://www.omim.org/entry/\"\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"hp\",\n",
      "        \"name\": \"human phenotype ontology\",\n",
      "        \"url\": \"http://purl.obolibrary.org/obo/hp.owl\",\n",
      "        \"version\": \"2022-12-15\",\n",
      "        \"namespacePrefix\": \"HP\",\n",
      "        \"iriPrefix\": \"http://purl.obolibrary.org/obo/HP_\"\n",
      "      }\n",
      "    ],\n",
      "    \"phenopacketSchemaVersion\": \"2.0\"\n",
      "  }\n",
      "}\n",
      "Wrote 1 phenopackets to ../../phenopackets/SMARCB1/\n"
     ]
    }
   ],
   "source": [
    "pmid = \"PMID: 34101994\"\n",
    "encoder = CohortEncoder(df=pd.DataFrame(dft.iloc[0,:]).T, hpo_cr=hpo_cr, column_mapper_d=column_mapper_d, \n",
    "                        individual_column_name=\"patient_id\", agemapper=ageMapper, sexmapper=sexMapper, metadata=metadata,\n",
    "                       pmid=pmid)\n",
    "encoder.set_disease(disease_id='609322', label='Rhabdoid tumor predisposition syndrome-1')\n",
    "individuals = encoder.get_individuals()\n",
    "i1 = individuals[0]\n",
    "phenopacket1 = i1.to_ga4gh_phenopacket(metadata=metadata.to_ga4gh())\n",
    "json_string = MessageToJson(phenopacket1)\n",
    "print(json_string)\n",
    "output_directory = \"../../phenopackets/SMARCB1/\"\n",
    "encoder.output_phenopackets(outdir=output_directory)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
