{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a39f3c65",
   "metadata": {},
   "source": [
    "<h1>Creation of phenopackets from tabular data (individuals in columns)</h1>\n",
    "<p>We will process <a href=\"https://pubmed.ncbi.nlm.nih.gov/16783569/\" target=\"__blank\">Budny, et al. (2006) A novel X-linked recessive mental retardation syndrome comprising macrocephaly and ciliary dysfunction is allelic to oral-facial-digital type I syndrome</a></p> in this notebook.\n",
    "<p>pyphetools provides a convenient way of extracting HPO terms from typical tables presented in supplemental material. Typical tables can have the individuals in columns or rows. In this case, we extract data from TABLE.</p>\n",
    "<p>This note shows how to work through the table and set up the pyphetools encoder. The table was not originally available in the table, but constructed using the data in the publication</p>\n",
    "<p>Users can work on one column at a time and then generate a collection of <a href=\"https://pubmed.ncbi.nlm.nih.gov/35705716/\" target=\"__blank\">GA4GH phenopackets</a> to represent each patient included in the original supplemental material. These phenopackets can then be used for a variety of downstream applications.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f0602c0e",
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [],
   "source": [
    "import phenopackets as php\n",
    "from google.protobuf.json_format import MessageToDict, MessageToJson\n",
    "from google.protobuf.json_format import Parse, ParseDict\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_colwidth', None) # show entire column contents, important!\n",
    "from collections import defaultdict\n",
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "from pyphetools.creation import *\n",
    "from pyphetools.creation.simple_column_mapper import try_mapping_columns, get_separate_hpos_from_df\n",
    "# Last tested with pyphetools 0.4.5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8606e7eb",
   "metadata": {},
   "source": [
    "<h2>Importing HPO data</h2>\n",
    "<p>pyphetools uses the Human Phenotype Ontology (HPO) to encode phenotypic features. The recommended way of doing this is to ingest the hp.json file using HpoParser, which in turn creates an HpoConceptRecognizer object. </p>\n",
    "<p>The HpoParser can accept a hpo_json_file argument if you want to use a specific file. If the argument is not passed, it will download the latext hp.json file from the HPO GitHub site and store it in a new subdirectory called hpo_data. It will not download the file if the file is already downloaded.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5a7789fc",
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [],
   "source": [
    "parser = HpoParser()\n",
    "hpo_cr = parser.get_hpo_concept_recognizer()\n",
    "hpo_version = parser.get_version()\n",
    "metadata = MetaData(created_by=\"ORCID:0000-0002-5648-2155\")\n",
    "metadata.default_versions_with_hpo(version=hpo_version)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5102722a",
   "metadata": {},
   "source": [
    "<h2>Importing the supplemental table</h2>\n",
    "<p>Here, we use the pandas library to import this file (note that the Python package called openpyxl must be installed to read Excel files with pandas, although the library does not need to be imported in this notebook). pyphetools expects a pandas DataFrame as input, and users can choose any input format available for pandas include CSV, TSV, and Excel, or can use any other method to transform their input data into a Pandas DataFrame before using pyphetools.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f17c3924",
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [],
   "source": [
    "df = pd.read_excel('input/PMID_16783569.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "06ec7dbd",
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>variant</th>\n",
       "      <th>Developmental delay</th>\n",
       "      <th>Abnormal respiratory system physiology</th>\n",
       "      <th>Macrocephaly</th>\n",
       "      <th>Recurrent respiratory infections</th>\n",
       "      <th>Intellectual disability, severe</th>\n",
       "      <th>High palate</th>\n",
       "      <th>Low-set ears</th>\n",
       "      <th>Broad thumb</th>\n",
       "      <th>Brachydactyly</th>\n",
       "      <th>Obesity</th>\n",
       "      <th>Large for gestational age</th>\n",
       "      <th>Postaxial polydactyly</th>\n",
       "      <th>Inguinal hernia</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>male</td>\n",
       "      <td>11</td>\n",
       "      <td>c.2122_2125dupGAAG</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>male</td>\n",
       "      <td>0</td>\n",
       "      <td>c.2122_2125dupGAAG</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>male</td>\n",
       "      <td>0</td>\n",
       "      <td>c.2122_2125dupGAAG</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>male</td>\n",
       "      <td>0</td>\n",
       "      <td>c.2122_2125dupGAAG</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>male</td>\n",
       "      <td>0</td>\n",
       "      <td>c.2122_2125dupGAAG</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>male</td>\n",
       "      <td>0</td>\n",
       "      <td>c.2122_2125dupGAAG</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>male</td>\n",
       "      <td>0</td>\n",
       "      <td>c.2122_2125dupGAAG</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>male</td>\n",
       "      <td>0</td>\n",
       "      <td>c.2122_2125dupGAAG</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>male</td>\n",
       "      <td>0</td>\n",
       "      <td>c.2122_2125dupGAAG</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   individual   sex  age             variant Developmental delay  \\\n",
       "0           1  male   11  c.2122_2125dupGAAG                   +   \n",
       "1           2  male    0  c.2122_2125dupGAAG                   +   \n",
       "2           3  male    0  c.2122_2125dupGAAG                   +   \n",
       "3           4  male    0  c.2122_2125dupGAAG                   +   \n",
       "4           5  male    0  c.2122_2125dupGAAG                   +   \n",
       "5           6  male    0  c.2122_2125dupGAAG                   +   \n",
       "6           7  male    0  c.2122_2125dupGAAG                   +   \n",
       "7           8  male    0  c.2122_2125dupGAAG                   +   \n",
       "8           9  male    0  c.2122_2125dupGAAG                   +   \n",
       "\n",
       "  Abnormal respiratory system physiology Macrocephaly  \\\n",
       "0                                      +            +   \n",
       "1                                      +            +   \n",
       "2                                      +            +   \n",
       "3                                      +            +   \n",
       "4                                      +            +   \n",
       "5                                      +            +   \n",
       "6                                      +            +   \n",
       "7                                      +            +   \n",
       "8                                      +            +   \n",
       "\n",
       "  Recurrent respiratory infections Intellectual disability, severe  \\\n",
       "0                                +                               +   \n",
       "1                                +                               +   \n",
       "2                                +                             NaN   \n",
       "3                                +                             NaN   \n",
       "4                                +                             NaN   \n",
       "5                                +                             NaN   \n",
       "6                                +                             NaN   \n",
       "7                                +                             NaN   \n",
       "8                                +                             NaN   \n",
       "\n",
       "  High palate Low-set ears Broad thumb Brachydactyly Obesity  \\\n",
       "0           +            +           +             +       +   \n",
       "1         NaN          NaN         NaN           NaN     NaN   \n",
       "2         NaN          NaN         NaN             +     NaN   \n",
       "3         NaN          NaN         NaN           NaN     NaN   \n",
       "4         NaN          NaN         NaN           NaN     NaN   \n",
       "5         NaN          NaN         NaN           NaN     NaN   \n",
       "6         NaN          NaN         NaN           NaN     NaN   \n",
       "7         NaN          NaN         NaN           NaN     NaN   \n",
       "8         NaN          NaN         NaN           NaN     NaN   \n",
       "\n",
       "  Large for gestational age Postaxial polydactyly Inguinal hernia  \n",
       "0                         -                   NaN             NaN  \n",
       "1                         +                     +             NaN  \n",
       "2                       NaN                   NaN               +  \n",
       "3                       NaN                   NaN             NaN  \n",
       "4                       NaN                   NaN             NaN  \n",
       "5                       NaN                   NaN             NaN  \n",
       "6                       NaN                   NaN             NaN  \n",
       "7                       NaN                   NaN             NaN  \n",
       "8                       NaN                   NaN             NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8314d6bd",
   "metadata": {},
   "source": [
    "<h2>Column mappers</h2>\n",
    "<p>Please see the notebook \"Create phenopackets from tabular data with individuals in rows\" for explanations. In the following cell we create a dictionary for the ColumnMappers. Note that the code is identical except that we use the df.loc function to get the corresponding row data</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2305a0d2",
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                      term    status\n",
      "0  Global developmental delay (HP:0001263)  observed\n",
      "1  Global developmental delay (HP:0001263)  observed\n",
      "2  Global developmental delay (HP:0001263)  observed\n",
      "3  Global developmental delay (HP:0001263)  observed\n",
      "4  Global developmental delay (HP:0001263)  observed\n",
      "5  Global developmental delay (HP:0001263)  observed\n",
      "6  Global developmental delay (HP:0001263)  observed\n",
      "7  Global developmental delay (HP:0001263)  observed\n",
      "8  Global developmental delay (HP:0001263)  observed\n",
      "                                                  term    status\n",
      "0  Abnormal respiratory system physiology (HP:0002795)  observed\n",
      "1  Abnormal respiratory system physiology (HP:0002795)  observed\n",
      "2  Abnormal respiratory system physiology (HP:0002795)  observed\n",
      "3  Abnormal respiratory system physiology (HP:0002795)  observed\n",
      "4  Abnormal respiratory system physiology (HP:0002795)  observed\n",
      "5  Abnormal respiratory system physiology (HP:0002795)  observed\n",
      "6  Abnormal respiratory system physiology (HP:0002795)  observed\n",
      "7  Abnormal respiratory system physiology (HP:0002795)  observed\n",
      "8  Abnormal respiratory system physiology (HP:0002795)  observed\n",
      "                        term    status\n",
      "0  Macrocephaly (HP:0000256)  observed\n",
      "1  Macrocephaly (HP:0000256)  observed\n",
      "2  Macrocephaly (HP:0000256)  observed\n",
      "3  Macrocephaly (HP:0000256)  observed\n",
      "4  Macrocephaly (HP:0000256)  observed\n",
      "5  Macrocephaly (HP:0000256)  observed\n",
      "6  Macrocephaly (HP:0000256)  observed\n",
      "7  Macrocephaly (HP:0000256)  observed\n",
      "8  Macrocephaly (HP:0000256)  observed\n",
      "                                            term    status\n",
      "0  Recurrent respiratory infections (HP:0002205)  observed\n",
      "1  Recurrent respiratory infections (HP:0002205)  observed\n",
      "2  Recurrent respiratory infections (HP:0002205)  observed\n",
      "3  Recurrent respiratory infections (HP:0002205)  observed\n",
      "4  Recurrent respiratory infections (HP:0002205)  observed\n",
      "5  Recurrent respiratory infections (HP:0002205)  observed\n",
      "6  Recurrent respiratory infections (HP:0002205)  observed\n",
      "7  Recurrent respiratory infections (HP:0002205)  observed\n",
      "8  Recurrent respiratory infections (HP:0002205)  observed\n",
      "                                   term        status\n",
      "0  Intellectual disability (HP:0001249)      observed\n",
      "1  Intellectual disability (HP:0001249)      observed\n",
      "2  Intellectual disability (HP:0001249)  not measured\n",
      "3  Intellectual disability (HP:0001249)  not measured\n",
      "4  Intellectual disability (HP:0001249)  not measured\n",
      "5  Intellectual disability (HP:0001249)  not measured\n",
      "6  Intellectual disability (HP:0001249)  not measured\n",
      "7  Intellectual disability (HP:0001249)  not measured\n",
      "8  Intellectual disability (HP:0001249)  not measured\n",
      "                       term        status\n",
      "0  High palate (HP:0000218)      observed\n",
      "1  High palate (HP:0000218)  not measured\n",
      "2  High palate (HP:0000218)  not measured\n",
      "3  High palate (HP:0000218)  not measured\n",
      "4  High palate (HP:0000218)  not measured\n",
      "5  High palate (HP:0000218)  not measured\n",
      "6  High palate (HP:0000218)  not measured\n",
      "7  High palate (HP:0000218)  not measured\n",
      "8  High palate (HP:0000218)  not measured\n",
      "                        term        status\n",
      "0  Low-set ears (HP:0000369)      observed\n",
      "1  Low-set ears (HP:0000369)  not measured\n",
      "2  Low-set ears (HP:0000369)  not measured\n",
      "3  Low-set ears (HP:0000369)  not measured\n",
      "4  Low-set ears (HP:0000369)  not measured\n",
      "5  Low-set ears (HP:0000369)  not measured\n",
      "6  Low-set ears (HP:0000369)  not measured\n",
      "7  Low-set ears (HP:0000369)  not measured\n",
      "8  Low-set ears (HP:0000369)  not measured\n",
      "                       term        status\n",
      "0  Broad thumb (HP:0011304)      observed\n",
      "1  Broad thumb (HP:0011304)  not measured\n",
      "2  Broad thumb (HP:0011304)  not measured\n",
      "3  Broad thumb (HP:0011304)  not measured\n",
      "4  Broad thumb (HP:0011304)  not measured\n",
      "5  Broad thumb (HP:0011304)  not measured\n",
      "6  Broad thumb (HP:0011304)  not measured\n",
      "7  Broad thumb (HP:0011304)  not measured\n",
      "8  Broad thumb (HP:0011304)  not measured\n",
      "                         term        status\n",
      "0  Brachydactyly (HP:0001156)      observed\n",
      "1  Brachydactyly (HP:0001156)  not measured\n",
      "2  Brachydactyly (HP:0001156)      observed\n",
      "3  Brachydactyly (HP:0001156)  not measured\n",
      "4  Brachydactyly (HP:0001156)  not measured\n",
      "5  Brachydactyly (HP:0001156)  not measured\n",
      "6  Brachydactyly (HP:0001156)  not measured\n",
      "7  Brachydactyly (HP:0001156)  not measured\n",
      "8  Brachydactyly (HP:0001156)  not measured\n",
      "                   term        status\n",
      "0  Obesity (HP:0001513)      observed\n",
      "1  Obesity (HP:0001513)  not measured\n",
      "2  Obesity (HP:0001513)  not measured\n",
      "3  Obesity (HP:0001513)  not measured\n",
      "4  Obesity (HP:0001513)  not measured\n",
      "5  Obesity (HP:0001513)  not measured\n",
      "6  Obesity (HP:0001513)  not measured\n",
      "7  Obesity (HP:0001513)  not measured\n",
      "8  Obesity (HP:0001513)  not measured\n",
      "                                     term        status\n",
      "0  Large for gestational age (HP:0001520)      excluded\n",
      "1  Large for gestational age (HP:0001520)      observed\n",
      "2  Large for gestational age (HP:0001520)  not measured\n",
      "3  Large for gestational age (HP:0001520)  not measured\n",
      "4  Large for gestational age (HP:0001520)  not measured\n",
      "5  Large for gestational age (HP:0001520)  not measured\n",
      "6  Large for gestational age (HP:0001520)  not measured\n",
      "7  Large for gestational age (HP:0001520)  not measured\n",
      "8  Large for gestational age (HP:0001520)  not measured\n",
      "                                 term        status\n",
      "0  Postaxial polydactyly (HP:0100259)  not measured\n",
      "1  Postaxial polydactyly (HP:0100259)      observed\n",
      "2  Postaxial polydactyly (HP:0100259)  not measured\n",
      "3  Postaxial polydactyly (HP:0100259)  not measured\n",
      "4  Postaxial polydactyly (HP:0100259)  not measured\n",
      "5  Postaxial polydactyly (HP:0100259)  not measured\n",
      "6  Postaxial polydactyly (HP:0100259)  not measured\n",
      "7  Postaxial polydactyly (HP:0100259)  not measured\n",
      "8  Postaxial polydactyly (HP:0100259)  not measured\n",
      "                           term        status\n",
      "0  Inguinal hernia (HP:0000023)  not measured\n",
      "1  Inguinal hernia (HP:0000023)  not measured\n",
      "2  Inguinal hernia (HP:0000023)      observed\n",
      "3  Inguinal hernia (HP:0000023)  not measured\n",
      "4  Inguinal hernia (HP:0000023)  not measured\n",
      "5  Inguinal hernia (HP:0000023)  not measured\n",
      "6  Inguinal hernia (HP:0000023)  not measured\n",
      "7  Inguinal hernia (HP:0000023)  not measured\n",
      "8  Inguinal hernia (HP:0000023)  not measured\n"
     ]
    }
   ],
   "source": [
    "hpo_cr = parser.get_hpo_concept_recognizer()\n",
    "column_mapper_d, col_not_found = try_mapping_columns(df=df,\n",
    "                                                    observed='+',\n",
    "                                                    excluded='-',\n",
    "                                                    hpo_cr=hpo_cr,\n",
    "                                                    preview=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fa281c54",
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['individual', 'sex', 'age', 'variant']\n"
     ]
    }
   ],
   "source": [
    "print(col_not_found)        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e53617d",
   "metadata": {},
   "source": [
    "<h2>Variant Data</h2>\n",
    "<p>The variant data (HGVS< transcript) is listed in the Variant (hg19, NM_015133.4) column.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "da4d5706",
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [],
   "source": [
    "genome = 'hg19'\n",
    "default_genotype = 'hemizygous'\n",
    "transcript='NM_003611.3'\n",
    "varMapper = VariantColumnMapper(assembly=genome,column_name='variant', \n",
    "                                transcript=transcript, default_genotype=default_genotype)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1af2052",
   "metadata": {},
   "source": [
    "<h1>Demographic data</h1>\n",
    "<p>pyphetools can be used to capture information about age, sex, and individual identifiers. This information is stored in a map of \"IndividualMapper\" objects. Special treatment may be required for the indifiers, which may be used as the column names or row index.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3e64dc08",
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original column contents</th>\n",
       "      <th>age</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>P11Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>P0Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   original column contents   age\n",
       "0                        11  P11Y\n",
       "1                         0   P0Y"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ageMapper = AgeColumnMapper.by_year('age')\n",
    "ageMapper.preview_column(df['age'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "71f664cc",
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original column contents</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>male</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>male</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>male</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>male</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>male</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>male</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>male</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>male</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>male</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  original column contents   sex\n",
       "0                     male  MALE\n",
       "1                     male  MALE\n",
       "2                     male  MALE\n",
       "3                     male  MALE\n",
       "4                     male  MALE\n",
       "5                     male  MALE\n",
       "6                     male  MALE\n",
       "7                     male  MALE\n",
       "8                     male  MALE"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sexMapper = SexColumnMapper(male_symbol='male', female_symbol='female', column_name='sex')\n",
    "sexMapper.preview_column(df['sex'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f6581a8a",
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [],
   "source": [
    "pmid = \"PMID:16783569\"\n",
    "encoder = CohortEncoder(df=df, hpo_cr=hpo_cr, column_mapper_d=column_mapper_d, \n",
    "                        individual_column_name=\"individual\", agemapper=ageMapper, sexmapper=sexMapper,\n",
    "                       variant_mapper=varMapper, metadata=metadata,\n",
    "                       pmid=pmid)\n",
    "encoder.set_disease(disease_id='300209', label='Simpson-Golabi-Behmel syndrome, type 2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fd367ed6",
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NM_003611.3%3Ac.2122_2125dupGAAG/NM_003611.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NM_003611.3%3Ac.2122_2125dupGAAG/NM_003611.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NM_003611.3%3Ac.2122_2125dupGAAG/NM_003611.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NM_003611.3%3Ac.2122_2125dupGAAG/NM_003611.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NM_003611.3%3Ac.2122_2125dupGAAG/NM_003611.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NM_003611.3%3Ac.2122_2125dupGAAG/NM_003611.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NM_003611.3%3Ac.2122_2125dupGAAG/NM_003611.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NM_003611.3%3Ac.2122_2125dupGAAG/NM_003611.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NM_003611.3%3Ac.2122_2125dupGAAG/NM_003611.3?content-type=application%2Fjson\n"
     ]
    }
   ],
   "source": [
    "individuals = encoder.get_individuals()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5d044b78",
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"id\": \"9\",\n",
      "  \"subject\": {\n",
      "    \"id\": \"9\",\n",
      "    \"timeAtLastEncounter\": {\n",
      "      \"age\": {\n",
      "        \"iso8601duration\": \"P0Y\"\n",
      "      }\n",
      "    },\n",
      "    \"sex\": \"MALE\"\n",
      "  },\n",
      "  \"phenotypicFeatures\": [\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0001263\",\n",
      "        \"label\": \"Global developmental delay\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0002795\",\n",
      "        \"label\": \"Abnormal respiratory system physiology\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0000256\",\n",
      "        \"label\": \"Macrocephaly\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": {\n",
      "        \"id\": \"HP:0002205\",\n",
      "        \"label\": \"Recurrent respiratory infections\"\n",
      "      }\n",
      "    }\n",
      "  ],\n",
      "  \"interpretations\": [\n",
      "    {\n",
      "      \"id\": \"9\",\n",
      "      \"progressStatus\": \"SOLVED\",\n",
      "      \"diagnosis\": {\n",
      "        \"disease\": {\n",
      "          \"id\": \"300209\",\n",
      "          \"label\": \"Simpson-Golabi-Behmel syndrome, type 2\"\n",
      "        },\n",
      "        \"genomicInterpretations\": [\n",
      "          {\n",
      "            \"subjectOrBiosampleId\": \"9\",\n",
      "            \"interpretationStatus\": \"CAUSATIVE\",\n",
      "            \"variantInterpretation\": {\n",
      "              \"variationDescriptor\": {\n",
      "                \"id\": \"var_syxlBTOPWsgJkEpbzfRaaZawK\",\n",
      "                \"geneContext\": {\n",
      "                  \"valueId\": \"HGNC:2567\",\n",
      "                  \"symbol\": \"OFD1\"\n",
      "                },\n",
      "                \"expressions\": [\n",
      "                  {\n",
      "                    \"syntax\": \"hgvs.c\",\n",
      "                    \"value\": \"NM_003611.3:c.2122_2125dup\"\n",
      "                  },\n",
      "                  {\n",
      "                    \"syntax\": \"hgvs.g\",\n",
      "                    \"value\": \"NC_000023.10:g.13778701_13778704dup\"\n",
      "                  }\n",
      "                ],\n",
      "                \"vcfRecord\": {\n",
      "                  \"genomeAssembly\": \"hg19\",\n",
      "                  \"chrom\": \"chrX\",\n",
      "                  \"pos\": \"13778700\",\n",
      "                  \"ref\": \"T\",\n",
      "                  \"alt\": \"TGAAG\"\n",
      "                },\n",
      "                \"moleculeContext\": \"genomic\",\n",
      "                \"allelicState\": {\n",
      "                  \"id\": \"GENO:0000134\",\n",
      "                  \"label\": \"hemizygous\"\n",
      "                }\n",
      "              }\n",
      "            }\n",
      "          }\n",
      "        ]\n",
      "      }\n",
      "    }\n",
      "  ],\n",
      "  \"metaData\": {\n",
      "    \"created\": \"2023-07-02T17:27:20.217549085Z\",\n",
      "    \"createdBy\": \"ORCID:0000-0002-5648-2155\",\n",
      "    \"resources\": [\n",
      "      {\n",
      "        \"id\": \"geno\",\n",
      "        \"name\": \"Genotype Ontology\",\n",
      "        \"url\": \"http://purl.obolibrary.org/obo/geno.owl\",\n",
      "        \"version\": \"2022-03-05\",\n",
      "        \"namespacePrefix\": \"GENO\",\n",
      "        \"iriPrefix\": \"http://purl.obolibrary.org/obo/GENO_\"\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"hgnc\",\n",
      "        \"name\": \"HUGO Gene Nomenclature Committee\",\n",
      "        \"url\": \"https://www.genenames.org\",\n",
      "        \"version\": \"06/01/23\",\n",
      "        \"namespacePrefix\": \"HGNC\",\n",
      "        \"iriPrefix\": \"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/\"\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"omim\",\n",
      "        \"name\": \"An Online Catalog of Human Genes and Genetic Disorders\",\n",
      "        \"url\": \"https://www.omim.org\",\n",
      "        \"version\": \"January 4, 2023\",\n",
      "        \"namespacePrefix\": \"OMIM\",\n",
      "        \"iriPrefix\": \"https://www.omim.org/entry/\"\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"so\",\n",
      "        \"name\": \"Sequence types and features ontology\",\n",
      "        \"url\": \"http://purl.obolibrary.org/obo/so.obo\",\n",
      "        \"version\": \"2021-11-22\",\n",
      "        \"namespacePrefix\": \"SO\",\n",
      "        \"iriPrefix\": \"http://purl.obolibrary.org/obo/SO_\"\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"hp\",\n",
      "        \"name\": \"human phenotype ontology\",\n",
      "        \"url\": \"http://purl.obolibrary.org/obo/hp.owl\",\n",
      "        \"version\": \"2023-06-17\",\n",
      "        \"namespacePrefix\": \"HP\",\n",
      "        \"iriPrefix\": \"http://purl.obolibrary.org/obo/HP_\"\n",
      "      }\n",
      "    ],\n",
      "    \"phenopacketSchemaVersion\": \"2.0\"\n",
      "  }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "i1 = individuals[-1]\n",
    "phenopacket1 = i1.to_ga4gh_phenopacket(metadata=metadata.to_ga4gh())\n",
    "json_string = MessageToJson(phenopacket1)\n",
    "print(json_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "23f1094f",
   "metadata": {
    "pycharm": {
     "is_executing": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NM_003611.3%3Ac.2122_2125dupGAAG/NM_003611.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NM_003611.3%3Ac.2122_2125dupGAAG/NM_003611.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NM_003611.3%3Ac.2122_2125dupGAAG/NM_003611.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NM_003611.3%3Ac.2122_2125dupGAAG/NM_003611.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NM_003611.3%3Ac.2122_2125dupGAAG/NM_003611.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NM_003611.3%3Ac.2122_2125dupGAAG/NM_003611.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NM_003611.3%3Ac.2122_2125dupGAAG/NM_003611.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NM_003611.3%3Ac.2122_2125dupGAAG/NM_003611.3?content-type=application%2Fjson\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NM_003611.3%3Ac.2122_2125dupGAAG/NM_003611.3?content-type=application%2Fjson\n",
      "Wrote 9 phenopackets to ../../phenopackets/OFD1/\n"
     ]
    }
   ],
   "source": [
    "output_directory = \"../../phenopackets/OFD1/\"\n",
    "encoder.output_phenopackets(outdir=output_directory)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
