{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "34284130",
   "metadata": {},
   "source": [
    "<h1>landscape of STXBP1-related disorders </h1>\n",
    "<p>Extract the clinical data from <a href=\"https://pubmed.ncbi.nlm.nih.gov/35190816/\"target=\"__blank\">Xian et al. (2022) Assessing the landscape of STXBP1-related disorders in 534 individuals. Brain.</a>.<p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0ecc4b45",
   "metadata": {},
   "outputs": [],
   "source": [
    "import phenopackets as php\n",
    "from google.protobuf.json_format import MessageToDict, MessageToJson\n",
    "from google.protobuf.json_format import Parse, ParseDict\n",
    "import pandas as pd\n",
    "from csv import DictReader\n",
    "pd.set_option('display.max_colwidth', None) # show entire column contents, important!\n",
    "from collections import defaultdict\n",
    "import re\n",
    "from pyphetools.creation import *\n",
    "# last tested with pyphetools version 0.2.22"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9e8891ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = HpoParser()\n",
    "hpo_cr = parser.get_hpo_concept_recognizer()\n",
    "hpo_version = parser.get_version()\n",
    "metadata = MetaData(created_by=\"ORCID:0000-0002-0736-9199\")\n",
    "metadata.default_versions_with_hpo(version=hpo_version)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "9dc8c7f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinical_df = pd.read_table(\"input/brain-2021-00642-File011.tsv\");\n",
    "genotype_df = pd.read_table(\"input/brain-2021-00642-File011-genotype.tsv\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "717ba1d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PatID</th>\n",
       "      <th>Source_Journal</th>\n",
       "      <th>Source_PMID*</th>\n",
       "      <th>Year</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Phenotypic_group**</th>\n",
       "      <th>age_onset_m</th>\n",
       "      <th>age_offset_m</th>\n",
       "      <th>age_eval_y</th>\n",
       "      <th>Base_HPO***</th>\n",
       "      <th>HPO_term</th>\n",
       "      <th>Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STX_18469812_Subject_11</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>18469812</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>M</td>\n",
       "      <td>EOEE</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>HP:0003593</td>\n",
       "      <td>Infantile onset</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STX_18469812_Subject_11</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>18469812</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>M</td>\n",
       "      <td>EOEE</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>HP:0010818</td>\n",
       "      <td>Generalized tonic seizures</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STX_18469812_Subject_11</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>18469812</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>M</td>\n",
       "      <td>EOEE</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>HP:0002069</td>\n",
       "      <td>Generalized tonic-clonic seizures</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STX_18469812_Subject_11</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>18469812</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>M</td>\n",
       "      <td>EOEE</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>HP:0010851</td>\n",
       "      <td>EEG with burst suppression</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STX_18469812_Subject_11</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>18469812</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>M</td>\n",
       "      <td>EOEE</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>HP:0002521</td>\n",
       "      <td>Hypsarrhythmia</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     PatID Source_Journal Source_PMID*    Year Sex  \\\n",
       "0  STX_18469812_Subject_11      Nat Genet     18469812  2008.0   M   \n",
       "1  STX_18469812_Subject_11      Nat Genet     18469812  2008.0   M   \n",
       "2  STX_18469812_Subject_11      Nat Genet     18469812  2008.0   M   \n",
       "3  STX_18469812_Subject_11      Nat Genet     18469812  2008.0   M   \n",
       "4  STX_18469812_Subject_11      Nat Genet     18469812  2008.0   M   \n",
       "\n",
       "  Phenotypic_group**  age_onset_m  age_offset_m  age_eval_y Base_HPO***  \\\n",
       "0               EOEE          2.0           NaN         8.0  HP:0003593   \n",
       "1               EOEE          2.0           NaN         8.0  HP:0010818   \n",
       "2               EOEE          2.0           NaN         8.0  HP:0002069   \n",
       "3               EOEE          2.0           NaN         8.0  HP:0010851   \n",
       "4               EOEE          2.0           NaN         8.0  HP:0002521   \n",
       "\n",
       "                            HPO_term Notes  \n",
       "0                    Infantile onset   NaN  \n",
       "1         Generalized tonic seizures   NaN  \n",
       "2  Generalized tonic-clonic seizures   NaN  \n",
       "3         EEG with burst suppression   NaN  \n",
       "4                     Hypsarrhythmia   NaN  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "a938b319",
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_d = defaultdict(list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aadcdbea",
   "metadata": {},
   "source": [
    "<h2>HpoTerm</h2>\n",
    "<p>We create classes called HpoTerm and PatientRow to help process the data from the input file. Note that\n",
    "the data for one patient is distributed across multiple rows of the input Excel file</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "c644123b",
   "metadata": {},
   "outputs": [],
   "source": [
    "class HpoTerm:\n",
    "    def __init__(self, row):\n",
    "        try: \n",
    "            self.age_onset_m = int(row[\"age_onset_m\"])\n",
    "        except:\n",
    "            self.age_onset_m = None\n",
    "        try:\n",
    "            self.age_offset_m = int(row[\"age_offset_m\"])\n",
    "        except:\n",
    "            self.age_offset_m = None\n",
    "        try:\n",
    "            self.age_eval_y = int(row[\"age_eval_y\"])\n",
    "        except:\n",
    "            self.age_eval_y = None\n",
    "        self.hpo_id = row[\"Base_HPO***\"]\n",
    "        self.hpo_label = row[\"HPO_term\"]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "class PatientRow:\n",
    "    def __init__(self, row):\n",
    "        self.patID = row[\"PatID\"]\n",
    "        self.sex = row[\"Sex\"]\n",
    "        self.phenogroup = row[\"Phenotypic_group**\"]\n",
    "        self.hpo_term = HpoTerm(row=row)\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "c657801d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We extracted data on 534 individuals\n"
     ]
    }
   ],
   "source": [
    "with open(\"input/brain-2021-00642-File011.tsv\") as f:\n",
    "    reader = DictReader(f, delimiter=\"\\t\")\n",
    "    for row in reader:\n",
    "        prow = PatientRow(row=row)\n",
    "        patient_d[prow.patID].append(prow)\n",
    "print(f\"We extracted data on {len(patient_d)} individuals\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "55a565a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PatID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Func.refGeneWithVer</th>\n",
       "      <th>Gene.refGeneWithVer</th>\n",
       "      <th>GeneDetail.refGeneWithVer</th>\n",
       "      <th>ExonicFunc.refGeneWithVer</th>\n",
       "      <th>...</th>\n",
       "      <th>Otherinfo3</th>\n",
       "      <th>Otherinfo4</th>\n",
       "      <th>Otherinfo5</th>\n",
       "      <th>Otherinfo7</th>\n",
       "      <th>Otherinfo8</th>\n",
       "      <th>Otherinfo9</th>\n",
       "      <th>Otherinfo10</th>\n",
       "      <th>Otherinfo11</th>\n",
       "      <th>bed</th>\n",
       "      <th>Unnamed: 134</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STX_18469812_Subject_11</td>\n",
       "      <td>9.0</td>\n",
       "      <td>130422313.0</td>\n",
       "      <td>130422313.0</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>STXBP1</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>9</td>\n",
       "      <td>130422313</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>Name=70.695764689</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STX_18469812_Subject_3</td>\n",
       "      <td>9.0</td>\n",
       "      <td>130444768.0</td>\n",
       "      <td>130444768.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>STXBP1</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>9</td>\n",
       "      <td>130444768</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>Name=75.862552050</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STX_18469812_Subject_6</td>\n",
       "      <td>9.0</td>\n",
       "      <td>130425593.0</td>\n",
       "      <td>130425593.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>STXBP1</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>9</td>\n",
       "      <td>130425593</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>Name=89.193399398</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STX_18469812_Subject_7</td>\n",
       "      <td>9.0</td>\n",
       "      <td>130439001.0</td>\n",
       "      <td>130439001.0</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>exonic</td>\n",
       "      <td>STXBP1</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>9</td>\n",
       "      <td>130439001</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>Name=20.833326089</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STX_19557857_Patient_1</td>\n",
       "      <td>9.0</td>\n",
       "      <td>130416076.0</td>\n",
       "      <td>130416076.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>splicing</td>\n",
       "      <td>STXBP1</td>\n",
       "      <td>NM_001032221.3:exon3:c.169+1G&gt;A;NM_003165.3:exon3:c.169+1G&gt;A</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>9</td>\n",
       "      <td>130416076</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 135 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     PatID  Chr        Start          End Ref Alt  \\\n",
       "0  STX_18469812_Subject_11  9.0  130422313.0  130422313.0   T   A   \n",
       "1   STX_18469812_Subject_3  9.0  130444768.0  130444768.0   G   A   \n",
       "2   STX_18469812_Subject_6  9.0  130425593.0  130425593.0   G   A   \n",
       "3   STX_18469812_Subject_7  9.0  130439001.0  130439001.0   T   G   \n",
       "4   STX_19557857_Patient_1  9.0  130416076.0  130416076.0   G   A   \n",
       "\n",
       "  Func.refGeneWithVer Gene.refGeneWithVer  \\\n",
       "0              exonic              STXBP1   \n",
       "1              exonic              STXBP1   \n",
       "2              exonic              STXBP1   \n",
       "3              exonic              STXBP1   \n",
       "4            splicing              STXBP1   \n",
       "\n",
       "                                      GeneDetail.refGeneWithVer  \\\n",
       "0                                                             .   \n",
       "1                                                             .   \n",
       "2                                                             .   \n",
       "3                                                             .   \n",
       "4  NM_001032221.3:exon3:c.169+1G>A;NM_003165.3:exon3:c.169+1G>A   \n",
       "\n",
       "  ExonicFunc.refGeneWithVer  ... Otherinfo3 Otherinfo4 Otherinfo5 Otherinfo7  \\\n",
       "0         nonsynonymous SNV  ...          .          9  130422313          T   \n",
       "1         nonsynonymous SNV  ...          .          9  130444768          G   \n",
       "2         nonsynonymous SNV  ...          .          9  130425593          G   \n",
       "3         nonsynonymous SNV  ...          .          9  130439001          T   \n",
       "4                         .  ...          .          9  130416076          G   \n",
       "\n",
       "  Otherinfo8 Otherinfo9 Otherinfo10 Otherinfo11                bed  \\\n",
       "0          A          .        PASS           .  Name=70.695764689   \n",
       "1          A          .        PASS           .  Name=75.862552050   \n",
       "2          A          .        PASS           .  Name=89.193399398   \n",
       "3          G          .        PASS           .  Name=20.833326089   \n",
       "4          A          .        PASS           .                  .   \n",
       "\n",
       "  Unnamed: 134  \n",
       "0          NaN  \n",
       "1          NaN  \n",
       "2          NaN  \n",
       "3          NaN  \n",
       "4          NaN  \n",
       "\n",
       "[5 rows x 135 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "b667810e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_var_inf(aachange):\n",
    "    \"\"\"\n",
    "    aachange: e.g., STXBP1:NM_001032221.3:exon6:c.T353G:p.L118R,STXBP1:NM_003165.3:exon6:c.T353G:p.L118R\n",
    "    \"\"\"\n",
    "    fields = aachange.split(\":\")\n",
    "    index = 0\n",
    "    i = 0\n",
    "    transcript = \"?\"\n",
    "    for f in fields:\n",
    "        if f == \"NM_001032221.3\" or f == \"NM_003165.3\":\n",
    "            index = i\n",
    "            transcript = f\n",
    "            break\n",
    "        i += 1\n",
    "    if (i + 2) < len(fields):\n",
    "        variant = fields[i+2]\n",
    "    else:\n",
    "        raise ValueError(f\"Could not get variant because of fields: {fields} and i={i}\")   \n",
    "    return transcript, variant\n",
    "\n",
    "def extract_splice_var(genedetail):\n",
    "    \"\"\"\n",
    "    genedetail: e.g., NM_001032221.3:exon3:c.169+1G>A;NM_003165.3:exon3:c.169+1G>A\n",
    "    \"\"\"\n",
    "    fields = genedetail.split(\":\")\n",
    "    index = 0\n",
    "    i = 0\n",
    "    transcript = \"?\"\n",
    "    for f in fields:\n",
    "        if f == \"NM_001032221.3\" or f == \"NM_003165.3\":\n",
    "            index = i\n",
    "            transcript = f\n",
    "            break\n",
    "        i += 1\n",
    "    if (i + 2) < len(fields):\n",
    "        variant = fields[i+2]\n",
    "    else:\n",
    "        raise ValueError(f\"Could not get variant because of fields: {fields} and i={i}\")   \n",
    "    return transcript, variant\n",
    "\n",
    "\n",
    "class GenotypeEntry:\n",
    "    def __init__(self, row):\n",
    "        self.patID = row[\"PatID\"]\n",
    "        self.chrom = row[\"Chr\"]\n",
    "        self.start = row[\"Start\"]\n",
    "        self.end = row[\"End\"]\n",
    "        self.ref = row[\"Ref\"]\n",
    "        self.alt = row[\"Alt\"]\n",
    "        self.transcript = \"?\"\n",
    "        func = row[\"Func.refGeneWithVer\"]\n",
    "        genenot = row[\"Gene.refGeneWithVer\"]\n",
    "        aachange = row[\"AAChange.refGeneWithVer\"]\n",
    "        if func == \"exonic\":\n",
    "            transcript, variant = extract_var_inf(aachange)\n",
    "            self.transcript = transcript\n",
    "            \n",
    "            regex_del = r\"c.\\d+_\\d+del\"\n",
    "            regex_single_nt_del = r\"(c.\\d+del)[ACGT]\"\n",
    "            regex_dup = r\"c.(\\d+)dup([A-Z]+)\"\n",
    "            regex_sub = r\"c.([A-Z]+)(\\d+)([A-Z]+)\"\n",
    "            regex_ins = r\"c.(\\d+)_(\\d+)ins([A-Z]+)\"  # e.g., 1372_1373insGCCGGAGCAA\n",
    "            result = re.search(regex_sub, variant)\n",
    "            result_dup = re.search(regex_dup, variant)\n",
    "            result_single_nt_del = re.search(regex_single_nt_del, variant)\n",
    "            result_ins = re.search(regex_ins, variant)\n",
    "            if re.match(regex_del, variant):\n",
    "                self.hgvs = variant\n",
    "            elif result:\n",
    "                ref = result.group(1)\n",
    "                position = result.group(2)\n",
    "                alt = result.group(3)\n",
    "                hgvs = f\"c.{position}{ref}>{alt}\"\n",
    "                self.hgvs = hgvs\n",
    "            elif result_dup:\n",
    "                position=result_dup.group(1)\n",
    "                hgvs = f\"c.{position}dup\"\n",
    "                self.hgvs = hgvs\n",
    "            elif result_single_nt_del:\n",
    "                self.hgvs = result_single_nt_del.group(1)\n",
    "            elif result_ins:\n",
    "                pos1 = result_ins.group(1)\n",
    "                pos2 = result_ins.group(2)\n",
    "                seq = result_ins.group(3)\n",
    "                self.hgvs = f\"c.{pos1}_{pos2}ins{seq}\"\n",
    "            else:\n",
    "                raise ValueError(f\"Could not parse variant {variant}\")\n",
    "        elif func == 'splicing':\n",
    "            print(row[\"GeneDetail.refGeneWithVer\"])\n",
    "            transcript, variant = extract_splice_var(aachange)\n",
    "            self.transcript = transcript\n",
    "            self.hgvs = variant\n",
    "            print(variant)\n",
    "            \n",
    "            \n",
    "        else:\n",
    "            raise ValueError(f\"Could not parse variant  for func {func}\\n{row}\")\n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "198cbd90",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NM_001032221.3:exon3:c.169+1G>A;NM_003165.3:exon3:c.169+1G>A\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Could not get variant because of fields: ['.'] and i=1",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[73], line 5\u001b[0m\n\u001b[1;32m      3\u001b[0m reader \u001b[38;5;241m=\u001b[39m DictReader(f, delimiter\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m row \u001b[38;5;129;01min\u001b[39;00m reader:\n\u001b[0;32m----> 5\u001b[0m     ge \u001b[38;5;241m=\u001b[39m \u001b[43mGenotypeEntry\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrow\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrow\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[72], line 90\u001b[0m, in \u001b[0;36mGenotypeEntry.__init__\u001b[0;34m(self, row)\u001b[0m\n\u001b[1;32m     88\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m func \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124msplicing\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[1;32m     89\u001b[0m     \u001b[38;5;28mprint\u001b[39m(row[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mGeneDetail.refGeneWithVer\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[0;32m---> 90\u001b[0m     transcript, variant \u001b[38;5;241m=\u001b[39m \u001b[43mextract_splice_var\u001b[49m\u001b[43m(\u001b[49m\u001b[43maachange\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     91\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtranscript \u001b[38;5;241m=\u001b[39m transcript\n\u001b[1;32m     92\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhgvs \u001b[38;5;241m=\u001b[39m variant\n",
      "Cell \u001b[0;32mIn[72], line 38\u001b[0m, in \u001b[0;36mextract_splice_var\u001b[0;34m(genedetail)\u001b[0m\n\u001b[1;32m     36\u001b[0m     variant \u001b[38;5;241m=\u001b[39m fields[i\u001b[38;5;241m+\u001b[39m\u001b[38;5;241m2\u001b[39m]\n\u001b[1;32m     37\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m---> 38\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCould not get variant because of fields: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfields\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m and i=\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mi\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)   \n\u001b[1;32m     39\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m transcript, variant\n",
      "\u001b[0;31mValueError\u001b[0m: Could not get variant because of fields: ['.'] and i=1"
     ]
    }
   ],
   "source": [
    "cc= 0\n",
    "with open(\"input/brain-2021-00642-File011-genotype.tsv\") as f:\n",
    "    reader = DictReader(f, delimiter=\"\\t\")\n",
    "    for row in reader:\n",
    "        ge = GenotypeEntry(row=row)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d2a0632",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09762f2a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7134daee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9079145b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60ffc759",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "c2csvenv",
   "language": "python",
   "name": "c2csvenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
